{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f92e2e4f-6704-4b3d-b8ec-c51427a7ca0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "474b1723-a2e0-4556-bd36-146ed07c1bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3c09f4af-4524-4a93-995f-a268efba3ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b571064-b110-4a4d-b182-bbcb19af6fc8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b3a514-cc78-40a4-87ff-0dfd817282e0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## List FASTQs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "022a6910-115b-4d46-b4bd-511b7a890893",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'MitoTracing': ['R1', 'R2'],\n",
    "}\n",
    "\n",
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae81c97b-0e5b-4aae-ae8a-977ca6384b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_barcode_genomic_fastqs(samples):\n",
    "    \n",
    "    samples['fastq'] = np.empty((len(samples), 0)).tolist()\n",
    "    fastqs = []\n",
    "    for sample, row in samples.iterrows():\n",
    "        S3_path = row['S3_path']\n",
    "        \n",
    "        # List FASTQ files\n",
    "        fastq_file_ids = fastq_map['MitoTracing']        \n",
    "        path_fastq = f\"{S3_path}/FASTQ/\"\n",
    "        fastq = get_fastqs(path_fastq, fastq_file_ids)\n",
    "        \n",
    "        fastqs.append(fastq)\n",
    "        \n",
    "    samples['fastq'] = fastqs\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd638296-c86e-4bdb-87ea-37983cdfbf89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract FASTQ sample name from list of files\n",
    "# Note: FASTQ name is file name up to lane id (e.g. L001, L002, etc.)\n",
    "def get_fastqs_name(fastqs):\n",
    "    fastq_name_re = r\".*/(.*)_S\\d+_L\\d{3}_[A-Za-z]\\d_\\d{3}.fastq.gz$\"\n",
    "    fastq_names = [re.match(fastq_name_re, x)[1] for x in fastqs]\n",
    "    assert len(set(fastq_names)) == 1 # make sure all names are same\n",
    "    return fastq_names[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6613242-152c-4b54-86c1-b245f3897c23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9773e79d-dfd6-4b33-9bd8-05781df1f038",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Get whitelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0b6e2570-8fae-4100-8741-ce5c8988ba64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_whitelist(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "):\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    whitelist = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\".txt$\")\n",
    "        )\n",
    "    whitelist = whitelist[0]\n",
    "    try:\n",
    "        whitelist = os.path.join(\"s3://\", bucket, whitelist)\n",
    "    except AssertionError as err:\n",
    "        logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "        return\n",
    "    return whitelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73b4978d-827f-45b7-8c39-ef8e5ecdb57a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5543869f-37e4-4d3b-a8d6-877973abb6ee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## DB queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3ca3bfa5-69fd-4d5f-9d45-3d5ac61c76c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "def get_sample_name(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Sample\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_names = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_names.append(result[0])\n",
    "        return sample_names\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "    \n",
    "    \n",
    "def get_aws_path(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_paths = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_paths.append(result[0])\n",
    "        return sample_paths\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "            \n",
    "def get_sample_id(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_ids = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_ids.append(result[0])\n",
    "        return sample_ids\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def format_sample_aws(querys, query_col, creds):\n",
    "    sample_names = []\n",
    "    sample_paths = []\n",
    "    sample_ids = []\n",
    "    \n",
    "    for query in querys:\n",
    "        sample_names += get_sample_name(query, query_col, creds)\n",
    "        sample_paths += get_aws_path(query, query_col, creds)\n",
    "        sample_ids += get_sample_id(query, query_col, creds)\n",
    "        \n",
    "    sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "    \n",
    "    samples = pd.DataFrame(\n",
    "        dict(S3_path=sample_paths, Sample_ID=sample_ids),\n",
    "        index=sample_names,\n",
    "        dtype=str,\n",
    "    )\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "25c817f6-9a33-4fcf-84b8-1d4ebbcb66ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomeIndex_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.genomeIndex_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            genomeIndex_id = result[0]\n",
    "            \n",
    "        return genomeIndex_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_scTech_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.scTech_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            scTech_id = result[0]\n",
    "            \n",
    "        return scTech_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_index(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.gIndex\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            index = result[0]\n",
    "            \n",
    "        return index\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_assay(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Run_name\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            assay = result[0]\n",
    "            \n",
    "        return assay\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_barcode(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.barcodes\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            barcode = result[0]\n",
    "            \n",
    "        return barcode\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "        \n",
    "def format_assay_barcode(samples, creds):\n",
    "    \n",
    "    # samples['genomeIndex_id'] = np.nan\n",
    "    samples['index'] = np.nan\n",
    "    # samples['scTech_id'] = np.nan\n",
    "    samples['assay'] = np.nan\n",
    "    samples['barcode'] = np.nan\n",
    "    \n",
    "    for sample, row in samples.iterrows():\n",
    "        sample_id = row['Sample_ID']\n",
    "        \n",
    "        genomeIndex_id  = get_genomeIndex_id(sample_id, 'id', creds)\n",
    "        index = get_index(genomeIndex_id, 'id', creds)\n",
    "        scTech_id = get_scTech_id(genomeIndex_id, 'id', creds)\n",
    "        assay = get_assay(scTech_id, 'id', creds)\n",
    "        barcode = get_barcode(scTech_id, 'id', creds)\n",
    "        \n",
    "        # samples.loc[sample, 'genomeIndex_id'] = genomeIndex_id\n",
    "        samples.loc[sample, 'index'] = index\n",
    "        # samples.loc[sample, 'scTech_id'] = scTech_id\n",
    "        samples.loc[sample, 'assay'] = assay\n",
    "        samples.loc[sample, 'barcode'] = barcode\n",
    "\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eb122238-79ee-4ecb-90d1-b2ea32e2b74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_project_id(sample_id, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fd74425-41d0-402c-9998-c3ed2c9f6aec",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c513b25-1067-4dfd-9ddc-092166012f7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e1c96c14-0cf1-4ab2-9beb-417bc02f1154",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"MitoTracing\": \"mito-tracing-outs\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"MitoTracing\": \"submit-processing.sh\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8ae3c158-3666-426f-94be-f9165e30c82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"MitoTracing\" # Workflow to run; also .wdl filename prefix\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "output_dirname = results_dirs[prefix]\n",
    "template_prefix = 'template'\n",
    "\n",
    "# If need to add comment, put here\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "61658425-6370-44c2-92c2-a3a977345bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = glob.glob(f\"{Path.home()}/scing/bin/*mito*\")[0]\n",
    "path_to_exec = sh_files[prefix]\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = glob.glob(f\"{workflow_dir}/*.options.aws.json\")[0]\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bb4d448a-5310-4b63-a073-66cf094e4467",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e2bc63-cf1f-4cf5-899b-f1e73c56f5aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2b28002d-85c9-4b9b-b0fc-5b86948c245c",
   "metadata": {},
   "source": [
    "# Excution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b988b1-df2f-4b32-a11b-786a188ba7b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2149bf16-9464-4654-bafd-a5b409f15321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_13_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             S3_path Sample_ID\n",
       "RA19_10_13_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...      3008\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...      3009\n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...      3010\n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...      3011\n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...      3012\n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...      3013"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "request_ids = ['BM-1244', ]\n",
    "\n",
    "samples = format_sample_aws(request_ids, 'request_id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "359fbd22-702d-4763-85c2-0410b0611bb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_13_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3008</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3009</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3010</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3011</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3012</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3013</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             S3_path  \\\n",
       "RA19_10_13_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            Sample_ID         index assay  barcode  \n",
       "RA19_10_13_FACs_MITO             3008  GRCh38-3.0.0    CR      NaN  \n",
       "RA19_10_14_FACs_MITO             3009  GRCh38-3.0.0    CR      NaN  \n",
       "RA19_10_17_FACs_MITO             3010  GRCh38-3.0.0    CR      NaN  \n",
       "RA19_10_18_FACs_MITO             3011  GRCh38-3.0.0    CR      NaN  \n",
       "RA19_10_23_FACs_MITO             3012  GRCh38-3.0.0    CR      NaN  \n",
       "RA19_10_14_FACS_citric_MITO      3013  GRCh38-3.0.0    CR      NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = format_assay_barcode(samples, creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a14b9418-dab4-4aaf-8abb-14f9e0534152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fastq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_13_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3008</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3009</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3010</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3011</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3012</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3013</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             S3_path  \\\n",
       "RA19_10_13_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            Sample_ID         index assay  barcode  \\\n",
       "RA19_10_13_FACs_MITO             3008  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_14_FACs_MITO             3009  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_17_FACs_MITO             3010  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_18_FACs_MITO             3011  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_23_FACs_MITO             3012  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_14_FACS_citric_MITO      3013  GRCh38-3.0.0    CR      NaN   \n",
       "\n",
       "                                                               fastq  \n",
       "RA19_10_13_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_14_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_17_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_18_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_23_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_14_FACS_citric_MITO  {'R1': ['s3://dp-lab-data/SCRI_Proje...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = get_barcode_genomic_fastqs(samples)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7fd5ce73-8d7f-4d23-9784-4774c7da80a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fastq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3009</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3010</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3011</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3012</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3013</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             S3_path  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            Sample_ID         index assay  barcode  \\\n",
       "RA19_10_14_FACs_MITO             3009  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_17_FACs_MITO             3010  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_18_FACs_MITO             3011  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_23_FACs_MITO             3012  GRCh38-3.0.0    CR      NaN   \n",
       "RA19_10_14_FACS_citric_MITO      3013  GRCh38-3.0.0    CR      NaN   \n",
       "\n",
       "                                                               fastq  \n",
       "RA19_10_14_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_17_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_18_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_23_FACs_MITO         {'R1': ['s3://dp-lab-data/SCRI_Proje...  \n",
       "RA19_10_14_FACS_citric_MITO  {'R1': ['s3://dp-lab-data/SCRI_Proje...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This sample has already been processed\n",
    "samples = samples[samples.index != 'RA19_10_13_FACs_MITO']\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9992a85d-6af9-449c-9eec-1ad4fc8aa2b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "85065328-a68e-46a7-9304-88953148416e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ce68536f-7029-4689-83a7-239421284398",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            MitoTracing.sampleName MitoTracing.fastqName  \\\n",
       "RA19_10_14_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_17_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_18_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_23_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_14_FACS_citric_MITO                    NaN                   NaN   \n",
       "\n",
       "                            MitoTracing.fastqR1 MitoTracing.fastqR2  \\\n",
       "RA19_10_14_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_17_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_18_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_23_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_14_FACS_citric_MITO                 NaN                 NaN   \n",
       "\n",
       "                            MitoTracing.reference MitoTracing.includeIntrons  \\\n",
       "RA19_10_14_FACs_MITO                          NaN                        NaN   \n",
       "RA19_10_17_FACs_MITO                          NaN                        NaN   \n",
       "RA19_10_18_FACs_MITO                          NaN                        NaN   \n",
       "RA19_10_23_FACs_MITO                          NaN                        NaN   \n",
       "RA19_10_14_FACS_citric_MITO                   NaN                        NaN   \n",
       "\n",
       "                            MitoTracing.expectCells MitoTracing.numCores  \\\n",
       "RA19_10_14_FACs_MITO                            NaN                  NaN   \n",
       "RA19_10_17_FACs_MITO                            NaN                  NaN   \n",
       "RA19_10_18_FACs_MITO                            NaN                  NaN   \n",
       "RA19_10_23_FACs_MITO                            NaN                  NaN   \n",
       "RA19_10_14_FACS_citric_MITO                     NaN                  NaN   \n",
       "\n",
       "                            MitoTracing.memory MitoTracing.whitelist  \\\n",
       "RA19_10_14_FACs_MITO                       NaN                   NaN   \n",
       "RA19_10_17_FACs_MITO                       NaN                   NaN   \n",
       "RA19_10_18_FACs_MITO                       NaN                   NaN   \n",
       "RA19_10_23_FACs_MITO                       NaN                   NaN   \n",
       "RA19_10_14_FACS_citric_MITO                NaN                   NaN   \n",
       "\n",
       "                            MitoTracing.dockerRegistry  \n",
       "RA19_10_14_FACs_MITO                               NaN  \n",
       "RA19_10_17_FACs_MITO                               NaN  \n",
       "RA19_10_18_FACs_MITO                               NaN  \n",
       "RA19_10_23_FACs_MITO                               NaN  \n",
       "RA19_10_14_FACS_citric_MITO                        NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from input template\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bd8a0e97-49ad-4295-a4ff-ca83e90732c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            MitoTracing.sampleName MitoTracing.fastqName  \\\n",
       "RA19_10_14_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_17_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_18_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_23_FACs_MITO                           NaN                   NaN   \n",
       "RA19_10_14_FACS_citric_MITO                    NaN                   NaN   \n",
       "\n",
       "                            MitoTracing.fastqR1 MitoTracing.fastqR2  \\\n",
       "RA19_10_14_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_17_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_18_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_23_FACs_MITO                        NaN                 NaN   \n",
       "RA19_10_14_FACS_citric_MITO                 NaN                 NaN   \n",
       "\n",
       "                            MitoTracing.reference  MitoTracing.includeIntrons  \\\n",
       "RA19_10_14_FACs_MITO                          NaN                       False   \n",
       "RA19_10_17_FACs_MITO                          NaN                       False   \n",
       "RA19_10_18_FACs_MITO                          NaN                       False   \n",
       "RA19_10_23_FACs_MITO                          NaN                       False   \n",
       "RA19_10_14_FACS_citric_MITO                   NaN                       False   \n",
       "\n",
       "                             MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "RA19_10_14_FACs_MITO                            5000                    16   \n",
       "RA19_10_17_FACs_MITO                            5000                    16   \n",
       "RA19_10_18_FACs_MITO                            5000                    16   \n",
       "RA19_10_23_FACs_MITO                            5000                    16   \n",
       "RA19_10_14_FACS_citric_MITO                     5000                    16   \n",
       "\n",
       "                             MitoTracing.memory MitoTracing.whitelist  \\\n",
       "RA19_10_14_FACs_MITO                        128                   NaN   \n",
       "RA19_10_17_FACs_MITO                        128                   NaN   \n",
       "RA19_10_18_FACs_MITO                        128                   NaN   \n",
       "RA19_10_23_FACs_MITO                        128                   NaN   \n",
       "RA19_10_14_FACS_citric_MITO                 128                   NaN   \n",
       "\n",
       "                            MitoTracing.dockerRegistry  \n",
       "RA19_10_14_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_17_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_18_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_23_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_14_FACS_citric_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Default inputs\n",
    "inputs[f\"{prefix}.includeIntrons\"] = False\n",
    "inputs[f\"{prefix}.expectCells\"] = 5000\n",
    "inputs[f\"{prefix}.numCores\"] = 16\n",
    "inputs[f\"{prefix}.memory\"] = 128\n",
    "\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "db50eb55-40dc-4840-bc61-63525c2afb2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>RA19_10_14_FACs_MITO</td>\n",
       "      <td>3009_RA19_10_14_FACs_MITO_IGO_12411_5</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>RA19_10_17_FACs_MITO</td>\n",
       "      <td>3010_RA19_10_17_FACs_MITO_IGO_12411_6</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>RA19_10_18_FACs_MITO</td>\n",
       "      <td>3011_RA19_10_18_FACs_MITO_IGO_12411_7</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>RA19_10_23_FACs_MITO</td>\n",
       "      <td>3012_RA19_10_23_FACs_MITO_IGO_12411_8</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>RA19_10_14_FACS_citric_MITO</td>\n",
       "      <td>3013_RA19_10_14_FACS_citric_MITO_IGO...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  MitoTracing.sampleName  \\\n",
       "RA19_10_14_FACs_MITO                RA19_10_14_FACs_MITO   \n",
       "RA19_10_17_FACs_MITO                RA19_10_17_FACs_MITO   \n",
       "RA19_10_18_FACs_MITO                RA19_10_18_FACs_MITO   \n",
       "RA19_10_23_FACs_MITO                RA19_10_23_FACs_MITO   \n",
       "RA19_10_14_FACS_citric_MITO  RA19_10_14_FACS_citric_MITO   \n",
       "\n",
       "                                               MitoTracing.fastqName  \\\n",
       "RA19_10_14_FACs_MITO           3009_RA19_10_14_FACs_MITO_IGO_12411_5   \n",
       "RA19_10_17_FACs_MITO           3010_RA19_10_17_FACs_MITO_IGO_12411_6   \n",
       "RA19_10_18_FACs_MITO           3011_RA19_10_18_FACs_MITO_IGO_12411_7   \n",
       "RA19_10_23_FACs_MITO           3012_RA19_10_23_FACs_MITO_IGO_12411_8   \n",
       "RA19_10_14_FACS_citric_MITO  3013_RA19_10_14_FACS_citric_MITO_IGO...   \n",
       "\n",
       "                                                 MitoTracing.fastqR1  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                                 MitoTracing.fastqR2  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                            MitoTracing.reference  MitoTracing.includeIntrons  \\\n",
       "RA19_10_14_FACs_MITO                          NaN                       False   \n",
       "RA19_10_17_FACs_MITO                          NaN                       False   \n",
       "RA19_10_18_FACs_MITO                          NaN                       False   \n",
       "RA19_10_23_FACs_MITO                          NaN                       False   \n",
       "RA19_10_14_FACS_citric_MITO                   NaN                       False   \n",
       "\n",
       "                             MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "RA19_10_14_FACs_MITO                            5000                    16   \n",
       "RA19_10_17_FACs_MITO                            5000                    16   \n",
       "RA19_10_18_FACs_MITO                            5000                    16   \n",
       "RA19_10_23_FACs_MITO                            5000                    16   \n",
       "RA19_10_14_FACS_citric_MITO                     5000                    16   \n",
       "\n",
       "                             MitoTracing.memory  \\\n",
       "RA19_10_14_FACs_MITO                        128   \n",
       "RA19_10_17_FACs_MITO                        128   \n",
       "RA19_10_18_FACs_MITO                        128   \n",
       "RA19_10_23_FACs_MITO                        128   \n",
       "RA19_10_14_FACS_citric_MITO                 128   \n",
       "\n",
       "                                               MitoTracing.whitelist  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            MitoTracing.dockerRegistry  \n",
       "RA19_10_14_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_17_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_18_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_23_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_14_FACS_citric_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample information\n",
    "inputs[f\"{prefix}.sampleName\"] = samples.index.tolist()\n",
    "\n",
    "inputs[f\"{prefix}.fastqR1\"] = samples['fastq'].str['R1']\n",
    "inputs[f\"{prefix}.fastqR2\"] = samples['fastq'].str['R2']\n",
    "\n",
    "for sample in inputs.index:\n",
    "    fastqR1_name = get_fastqs_name(inputs.loc[sample, 'MitoTracing.fastqR1'])\n",
    "    fastqR2_name = get_fastqs_name(inputs.loc[sample, 'MitoTracing.fastqR2'])\n",
    "    \n",
    "    assert(fastqR1_name == fastqR1_name)\n",
    "    inputs.loc[sample, f\"{prefix}.fastqName\"] = fastqR1_name\n",
    "    \n",
    "    # Check the whitelist to make sure that barcodes are in nucleotide format\n",
    "    inputs.loc[sample, f\"{prefix}.whitelist\"] = get_whitelist(samples.loc[sample, 'S3_path'])\n",
    "\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "101eb0d4-4ca6-4639-bb67-afe47109a251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>RA19_10_14_FACs_MITO</td>\n",
       "      <td>3009_RA19_10_14_FACs_MITO_IGO_12411_5</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>RA19_10_17_FACs_MITO</td>\n",
       "      <td>3010_RA19_10_17_FACs_MITO_IGO_12411_6</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>RA19_10_18_FACs_MITO</td>\n",
       "      <td>3011_RA19_10_18_FACs_MITO_IGO_12411_7</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>RA19_10_23_FACs_MITO</td>\n",
       "      <td>3012_RA19_10_23_FACs_MITO_IGO_12411_8</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>RA19_10_14_FACS_citric_MITO</td>\n",
       "      <td>3013_RA19_10_14_FACS_citric_MITO_IGO...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  MitoTracing.sampleName  \\\n",
       "RA19_10_14_FACs_MITO                RA19_10_14_FACs_MITO   \n",
       "RA19_10_17_FACs_MITO                RA19_10_17_FACs_MITO   \n",
       "RA19_10_18_FACs_MITO                RA19_10_18_FACs_MITO   \n",
       "RA19_10_23_FACs_MITO                RA19_10_23_FACs_MITO   \n",
       "RA19_10_14_FACS_citric_MITO  RA19_10_14_FACS_citric_MITO   \n",
       "\n",
       "                                               MitoTracing.fastqName  \\\n",
       "RA19_10_14_FACs_MITO           3009_RA19_10_14_FACs_MITO_IGO_12411_5   \n",
       "RA19_10_17_FACs_MITO           3010_RA19_10_17_FACs_MITO_IGO_12411_6   \n",
       "RA19_10_18_FACs_MITO           3011_RA19_10_18_FACs_MITO_IGO_12411_7   \n",
       "RA19_10_23_FACs_MITO           3012_RA19_10_23_FACs_MITO_IGO_12411_8   \n",
       "RA19_10_14_FACS_citric_MITO  3013_RA19_10_14_FACS_citric_MITO_IGO...   \n",
       "\n",
       "                                                 MitoTracing.fastqR1  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                                 MitoTracing.fastqR2  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                               MitoTracing.reference  \\\n",
       "RA19_10_14_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_17_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_18_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_23_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_14_FACS_citric_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                             MitoTracing.includeIntrons  \\\n",
       "RA19_10_14_FACs_MITO                              False   \n",
       "RA19_10_17_FACs_MITO                              False   \n",
       "RA19_10_18_FACs_MITO                              False   \n",
       "RA19_10_23_FACs_MITO                              False   \n",
       "RA19_10_14_FACS_citric_MITO                       False   \n",
       "\n",
       "                             MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "RA19_10_14_FACs_MITO                            5000                    16   \n",
       "RA19_10_17_FACs_MITO                            5000                    16   \n",
       "RA19_10_18_FACs_MITO                            5000                    16   \n",
       "RA19_10_23_FACs_MITO                            5000                    16   \n",
       "RA19_10_14_FACS_citric_MITO                     5000                    16   \n",
       "\n",
       "                             MitoTracing.memory  \\\n",
       "RA19_10_14_FACs_MITO                        128   \n",
       "RA19_10_17_FACs_MITO                        128   \n",
       "RA19_10_18_FACs_MITO                        128   \n",
       "RA19_10_23_FACs_MITO                        128   \n",
       "RA19_10_14_FACS_citric_MITO                 128   \n",
       "\n",
       "                                               MitoTracing.whitelist  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            MitoTracing.dockerRegistry  \n",
       "RA19_10_14_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_17_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_18_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_23_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_14_FACS_citric_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reference created by Jaeyoung\n",
    "cellRangerRefPkg=\"s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/reference-packages/refdata-gex-mito-GRCh38-ensemble98.tar.gz\"\n",
    "mitoFastaRefPkg=\"s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/reference-packages/mito-fasta-GRCh38.tar.gz\"\n",
    "\n",
    "reference = [dict(cellRangerRefPkg=cellRangerRefPkg,\n",
    "                  mitoFastaRefPkg=mitoFastaRefPkg)] * len(inputs)\n",
    "inputs[f\"{prefix}.reference\"] = reference\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f12841-72fe-49c6-b91c-2381f7fc62be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0f737133-5115-4ee6-bf93-5443e5c652d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make label file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6ca8c556-8254-4051-93ee-5447d43abfbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            pipelineType project sample owner destination  \\\n",
       "RA19_10_14_FACs_MITO                 NaN     NaN    NaN   NaN         NaN   \n",
       "RA19_10_17_FACs_MITO                 NaN     NaN    NaN   NaN         NaN   \n",
       "RA19_10_18_FACs_MITO                 NaN     NaN    NaN   NaN         NaN   \n",
       "RA19_10_23_FACs_MITO                 NaN     NaN    NaN   NaN         NaN   \n",
       "RA19_10_14_FACS_citric_MITO          NaN     NaN    NaN   NaN         NaN   \n",
       "\n",
       "                            transfer comment  \n",
       "RA19_10_14_FACs_MITO             NaN     NaN  \n",
       "RA19_10_17_FACs_MITO             NaN     NaN  \n",
       "RA19_10_18_FACs_MITO             NaN     NaN  \n",
       "RA19_10_23_FACs_MITO             NaN     NaN  \n",
       "RA19_10_14_FACS_citric_MITO      NaN     NaN  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from labels template\n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow labels\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0731efde-30bb-419b-b5a4-b8420911f309",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_14_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_17_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_18_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_23_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_14_FACS_citric_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            pipelineType       project  \\\n",
       "RA19_10_14_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_17_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_18_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_23_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_14_FACS_citric_MITO  MitoTracing  HTAN_CITEseq   \n",
       "\n",
       "                                                  sample    owner  \\\n",
       "RA19_10_14_FACs_MITO                RA19_10_14_FACs_MITO  sohailn   \n",
       "RA19_10_17_FACs_MITO                RA19_10_17_FACs_MITO  sohailn   \n",
       "RA19_10_18_FACs_MITO                RA19_10_18_FACs_MITO  sohailn   \n",
       "RA19_10_23_FACs_MITO                RA19_10_23_FACs_MITO  sohailn   \n",
       "RA19_10_14_FACS_citric_MITO  RA19_10_14_FACS_citric_MITO  sohailn   \n",
       "\n",
       "                                                         destination transfer  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "\n",
       "                             comment  \n",
       "RA19_10_14_FACs_MITO         sohailn  \n",
       "RA19_10_17_FACs_MITO         sohailn  \n",
       "RA19_10_18_FACs_MITO         sohailn  \n",
       "RA19_10_23_FACs_MITO         sohailn  \n",
       "RA19_10_14_FACS_citric_MITO  sohailn  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5973e09-e9bf-4415-acee-9e1809afdf5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "84007e1e-ae11-4665-ba47-aab5b803af27",
   "metadata": {},
   "source": [
    "# Submit job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "68d260a1-cca0-41d6-9c84-d5f1aef94128",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>RA19_10_14_FACs_MITO</td>\n",
       "      <td>3009_RA19_10_14_FACs_MITO_IGO_12411_5</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>RA19_10_17_FACs_MITO</td>\n",
       "      <td>3010_RA19_10_17_FACs_MITO_IGO_12411_6</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>RA19_10_18_FACs_MITO</td>\n",
       "      <td>3011_RA19_10_18_FACs_MITO_IGO_12411_7</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>RA19_10_23_FACs_MITO</td>\n",
       "      <td>3012_RA19_10_23_FACs_MITO_IGO_12411_8</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>RA19_10_14_FACS_citric_MITO</td>\n",
       "      <td>3013_RA19_10_14_FACS_citric_MITO_IGO...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  MitoTracing.sampleName  \\\n",
       "RA19_10_14_FACs_MITO                RA19_10_14_FACs_MITO   \n",
       "RA19_10_17_FACs_MITO                RA19_10_17_FACs_MITO   \n",
       "RA19_10_18_FACs_MITO                RA19_10_18_FACs_MITO   \n",
       "RA19_10_23_FACs_MITO                RA19_10_23_FACs_MITO   \n",
       "RA19_10_14_FACS_citric_MITO  RA19_10_14_FACS_citric_MITO   \n",
       "\n",
       "                                               MitoTracing.fastqName  \\\n",
       "RA19_10_14_FACs_MITO           3009_RA19_10_14_FACs_MITO_IGO_12411_5   \n",
       "RA19_10_17_FACs_MITO           3010_RA19_10_17_FACs_MITO_IGO_12411_6   \n",
       "RA19_10_18_FACs_MITO           3011_RA19_10_18_FACs_MITO_IGO_12411_7   \n",
       "RA19_10_23_FACs_MITO           3012_RA19_10_23_FACs_MITO_IGO_12411_8   \n",
       "RA19_10_14_FACS_citric_MITO  3013_RA19_10_14_FACS_citric_MITO_IGO...   \n",
       "\n",
       "                                                 MitoTracing.fastqR1  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                                 MitoTracing.fastqR2  \\\n",
       "RA19_10_14_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_17_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_18_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_23_FACs_MITO         [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "RA19_10_14_FACS_citric_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                               MitoTracing.reference  \\\n",
       "RA19_10_14_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_17_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_18_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_23_FACs_MITO         {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "RA19_10_14_FACS_citric_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                             MitoTracing.includeIntrons  \\\n",
       "RA19_10_14_FACs_MITO                              False   \n",
       "RA19_10_17_FACs_MITO                              False   \n",
       "RA19_10_18_FACs_MITO                              False   \n",
       "RA19_10_23_FACs_MITO                              False   \n",
       "RA19_10_14_FACS_citric_MITO                       False   \n",
       "\n",
       "                             MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "RA19_10_14_FACs_MITO                            5000                    16   \n",
       "RA19_10_17_FACs_MITO                            5000                    16   \n",
       "RA19_10_18_FACs_MITO                            5000                    16   \n",
       "RA19_10_23_FACs_MITO                            5000                    16   \n",
       "RA19_10_14_FACS_citric_MITO                     5000                    16   \n",
       "\n",
       "                             MitoTracing.memory  \\\n",
       "RA19_10_14_FACs_MITO                        128   \n",
       "RA19_10_17_FACs_MITO                        128   \n",
       "RA19_10_18_FACs_MITO                        128   \n",
       "RA19_10_23_FACs_MITO                        128   \n",
       "RA19_10_14_FACS_citric_MITO                 128   \n",
       "\n",
       "                                               MitoTracing.whitelist  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                            MitoTracing.dockerRegistry  \n",
       "RA19_10_14_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_17_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_18_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_23_FACs_MITO                   quay.io/hisplan  \n",
       "RA19_10_14_FACS_citric_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5713d9e-a923-446e-9df8-fb9619590923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_14_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_17_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_17_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_18_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_18_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_23_FACs_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_23_FACs_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA19_10_14_FACS_citric_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>HTAN_CITEseq</td>\n",
       "      <td>RA19_10_14_FACS_citric_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            pipelineType       project  \\\n",
       "RA19_10_14_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_17_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_18_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_23_FACs_MITO         MitoTracing  HTAN_CITEseq   \n",
       "RA19_10_14_FACS_citric_MITO  MitoTracing  HTAN_CITEseq   \n",
       "\n",
       "                                                  sample    owner  \\\n",
       "RA19_10_14_FACs_MITO                RA19_10_14_FACs_MITO  sohailn   \n",
       "RA19_10_17_FACs_MITO                RA19_10_17_FACs_MITO  sohailn   \n",
       "RA19_10_18_FACs_MITO                RA19_10_18_FACs_MITO  sohailn   \n",
       "RA19_10_23_FACs_MITO                RA19_10_23_FACs_MITO  sohailn   \n",
       "RA19_10_14_FACS_citric_MITO  RA19_10_14_FACS_citric_MITO  sohailn   \n",
       "\n",
       "                                                         destination transfer  \\\n",
       "RA19_10_14_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_17_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_18_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_23_FACs_MITO         s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "RA19_10_14_FACS_citric_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...        -   \n",
       "\n",
       "                             comment  \n",
       "RA19_10_14_FACs_MITO         sohailn  \n",
       "RA19_10_17_FACs_MITO         sohailn  \n",
       "RA19_10_18_FACs_MITO         sohailn  \n",
       "RA19_10_23_FACs_MITO         sohailn  \n",
       "RA19_10_14_FACS_citric_MITO  sohailn  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3bd30985-9d8f-4a6a-afb4-0f93357ad25b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f68dbeab6b234b3c9f22a064e7cfb506",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())\n",
    "\n",
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(inputs.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))\n",
    "            \n",
    "        # Sometimes causes problems if too many samples are run at once\n",
    "        time.sleep(20) # Delay for 20 seconds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "25c94191-2218-4edc-b3cf-d705e0f36721",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACS_citric_MITO_MitoTracing.inputs.json\n",
      "/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACS_citric_MITO_MitoTracing.labels.json\n"
     ]
    }
   ],
   "source": [
    "print(path_to_inputs)\n",
    "print(path_to_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0a34cef8-6aef-46ea-a3ec-b5fe3e674213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACs_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACs_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"bea801a8-dd25-489e-87c6-bb07e87d3548\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_17_FACs_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_17_FACs_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"174cec20-8e12-4ea3-9d49-eb022bf33fd6\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_18_FACs_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_18_FACs_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"910de209-1ff7-4b7e-822b-6233cb252fbd\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_23_FACs_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_23_FACs_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"9e8049a7-344d-44e4-a0ba-7d5e3795dcc0\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACS_citric_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/RA19_10_14_FACS_citric_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"c6eaa181-4425-44b3-94cd-89e34047d464\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6c483fed-f39b-467b-a8e2-43a1b2fa5b9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/RA19_10_14_FACs_MITO/mito-tracing-outs'\n",
      " 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/RA19_10_17_FACs_MITO/mito-tracing-outs'\n",
      " 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/RA19_10_18_FACs_MITO/mito-tracing-outs'\n",
      " 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/RA19_10_23_FACs_MITO/mito-tracing-outs'\n",
      " 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/RA19_10_14_FACS_citric_MITO/mito-tracing-outs']\n"
     ]
    }
   ],
   "source": [
    "print(labels['destination'].values)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
