{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## List FASTQs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'genomic': ['R2'],\n",
    "    'barcode': ['R1']\n",
    "}\n",
    "\n",
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_barcode_genomic_fastqs(samples):\n",
    "    \n",
    "    samples['fastqBarcode'] = np.empty((len(samples), 0)).tolist()\n",
    "    samples['fastqGenomic'] = np.empty((len(samples), 0)).tolist()\n",
    "    \n",
    "    for sample, row in samples.iterrows():\n",
    "        S3_path = row['S3_path']\n",
    "        \n",
    "        # Barcode FASTQ files\n",
    "        fastq_file_ids = fastq_map['barcode']        \n",
    "        barcode_path = f\"{S3_path}/barcode/\"\n",
    "        fastqBarcode = get_fastqs(barcode_path, fastq_file_ids)['R1']\n",
    "        \n",
    "        # Genomic FASTQ files\n",
    "        fastq_file_ids = fastq_map['genomic']        \n",
    "        genomic_path = f\"{S3_path}/genomic/\"\n",
    "        fastqGenomic = get_fastqs(genomic_path, fastq_file_ids)['R2']\n",
    "        \n",
    "        samples.loc[sample, 'fastqBarcode'] += fastqBarcode\n",
    "        samples.loc[sample, 'fastqGenomic'] += fastqGenomic\n",
    "    \n",
    "    return samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## DB queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "def get_sample_name(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Sample\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_names = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_names.append(result[0])\n",
    "        return sample_names\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "    \n",
    "    \n",
    "def get_aws_path(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_paths = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_paths.append(result[0])\n",
    "        return sample_paths\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "            \n",
    "def get_sample_id(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_ids = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_ids.append(result[0])\n",
    "        return sample_ids\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def format_sample_aws(querys, query_col, creds):\n",
    "    sample_names = []\n",
    "    sample_paths = []\n",
    "    sample_ids = []\n",
    "    \n",
    "    for query in querys:\n",
    "        sample_names += get_sample_name(query, query_col, creds)\n",
    "        sample_paths += get_aws_path(query, query_col, creds)\n",
    "        sample_ids += get_sample_id(query, query_col, creds)\n",
    "        \n",
    "    sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "    \n",
    "    samples = pd.DataFrame(\n",
    "        dict(S3_path=sample_paths, Sample_ID=sample_ids),\n",
    "        index=sample_names,\n",
    "        dtype=str,\n",
    "    )\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomeIndex_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.genomeIndex_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            genomeIndex_id = result[0]\n",
    "            \n",
    "        return genomeIndex_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_scTech_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.scTech_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            scTech_id = result[0]\n",
    "            \n",
    "        return scTech_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_index(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.gIndex\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            index = result[0]\n",
    "            \n",
    "        return index\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_assay(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Run_name\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            assay = result[0]\n",
    "            \n",
    "        return assay\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_barcode(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.barcodes\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            barcode = result[0]\n",
    "            \n",
    "        return barcode\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "        \n",
    "def format_assay_barcode(samples, creds):\n",
    "    \n",
    "    # samples['genomeIndex_id'] = np.nan\n",
    "    samples['index'] = np.nan\n",
    "    # samples['scTech_id'] = np.nan\n",
    "    samples['assay'] = np.nan\n",
    "    samples['barcode'] = np.nan\n",
    "    \n",
    "    for sample, row in samples.iterrows():\n",
    "        sample_id = row['Sample_ID']\n",
    "        \n",
    "        genomeIndex_id  = get_genomeIndex_id(sample_id, 'id', creds)\n",
    "        index = get_index(genomeIndex_id, 'id', creds)\n",
    "        scTech_id = get_scTech_id(genomeIndex_id, 'id', creds)\n",
    "        assay = get_assay(scTech_id, 'id', creds)\n",
    "        barcode = get_barcode(scTech_id, 'id', creds)\n",
    "        \n",
    "        # samples.loc[sample, 'genomeIndex_id'] = genomeIndex_id\n",
    "        samples.loc[sample, 'index'] = index\n",
    "        # samples.loc[sample, 'scTech_id'] = scTech_id\n",
    "        samples.loc[sample, 'assay'] = assay\n",
    "        samples.loc[sample, 'barcode'] = barcode\n",
    "\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_project_id(sample_id, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"SeqcAda\": \"seqc-ada-results\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"SeqcAda\": \"submit.sh\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"SeqcAda\" # Workflow to run; also .wdl filename prefix\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "output_dirname = results_dirs[prefix]\n",
    "\n",
    "# If need to add comment, put here\n",
    "email = '20noor.sohail@gmail.com'\n",
    "comment = \"sohailn\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = glob.glob(f\"{Path.home()}/scing/bin/*ada*\")[0]\n",
    "path_to_exec = sh_files[prefix]\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = glob.glob(f\"{workflow_dir}/*.options.aws.json\")[0]\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Excution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3251</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_path Sample_ID\n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...      3249\n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...      3250\n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...      3251"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "request_ids = ['ARN-1167']\n",
    "\n",
    "samples = format_sample_aws(request_ids, 'request_id', creds)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3249</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3250</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3251</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_path Sample_ID  \\\n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...      3249   \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...      3250   \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...      3251   \n",
       "\n",
       "                                                   index     assay  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "ARN-1167_Normal  s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "\n",
       "                                                 barcode  \n",
       "ARN-1167_M-4T1   s3://seqc-public/barcodes/ten_x_v3/f...  \n",
       "ARN-1167_Normal  s3://seqc-public/barcodes/ten_x_v3/f...  \n",
       "ARN-1167_PM-4T1  s3://seqc-public/barcodes/ten_x_v3/f...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = format_assay_barcode(samples, creds)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fastqBarcode</th>\n",
       "      <th>fastqGenomic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3249</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3250</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3251</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_path Sample_ID  \\\n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...      3249   \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...      3250   \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...      3251   \n",
       "\n",
       "                                                   index     assay  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "ARN-1167_Normal  s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/genomes/mm38_long_p...  ten_x_v3   \n",
       "\n",
       "                                                 barcode  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_Normal  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "\n",
       "                                            fastqBarcode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...   \n",
       "\n",
       "                                            fastqGenomic  \n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...  \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...  \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = get_barcode_genomic_fastqs(samples)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqcAda.version</th>\n",
       "      <th>SeqcAda.assay</th>\n",
       "      <th>SeqcAda.index</th>\n",
       "      <th>SeqcAda.barcodeFiles</th>\n",
       "      <th>SeqcAda.fastqBarcode</th>\n",
       "      <th>SeqcAda.fastqGenomic</th>\n",
       "      <th>SeqcAda.filterMode</th>\n",
       "      <th>SeqcAda.outputPrefix</th>\n",
       "      <th>SeqcAda.starArguments</th>\n",
       "      <th>SeqcAda.email</th>\n",
       "      <th>SeqcAda.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SeqcAda.version SeqcAda.assay SeqcAda.index  \\\n",
       "ARN-1167_M-4T1              NaN           NaN           NaN   \n",
       "ARN-1167_Normal             NaN           NaN           NaN   \n",
       "ARN-1167_PM-4T1             NaN           NaN           NaN   \n",
       "\n",
       "                SeqcAda.barcodeFiles SeqcAda.fastqBarcode  \\\n",
       "ARN-1167_M-4T1                   NaN                  NaN   \n",
       "ARN-1167_Normal                  NaN                  NaN   \n",
       "ARN-1167_PM-4T1                  NaN                  NaN   \n",
       "\n",
       "                SeqcAda.fastqGenomic SeqcAda.filterMode SeqcAda.outputPrefix  \\\n",
       "ARN-1167_M-4T1                   NaN                NaN                  NaN   \n",
       "ARN-1167_Normal                  NaN                NaN                  NaN   \n",
       "ARN-1167_PM-4T1                  NaN                NaN                  NaN   \n",
       "\n",
       "                SeqcAda.starArguments SeqcAda.email SeqcAda.dockerRegistry  \n",
       "ARN-1167_M-4T1                    NaN           NaN                    NaN  \n",
       "ARN-1167_Normal                   NaN           NaN                    NaN  \n",
       "ARN-1167_PM-4T1                   NaN           NaN                    NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_prefix = 'template'\n",
    "# Load minimum inputs and labels fields from input template\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqcAda.version</th>\n",
       "      <th>SeqcAda.assay</th>\n",
       "      <th>SeqcAda.index</th>\n",
       "      <th>SeqcAda.barcodeFiles</th>\n",
       "      <th>SeqcAda.fastqBarcode</th>\n",
       "      <th>SeqcAda.fastqGenomic</th>\n",
       "      <th>SeqcAda.filterMode</th>\n",
       "      <th>SeqcAda.outputPrefix</th>\n",
       "      <th>SeqcAda.starArguments</th>\n",
       "      <th>SeqcAda.email</th>\n",
       "      <th>SeqcAda.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SeqcAda.version SeqcAda.assay SeqcAda.index  \\\n",
       "ARN-1167_M-4T1           0.2.11           NaN           NaN   \n",
       "ARN-1167_Normal          0.2.11           NaN           NaN   \n",
       "ARN-1167_PM-4T1          0.2.11           NaN           NaN   \n",
       "\n",
       "                SeqcAda.barcodeFiles SeqcAda.fastqBarcode  \\\n",
       "ARN-1167_M-4T1                   NaN                  NaN   \n",
       "ARN-1167_Normal                  NaN                  NaN   \n",
       "ARN-1167_PM-4T1                  NaN                  NaN   \n",
       "\n",
       "                SeqcAda.fastqGenomic SeqcAda.filterMode SeqcAda.outputPrefix  \\\n",
       "ARN-1167_M-4T1                   NaN          scRNA-seq                  NaN   \n",
       "ARN-1167_Normal                  NaN          scRNA-seq                  NaN   \n",
       "ARN-1167_PM-4T1                  NaN          scRNA-seq                  NaN   \n",
       "\n",
       "                SeqcAda.starArguments            SeqcAda.email  \\\n",
       "ARN-1167_M-4T1           runRNGseed=0  20noor.sohail@gmail.com   \n",
       "ARN-1167_Normal          runRNGseed=0  20noor.sohail@gmail.com   \n",
       "ARN-1167_PM-4T1          runRNGseed=0  20noor.sohail@gmail.com   \n",
       "\n",
       "                SeqcAda.dockerRegistry  \n",
       "ARN-1167_M-4T1         quay.io/hisplan  \n",
       "ARN-1167_Normal        quay.io/hisplan  \n",
       "ARN-1167_PM-4T1        quay.io/hisplan  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Default inputs\n",
    "inputs[f\"{prefix}.version\"] = \"0.2.11\"\n",
    "inputs[f\"{prefix}.filterMode\"] = \"scRNA-seq\"\n",
    "inputs[f\"{prefix}.starArguments\"] = \"runRNGseed=0\"\n",
    "inputs[f\"{prefix}.email\"] = email\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqcAda.version</th>\n",
       "      <th>SeqcAda.assay</th>\n",
       "      <th>SeqcAda.index</th>\n",
       "      <th>SeqcAda.barcodeFiles</th>\n",
       "      <th>SeqcAda.fastqBarcode</th>\n",
       "      <th>SeqcAda.fastqGenomic</th>\n",
       "      <th>SeqcAda.filterMode</th>\n",
       "      <th>SeqcAda.outputPrefix</th>\n",
       "      <th>SeqcAda.starArguments</th>\n",
       "      <th>SeqcAda.email</th>\n",
       "      <th>SeqcAda.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_M-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_Normal</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_PM-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SeqcAda.version SeqcAda.assay  \\\n",
       "ARN-1167_M-4T1           0.2.11      ten_x_v3   \n",
       "ARN-1167_Normal          0.2.11      ten_x_v3   \n",
       "ARN-1167_PM-4T1          0.2.11      ten_x_v3   \n",
       "\n",
       "                                           SeqcAda.index  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/genomes/mm38_long_p...   \n",
       "ARN-1167_Normal  s3://seqc-public/genomes/mm38_long_p...   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/genomes/mm38_long_p...   \n",
       "\n",
       "                                    SeqcAda.barcodeFiles  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_Normal  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "\n",
       "                                    SeqcAda.fastqBarcode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...   \n",
       "\n",
       "                                    SeqcAda.fastqGenomic SeqcAda.filterMode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "\n",
       "                SeqcAda.outputPrefix SeqcAda.starArguments  \\\n",
       "ARN-1167_M-4T1        ARN-1167_M-4T1          runRNGseed=0   \n",
       "ARN-1167_Normal      ARN-1167_Normal          runRNGseed=0   \n",
       "ARN-1167_PM-4T1      ARN-1167_PM-4T1          runRNGseed=0   \n",
       "\n",
       "                           SeqcAda.email SeqcAda.dockerRegistry  \n",
       "ARN-1167_M-4T1   20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_Normal  20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_PM-4T1  20noor.sohail@gmail.com        quay.io/hisplan  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample inputs\n",
    "inputs[f\"{prefix}.assay\"] = samples[\"assay\"]\n",
    "inputs[f\"{prefix}.index\"] = samples[\"index\"]\n",
    "inputs[f\"{prefix}.barcodeFiles\"] = samples[\"barcode\"]\n",
    "inputs[f\"{prefix}.fastqBarcode\"] = samples[\"fastqBarcode\"]\n",
    "inputs[f\"{prefix}.fastqGenomic\"] = samples[\"fastqGenomic\"]\n",
    "\n",
    "inputs[f\"{prefix}.outputPrefix\"] = samples.index.tolist()\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqcAda.version</th>\n",
       "      <th>SeqcAda.assay</th>\n",
       "      <th>SeqcAda.index</th>\n",
       "      <th>SeqcAda.barcodeFiles</th>\n",
       "      <th>SeqcAda.fastqBarcode</th>\n",
       "      <th>SeqcAda.fastqGenomic</th>\n",
       "      <th>SeqcAda.filterMode</th>\n",
       "      <th>SeqcAda.outputPrefix</th>\n",
       "      <th>SeqcAda.starArguments</th>\n",
       "      <th>SeqcAda.email</th>\n",
       "      <th>SeqcAda.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_M-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_Normal</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_PM-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SeqcAda.version SeqcAda.assay  \\\n",
       "ARN-1167_M-4T1           0.2.11      ten_x_v3   \n",
       "ARN-1167_Normal          0.2.11      ten_x_v3   \n",
       "ARN-1167_PM-4T1          0.2.11      ten_x_v3   \n",
       "\n",
       "                                           SeqcAda.index  \\\n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...   \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...   \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...   \n",
       "\n",
       "                                    SeqcAda.barcodeFiles  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_Normal  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "\n",
       "                                    SeqcAda.fastqBarcode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...   \n",
       "\n",
       "                                    SeqcAda.fastqGenomic SeqcAda.filterMode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "\n",
       "                SeqcAda.outputPrefix SeqcAda.starArguments  \\\n",
       "ARN-1167_M-4T1        ARN-1167_M-4T1          runRNGseed=0   \n",
       "ARN-1167_Normal      ARN-1167_Normal          runRNGseed=0   \n",
       "ARN-1167_PM-4T1      ARN-1167_PM-4T1          runRNGseed=0   \n",
       "\n",
       "                           SeqcAda.email SeqcAda.dockerRegistry  \n",
       "ARN-1167_M-4T1   20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_Normal  20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_PM-4T1  20noor.sohail@gmail.com        quay.io/hisplan  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Additional changes\n",
    "inputs[f\"{prefix}.index\"] = \"s3://dp-lab-data/collaborators/aboire/LeptomeningealMetHeterogeneity/transgene_reference/refdata-seqc/STAR-index/\"\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make label file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pipelineType project sample owner destination transfer comment\n",
       "ARN-1167_M-4T1           NaN     NaN    NaN   NaN         NaN      NaN     NaN\n",
       "ARN-1167_Normal          NaN     NaN    NaN   NaN         NaN      NaN     NaN\n",
       "ARN-1167_PM-4T1          NaN     NaN    NaN   NaN         NaN      NaN     NaN"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from labels template\n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow labels\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_M-4T1</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_Normal</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_PM-4T1</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pipelineType                                  project  \\\n",
       "ARN-1167_M-4T1       SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_Normal      SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_PM-4T1      SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "\n",
       "                          sample    owner  \\\n",
       "ARN-1167_M-4T1    ARN-1167_M-4T1  sohailn   \n",
       "ARN-1167_Normal  ARN-1167_Normal  sohailn   \n",
       "ARN-1167_PM-4T1  ARN-1167_PM-4T1  sohailn   \n",
       "\n",
       "                                             destination transfer  comment  \n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...        -  sohailn  \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...        -  sohailn  \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...        -  sohailn  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Submit job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqcAda.version</th>\n",
       "      <th>SeqcAda.assay</th>\n",
       "      <th>SeqcAda.index</th>\n",
       "      <th>SeqcAda.barcodeFiles</th>\n",
       "      <th>SeqcAda.fastqBarcode</th>\n",
       "      <th>SeqcAda.fastqGenomic</th>\n",
       "      <th>SeqcAda.filterMode</th>\n",
       "      <th>SeqcAda.outputPrefix</th>\n",
       "      <th>SeqcAda.starArguments</th>\n",
       "      <th>SeqcAda.email</th>\n",
       "      <th>SeqcAda.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_M-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_Normal</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>0.2.11</td>\n",
       "      <td>ten_x_v3</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>s3://seqc-public/barcodes/ten_x_v3/f...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/aboi...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "      <td>ARN-1167_PM-4T1</td>\n",
       "      <td>runRNGseed=0</td>\n",
       "      <td>20noor.sohail@gmail.com</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SeqcAda.version SeqcAda.assay  \\\n",
       "ARN-1167_M-4T1           0.2.11      ten_x_v3   \n",
       "ARN-1167_Normal          0.2.11      ten_x_v3   \n",
       "ARN-1167_PM-4T1          0.2.11      ten_x_v3   \n",
       "\n",
       "                                           SeqcAda.index  \\\n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...   \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...   \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...   \n",
       "\n",
       "                                    SeqcAda.barcodeFiles  \\\n",
       "ARN-1167_M-4T1   s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_Normal  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "ARN-1167_PM-4T1  s3://seqc-public/barcodes/ten_x_v3/f...   \n",
       "\n",
       "                                    SeqcAda.fastqBarcode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...   \n",
       "\n",
       "                                    SeqcAda.fastqGenomic SeqcAda.filterMode  \\\n",
       "ARN-1167_M-4T1   [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_Normal  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "ARN-1167_PM-4T1  [s3://dp-lab-data/collaborators/aboi...          scRNA-seq   \n",
       "\n",
       "                SeqcAda.outputPrefix SeqcAda.starArguments  \\\n",
       "ARN-1167_M-4T1        ARN-1167_M-4T1          runRNGseed=0   \n",
       "ARN-1167_Normal      ARN-1167_Normal          runRNGseed=0   \n",
       "ARN-1167_PM-4T1      ARN-1167_PM-4T1          runRNGseed=0   \n",
       "\n",
       "                           SeqcAda.email SeqcAda.dockerRegistry  \n",
       "ARN-1167_M-4T1   20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_Normal  20noor.sohail@gmail.com        quay.io/hisplan  \n",
       "ARN-1167_PM-4T1  20noor.sohail@gmail.com        quay.io/hisplan  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_M-4T1</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_Normal</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>SeqcAda</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>ARN-1167_PM-4T1</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pipelineType                                  project  \\\n",
       "ARN-1167_M-4T1       SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_Normal      SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_PM-4T1      SeqcAda  Leptomeningeal metastasis heterogeneity   \n",
       "\n",
       "                          sample    owner  \\\n",
       "ARN-1167_M-4T1    ARN-1167_M-4T1  sohailn   \n",
       "ARN-1167_Normal  ARN-1167_Normal  sohailn   \n",
       "ARN-1167_PM-4T1  ARN-1167_PM-4T1  sohailn   \n",
       "\n",
       "                                             destination transfer  comment  \n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...        -  sohailn  \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...        -  sohailn  \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...        -  sohailn  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6bb353be8e5b4ac58269f7bc81977ab9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(samples.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_PM-4T1_SeqcAda.inputs.json\n",
      "/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_PM-4T1_SeqcAda.labels.json\n"
     ]
    }
   ],
   "source": [
    "print(path_to_inputs)\n",
    "print(path_to_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/seqc-ada-0.0.4/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_M-4T1_SeqcAda.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_M-4T1_SeqcAda.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/SeqcAda.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"b29e38bf-8da9-4dd2-b1d6-68421dd00357\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/seqc-ada-0.0.4/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_Normal_SeqcAda.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_Normal_SeqcAda.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/SeqcAda.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"870495d0-1c22-44bb-ab21-51ad8e4655ad\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/seqc-ada-0.0.4/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_PM-4T1_SeqcAda.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/configs/ARN-1167_PM-4T1_SeqcAda.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/seqc-ada-0.0.4/SeqcAda.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"bbd1d14b-f1da-4317-a82a-c1129dbdbdac\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
