{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a480ca5b-5dad-4673-9cd4-b9fb5137f40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "\n",
    "import glob\n",
    "import os\n",
    "\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b1b09b4-7cde-45a8-8ae4-a401972d0acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb85f7c-d187-4a57-b795-e07c002eddc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "10b1a35a-406c-4f66-882d-10eeaabce91b",
   "metadata": {},
   "source": [
    "# AWS setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4370e93d-2565-48cf-92e1-3bd6eac525d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: AWS_ACCESS_KEY_ID=ASIAYPY66CWMGNQHAG7V\n",
      "env: AWS_SECRET_ACCESS_KEY=Zt6lmgFmhx1Qn1/mTajuJHxORPKjMd7zHo9yxCv3\n",
      "env: AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEL///////////wEaCXVzLWVhc3QtMSJHMEUCIBOn2n7FYhkwDrJArswLQBQzsB8bEcfdUZp6ZJWuQkNlAiEA/ZQVvnqwuRjHE0/2cTzrTEwF8bS0pfOp5sd5XU9QANMq+AEIqP//////////ARADGgw1ODM2NDM1Njc1MTIiDMFsXq9WBJoyLEdVQyrMAdIJg0FT0qo+cpTVLPKnFLvyQCjzH/7ZlWLd5w1NkTkMDIR4W/d+02fgfncbF3cGwglqlO18saOqvMTOjEUvvvhihRLRzuSajiF3sqG6wsdwu6IgCyk6xdYap84Wt0uo5LshyJAma2fK31yVPTip6n6nxoCcTmsJjhvFtMRGlgiC4bIAeW2lC8lqGHPmyz//tutqj9fZeUR0qmGtr1ium7Gg39WX5IzOgHAWTbb/HSspeP0+xDVfdzOHZyA47RVNc2EIwIXoHCntzPmUEjCNjYqbBjqYAcTJn0OZLrzdFL5RsIVXPaoQMedyiuxr27Bmz4QZoB32whcGu/0osFhwL4la0v+BUXuz2zGPB+byyrGkMB249QQitouR1DIZ/mCUnEz6HvCEQwdH/yIFkeNcY6p+rV/1sOa7op3p43reWd7s9ui3mlQ4QDi63ZJG/aBmGMwmPrCMQVgW9vJR5vr+LjV4YWxqVJ1PqOiZnGf+\n",
      "2021-10-07 15:31:32 agc-583643567512-us-east-1\n",
      "2021-10-07 15:28:07 cdktoolkit-stagingbucket-d49u1xfb0sc4\n",
      "2019-10-09 13:04:06 cf-templates-umiwbnq3566w-us-east-1\n",
      "2019-10-10 12:46:54 dp-daily-reports\n",
      "2017-09-26 10:15:15 dp-lab-data\n",
      "2019-05-23 12:34:58 dp-lab-data-public\n",
      "2018-03-01 13:29:07 dp-lab-glacier\n",
      "2020-11-06 23:02:27 dp-lab-gwf-core\n",
      "2021-07-15 18:14:01 dp-lab-gwf-core2\n",
      "2017-09-26 10:17:18 dp-lab-home\n",
      "2019-08-14 16:42:43 dp-lab-test\n",
      "2019-04-25 12:35:35 elasticbeanstalk-us-east-1-583643567512\n",
      "2019-12-13 15:53:57 scri-computational\n",
      "2017-09-26 10:23:50 seqc-public\n"
     ]
    }
   ],
   "source": [
    "%env AWS_ACCESS_KEY_ID=ASIAYPY66CWMGNQHAG7V\n",
    "%env AWS_SECRET_ACCESS_KEY=Zt6lmgFmhx1Qn1/mTajuJHxORPKjMd7zHo9yxCv3\n",
    "%env AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEL///////////wEaCXVzLWVhc3QtMSJHMEUCIBOn2n7FYhkwDrJArswLQBQzsB8bEcfdUZp6ZJWuQkNlAiEA/ZQVvnqwuRjHE0/2cTzrTEwF8bS0pfOp5sd5XU9QANMq+AEIqP//////////ARADGgw1ODM2NDM1Njc1MTIiDMFsXq9WBJoyLEdVQyrMAdIJg0FT0qo+cpTVLPKnFLvyQCjzH/7ZlWLd5w1NkTkMDIR4W/d+02fgfncbF3cGwglqlO18saOqvMTOjEUvvvhihRLRzuSajiF3sqG6wsdwu6IgCyk6xdYap84Wt0uo5LshyJAma2fK31yVPTip6n6nxoCcTmsJjhvFtMRGlgiC4bIAeW2lC8lqGHPmyz//tutqj9fZeUR0qmGtr1ium7Gg39WX5IzOgHAWTbb/HSspeP0+xDVfdzOHZyA47RVNc2EIwIXoHCntzPmUEjCNjYqbBjqYAcTJn0OZLrzdFL5RsIVXPaoQMedyiuxr27Bmz4QZoB32whcGu/0osFhwL4la0v+BUXuz2zGPB+byyrGkMB249QQitouR1DIZ/mCUnEz6HvCEQwdH/yIFkeNcY6p+rV/1sOa7op3p43reWd7s9ui3mlQ4QDi63ZJG/aBmGMwmPrCMQVgW9vJR5vr+LjV4YWxqVJ1PqOiZnGf+\n",
    "\n",
    "!aws s3 ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1be9e8c4-2c18-4cb5-85cc-ee33b510f23c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7091dcda-0a18-43a2-9c3f-077662f3bbc3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "66d94c6b-ca49-4e15-be39-13ee21ec94e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"TransgenesSeqc\": \"refdata-seqc\",\n",
    "    \"TransgenesCellRanger\": \"refdata-cellranger\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"TransgenesSeqc\": \"submit-seqc.sh\",\n",
    "    \"TransgenesCellRanger\": \"submit-cellranger.sh\",\n",
    "}\n",
    "\n",
    "# Version (CellRanger or Star) to use; Should be in database\n",
    "latest_STAR = \"2.5.3a\"\n",
    "latest_CellRanger = \"6.1.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f3d482b7-cc91-44f8-9b80-93847138c22c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which pipeline are you running\n",
    "prefix = \"TransgenesCellRanger\" # Workflow to run; also .wdl filename prefix\n",
    "output_dirname = \"refdata-cellranger\"\n",
    "\n",
    "workflow_dir = glob.glob(f\"{Path.home()}/scing/bin/transgenes*\")[0]\n",
    "path_to_exec = f\"{workflow_dir}/submit.sh\" # CHANGE THIS FOR SHARP\n",
    "\n",
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\"\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cca545a8-ccc3-4e15-8813-ae10e2d07ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bdb9c13b-3129-4069-81a6-19919996f094",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Workflow file paths\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = f\"{workflow_dir}/Sharp.options.aws.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "48f3391c-cd0e-48da-b05c-4adcc4e44f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6670e0cc-8731-41cc-8d31-761deb3f3a1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e88c91c2-6044-463e-9bd9-14e1631963b4",
   "metadata": {},
   "source": [
    "# Sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "328b93a1-447a-4f30-9d5c-8cf82497fa43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AWS_storage</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>sc_tech</th>\n",
       "      <th>project_id</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARN-1167_M-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3249</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10X_V3.1</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_Normal</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3250</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10X_V3.1</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARN-1167_PM-4T1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/aboir...</td>\n",
       "      <td>3251</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10X_V3.1</td>\n",
       "      <td>Leptomeningeal metastasis heterogeneity</td>\n",
       "      <td>s3://seqc-public/genomes/mm38_long_p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             AWS_storage    id species  \\\n",
       "Sample                                                                   \n",
       "ARN-1167_M-4T1   s3://dp-lab-data/collaborators/aboir...  3249   mouse   \n",
       "ARN-1167_Normal  s3://dp-lab-data/collaborators/aboir...  3250   mouse   \n",
       "ARN-1167_PM-4T1  s3://dp-lab-data/collaborators/aboir...  3251   mouse   \n",
       "\n",
       "                  sc_tech                               project_id  \\\n",
       "Sample                                                               \n",
       "ARN-1167_M-4T1   10X_V3.1  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_Normal  10X_V3.1  Leptomeningeal metastasis heterogeneity   \n",
       "ARN-1167_PM-4T1  10X_V3.1  Leptomeningeal metastasis heterogeneity   \n",
       "\n",
       "                                               reference  \n",
       "Sample                                                    \n",
       "ARN-1167_M-4T1   s3://seqc-public/genomes/mm38_long_p...  \n",
       "ARN-1167_Normal  s3://seqc-public/genomes/mm38_long_p...  \n",
       "ARN-1167_PM-4T1  s3://seqc-public/genomes/mm38_long_p...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can supply a list of IDs or request IDs\n",
    "\n",
    "request_ids = ['ARN-1167']\n",
    "samples = sample_scridb_info(request_ids, 'request_id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b837d999-a7e3-431b-ab24-1076ed1876ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aa4f0df0-2405-4d1a-aaf2-d2f10246690e",
   "metadata": {},
   "source": [
    "# Manually add AWS path to fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d229087-fb55-479e-aa12-9a98f323e41d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2529ec18-37f3-42e9-a91a-8f0be5b0c2e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5b371e6a-d0c1-4d6e-bf7e-0af3a3317e53",
   "metadata": {},
   "source": [
    "# Make labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c6cfd3b-c96e-4605-8961-00f210584517",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "pipeline_type = prefix.split('Transgenes')[-1].lower()\n",
    "with open(f\"{config_dir}/template.{pipeline_type}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "with open(f\"{config_dir}/template.{pipeline_type}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=submissions.index, columns=std_inputs_fields,)\n",
    "labels = pd.DataFrame(index=submissions.index, columns=std_labels_fields,)\n",
    "\n",
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.referenceName\"] = \\\n",
    "    submissions[\"Project_Name\"] + \"-\" + \\\n",
    "    submissions[\"Parsed_Species\"].apply(lambda x: f\"{x[1]}-Ensembl-{x[2]}\") + \"-\" \\\n",
    "    \"transgenes\"\n",
    "inputs[f\"{prefix}.genomeReferenceFasta\"] = \\\n",
    "    submissions[\"Parsed_Species\"].apply(lambda x: ref_map[x[1:]][1])\n",
    "inputs[f\"{prefix}.annotationGtf\"] = \\\n",
    "    submissions[\"Parsed_Species\"].apply(lambda x: ref_map[x[1:]][0])\n",
    "inputs[f\"{prefix}.customFastaFiles\"] = \\\n",
    "     submissions[\"AWS S3 location(s) of the FASTA sequence of your reporter genes\"].apply(lambda x: x.split(\"\\n\"))\n",
    "inputs[f\"{prefix}.ensembleIdPrefix\"] = \\\n",
    "    submissions[\"Parsed_Species\"].apply(lambda x: ensembl_map[x[0]])\n",
    "inputs[f\"{prefix}.ensembleIds\"] = submissions[\"Ensembl ID of transgene\"].apply(lambda x: x.split(\"\\n\"))\n",
    "if pipeline_type == \"TransgenesSEQC\":\n",
    "    inputs[f\"{prefix}.starVersion\"] = latest_STAR\n",
    "else:\n",
    "    inputs[f\"{prefix}.cellRangerVersion\"] = latest_CellRanger\n",
    "standard_biotypes = [\n",
    "    \"protein_coding\", \"lincRNA\", \"antisense\",\n",
    "    \"IG_V_gene\", \"IG_D_gene\", \"IG_J_gene\", \"IG_C_gene\",\n",
    "    \"TR_V_gene\", \"TR_D_gene\", \"TR_J_gene\", \"TR_C_gene\",\n",
    "]\n",
    "inputs[f\"{prefix}.biotypes\"] = inputs.iloc[:, 0].apply(lambda x: standard_biotypes)\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be14a08e-71f3-41ea-b716-e9e69ee7c44c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = prefix\n",
    "labels[\"project\"] = submissions[\"Project_Name\"]\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = \\\n",
    "    submissions[\"AWS S3 location where you want the newly built genome to be stored\"].str.strip(\"/\") + \\\n",
    "    \"/\" + output_dirname\n",
    "    \n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
