{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f92e2e4f-6704-4b3d-b8ec-c51427a7ca0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "474b1723-a2e0-4556-bd36-146ed07c1bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3c09f4af-4524-4a93-995f-a268efba3ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b571064-b110-4a4d-b182-bbcb19af6fc8",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b3a514-cc78-40a4-87ff-0dfd817282e0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## List FASTQs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "022a6910-115b-4d46-b4bd-511b7a890893",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'MitoTracing': ['R1', 'R2'],\n",
    "}\n",
    "\n",
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ae81c97b-0e5b-4aae-ae8a-977ca6384b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_barcode_genomic_fastqs(samples):\n",
    "    \n",
    "    samples['fastq'] = np.empty((len(samples), 0)).tolist()\n",
    "    fastqs = []\n",
    "    for sample, row in samples.iterrows():\n",
    "        S3_path = row['S3_path']\n",
    "        \n",
    "        # List FASTQ files\n",
    "        fastq_file_ids = fastq_map['MitoTracing']        \n",
    "        path_fastq = f\"{S3_path}/FASTQ/\"\n",
    "        fastq = get_fastqs(path_fastq, fastq_file_ids)\n",
    "        \n",
    "        fastqs.append(fastq)\n",
    "        \n",
    "    samples['fastq'] = fastqs\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd638296-c86e-4bdb-87ea-37983cdfbf89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract FASTQ sample name from list of files\n",
    "# Note: FASTQ name is file name up to lane id (e.g. L001, L002, etc.)\n",
    "def get_fastqs_name(fastqs):\n",
    "    fastq_name_re = r\".*/(.*)_S\\d+_L\\d{3}_[A-Za-z]\\d_\\d{3}.fastq.gz$\"\n",
    "    fastq_names = [re.match(fastq_name_re, x)[1] for x in fastqs]\n",
    "    assert len(set(fastq_names)) == 1 # make sure all names are same\n",
    "    return fastq_names[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6613242-152c-4b54-86c1-b245f3897c23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5543869f-37e4-4d3b-a8d6-877973abb6ee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## DB queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3ca3bfa5-69fd-4d5f-9d45-3d5ac61c76c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "def get_sample_name(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Sample\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_names = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_names.append(result[0])\n",
    "        return sample_names\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "    \n",
    "    \n",
    "def get_aws_path(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_paths = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_paths.append(result[0])\n",
    "        return sample_paths\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "            \n",
    "def get_sample_id(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_ids = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_ids.append(result[0])\n",
    "        return sample_ids\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def format_sample_aws(querys, query_col, creds):\n",
    "    sample_names = []\n",
    "    sample_paths = []\n",
    "    sample_ids = []\n",
    "    \n",
    "    for query in querys:\n",
    "        sample_names += get_sample_name(query, query_col, creds)\n",
    "        sample_paths += get_aws_path(query, query_col, creds)\n",
    "        sample_ids += get_sample_id(query, query_col, creds)\n",
    "        \n",
    "    sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "    \n",
    "    samples = pd.DataFrame(\n",
    "        dict(S3_path=sample_paths, Sample_ID=sample_ids),\n",
    "        index=sample_names,\n",
    "        dtype=str,\n",
    "    )\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "25c817f6-9a33-4fcf-84b8-1d4ebbcb66ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomeIndex_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.genomeIndex_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            genomeIndex_id = result[0]\n",
    "            \n",
    "        return genomeIndex_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_scTech_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.scTech_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            scTech_id = result[0]\n",
    "            \n",
    "        return scTech_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_index(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.gIndex\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            index = result[0]\n",
    "            \n",
    "        return index\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_assay(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Run_name\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            assay = result[0]\n",
    "            \n",
    "        return assay\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_barcode(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.barcodes\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            barcode = result[0]\n",
    "            \n",
    "        return barcode\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "        \n",
    "def format_assay_barcode(samples, creds):\n",
    "    \n",
    "    # samples['genomeIndex_id'] = np.nan\n",
    "    samples['index'] = np.nan\n",
    "    # samples['scTech_id'] = np.nan\n",
    "    samples['assay'] = np.nan\n",
    "    samples['barcode'] = np.nan\n",
    "    \n",
    "    for sample, row in samples.iterrows():\n",
    "        sample_id = row['Sample_ID']\n",
    "        \n",
    "        genomeIndex_id  = get_genomeIndex_id(sample_id, 'id', creds)\n",
    "        index = get_index(genomeIndex_id, 'id', creds)\n",
    "        scTech_id = get_scTech_id(genomeIndex_id, 'id', creds)\n",
    "        assay = get_assay(scTech_id, 'id', creds)\n",
    "        barcode = get_barcode(scTech_id, 'id', creds)\n",
    "        \n",
    "        # samples.loc[sample, 'genomeIndex_id'] = genomeIndex_id\n",
    "        samples.loc[sample, 'index'] = index\n",
    "        # samples.loc[sample, 'scTech_id'] = scTech_id\n",
    "        samples.loc[sample, 'assay'] = assay\n",
    "        samples.loc[sample, 'barcode'] = barcode\n",
    "\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "eb122238-79ee-4ecb-90d1-b2ea32e2b74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_project_id(sample_id, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fd74425-41d0-402c-9998-c3ed2c9f6aec",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c513b25-1067-4dfd-9ddc-092166012f7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e1c96c14-0cf1-4ab2-9beb-417bc02f1154",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"MitoTracing\": \"mito-tracing-outs\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"MitoTracing\": \"submit-processing.sh\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8ae3c158-3666-426f-94be-f9165e30c82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"MitoTracing\" # Workflow to run; also .wdl filename prefix\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "output_dirname = results_dirs[prefix]\n",
    "template_prefix = 'template'\n",
    "\n",
    "# If need to add comment, put here\n",
    "comment = \"sohailn\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "61658425-6370-44c2-92c2-a3a977345bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = glob.glob(f\"{Path.home()}/scing/bin/*mito*\")[0]\n",
    "path_to_exec = sh_files[prefix]\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = glob.glob(f\"{workflow_dir}/*.options.aws.json\")[0]\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bb4d448a-5310-4b63-a073-66cf094e4467",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e2bc63-cf1f-4cf5-899b-f1e73c56f5aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7c4de0dd-3c84-48b0-8e38-38b83d2abef8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: AWS_ACCESS_KEY_ID=ASIAYPY66CWMPLW5T6NP\n",
      "env: AWS_SECRET_ACCESS_KEY=1SX9bRl+eDSWhPQZoE3veOnjWMS1q4pOi+H2h4CF\n",
      "env: AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEKf//////////wEaCXVzLWVhc3QtMSJGMEQCIG8pw61xL0cjlnxgAmlhtJuGvjKrly9zppW9jmxWKct1AiBSCMeORMMiw34vMpf2Dy7lfNTer4K0ZYHjiIDezosCPSr4AQjw//////////8BEAMaDDU4MzY0MzU2NzUxMiIMcwOotqu75F46WP4sKswBtpwzuDdeano9gSn3VxI01yB4h7fxYVnQQiREf7FBGcmfYrMJ7B9DnL62fHTzdFdNRNsW9BETcVnuQrW5rIEtUJ50XQbsoWsRDBdF9DSXJL26qMr0l7ZOs4IkMn/MYfuGbN9OwRolOvPAh8ewB1N5NqOR+8QO29kRrshV11UBx5CfZwAL00os+sPJgweB+pWsniVheoPRaJLonnsTisSfQYRzoidAVL4ZkZv48eFzaodcAGD5M8VbZQpZ3njSQrmqKQ0jYP7fH0PytXxOMK3FipcGOpkB+HPz9ENB0ACPcyAAz9L3z3uWF/IqP8Q+UbU8Mov3FXhSQuTLEzdZQUxMtBiTT8MAFxyQKabxCrUKpctEWTFL6TeXIoexFp8BmqetPNFP0xwCNGEN/E6KaXs9PKQi5hak1EFX2TVvg3w8DJhsy9ApMktj/wNiU8+ZQvGXX2xIW/pakOBI4H9qoOwC9IIXofUbSFpino3BKg6w\n",
      "2021-10-07 15:31:32 agc-583643567512-us-east-1\n",
      "2021-10-07 15:28:07 cdktoolkit-stagingbucket-d49u1xfb0sc4\n",
      "2019-10-09 13:04:06 cf-templates-umiwbnq3566w-us-east-1\n",
      "2019-10-10 12:46:54 dp-daily-reports\n",
      "2017-09-26 10:15:15 dp-lab-data\n",
      "2019-05-23 12:34:58 dp-lab-data-public\n",
      "2018-03-01 13:29:07 dp-lab-glacier\n",
      "2020-11-06 23:02:27 dp-lab-gwf-core\n",
      "2021-07-15 18:14:01 dp-lab-gwf-core2\n",
      "2017-09-26 10:17:18 dp-lab-home\n",
      "2019-08-14 16:42:43 dp-lab-test\n",
      "2019-04-25 12:35:35 elasticbeanstalk-us-east-1-583643567512\n",
      "2019-12-13 15:53:57 scri-computational\n",
      "2017-09-26 10:23:50 seqc-public\n"
     ]
    }
   ],
   "source": [
    "%env AWS_ACCESS_KEY_ID=ASIAYPY66CWMPLW5T6NP\n",
    "%env AWS_SECRET_ACCESS_KEY=1SX9bRl+eDSWhPQZoE3veOnjWMS1q4pOi+H2h4CF\n",
    "%env AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEKf//////////wEaCXVzLWVhc3QtMSJGMEQCIG8pw61xL0cjlnxgAmlhtJuGvjKrly9zppW9jmxWKct1AiBSCMeORMMiw34vMpf2Dy7lfNTer4K0ZYHjiIDezosCPSr4AQjw//////////8BEAMaDDU4MzY0MzU2NzUxMiIMcwOotqu75F46WP4sKswBtpwzuDdeano9gSn3VxI01yB4h7fxYVnQQiREf7FBGcmfYrMJ7B9DnL62fHTzdFdNRNsW9BETcVnuQrW5rIEtUJ50XQbsoWsRDBdF9DSXJL26qMr0l7ZOs4IkMn/MYfuGbN9OwRolOvPAh8ewB1N5NqOR+8QO29kRrshV11UBx5CfZwAL00os+sPJgweB+pWsniVheoPRaJLonnsTisSfQYRzoidAVL4ZkZv48eFzaodcAGD5M8VbZQpZ3njSQrmqKQ0jYP7fH0PytXxOMK3FipcGOpkB+HPz9ENB0ACPcyAAz9L3z3uWF/IqP8Q+UbU8Mov3FXhSQuTLEzdZQUxMtBiTT8MAFxyQKabxCrUKpctEWTFL6TeXIoexFp8BmqetPNFP0xwCNGEN/E6KaXs9PKQi5hak1EFX2TVvg3w8DJhsy9ApMktj/wNiU8+ZQvGXX2xIW/pakOBI4H9qoOwC9IIXofUbSFpino3BKg6w\n",
    "!aws s3 ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f6c614-d304-433c-a318-81182ba675b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2b28002d-85c9-4b9b-b0fc-5b86948c245c",
   "metadata": {},
   "source": [
    "# Excution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b988b1-df2f-4b32-a11b-786a188ba7b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2149bf16-9464-4654-bafd-a5b409f15321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1759_Ru1083_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1760_MSK_LX_1083c_T_2_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1761_POSIE_101920_T_1_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1762_Ru1083d_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1763_Ru1250C_T_1_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1764_MSK_LX_1250b_PM_1_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1764_Ru1250D_T_1_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1766_MSK_LX_1250f_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1760_Ru263_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                S3_path  \\\n",
       "AV-1759_Ru1083_MITO             s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1761_POSIE_101920_T_1_MITO   s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1762_Ru1083d_MITO            s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1763_Ru1250C_T_1_MITO        s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1764_Ru1250D_T_1_MITO        s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1765_Ru1250e_MITO            s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1766_MSK_LX_1250f_MITO       s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1760_Ru263_MITO              s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                               Sample_ID  \n",
       "AV-1759_Ru1083_MITO                 3924  \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO       3925  \n",
       "AV-1761_POSIE_101920_T_1_MITO       3926  \n",
       "AV-1762_Ru1083d_MITO                3927  \n",
       "AV-1763_Ru1250C_T_1_MITO            3928  \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO      3929  \n",
       "AV-1764_Ru1250D_T_1_MITO            3930  \n",
       "AV-1765_Ru1250e_MITO                3931  \n",
       "AV-1766_MSK_LX_1250f_MITO           3932  \n",
       "AV-1760_Ru263_MITO                  3933  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "sample_id = list(range(3924, 3934))\n",
    "\n",
    "samples = format_sample_aws(sample_id, 'id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "40640f5a-1e75-4947-98ba-05aec427e5e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Modification for Joe's samples\n",
    "# samples.loc['Ru581b_T1_MITO', 'S3_path'] = 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581b_T1_MITO'\n",
    "# samples.loc['Ru581c-LN1_MITO', 'S3_path'] = 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581c-LN1_MITO'\n",
    "\n",
    "# samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dc80d4e5-c699-4be7-96f4-0475e3f1f12d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3931</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      S3_path Sample_ID\n",
       "AV-1765_Ru1250e_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...      3931"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Re-running some of the samples\n",
    "# samples = samples.loc[samples.index.isin(['AV-1760_Ru263_MITO', 'AV-1762_Ru1083d_MITO'])]\n",
    "\n",
    "# barcodes_in_RNA.non-epithelial.RU1083_ST.txt\n",
    "# samples = samples.loc[samples.index.isin(['AV-1762_Ru1083d_MITO'])]\n",
    "\n",
    "\n",
    "# barcodes_in_RNA.epithelial.RU1250_ASC1.txt\n",
    "samples = samples.loc[samples.index.isin(['AV-1765_Ru1250e_MITO'])]\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "359fbd22-702d-4763-85c2-0410b0611bb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/2360911148.py:110: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['index'] = np.nan\n",
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/2360911148.py:112: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['assay'] = np.nan\n",
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/2360911148.py:113: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['barcode'] = np.nan\n",
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/2360911148.py:125: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples.loc[sample, 'index'] = index\n",
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/2360911148.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples.loc[sample, 'assay'] = assay\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3931</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      S3_path Sample_ID  \\\n",
       "AV-1765_Ru1250e_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...      3931   \n",
       "\n",
       "                             index assay  barcode  \n",
       "AV-1765_Ru1250e_MITO  GRCh38-3.0.0    CR      NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = format_assay_barcode(samples, creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a14b9418-dab4-4aaf-8abb-14f9e0534152",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/638862739.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['fastq'] = np.empty((len(samples), 0)).tolist()\n",
      "/var/folders/9v/1lnyrkxd0yq8l2pgl002wh1s7c6fqv/T/ipykernel_34982/638862739.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['fastq'] = fastqs\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>index</th>\n",
       "      <th>assay</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fastq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>3931</td>\n",
       "      <td>GRCh38-3.0.0</td>\n",
       "      <td>CR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/SCRI_Proje...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      S3_path Sample_ID  \\\n",
       "AV-1765_Ru1250e_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...      3931   \n",
       "\n",
       "                             index assay  barcode  \\\n",
       "AV-1765_Ru1250e_MITO  GRCh38-3.0.0    CR      NaN   \n",
       "\n",
       "                                                        fastq  \n",
       "AV-1765_Ru1250e_MITO  {'R1': ['s3://dp-lab-data/SCRI_Proje...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = get_barcode_genomic_fastqs(samples)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bc16e8ed-5503-421b-9ce9-b31496670188",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples = pd.DataFrame(samples.loc['Ru581b_T1_MITO']).T\n",
    "# samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85065328-a68e-46a7-9304-88953148416e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ce68536f-7029-4689-83a7-239421284398",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MitoTracing.sampleName MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO                    NaN                   NaN   \n",
       "\n",
       "                     MitoTracing.fastqR1 MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO                 NaN                 NaN   \n",
       "\n",
       "                     MitoTracing.reference MitoTracing.includeIntrons  \\\n",
       "AV-1765_Ru1250e_MITO                   NaN                        NaN   \n",
       "\n",
       "                     MitoTracing.expectCells MitoTracing.numCores  \\\n",
       "AV-1765_Ru1250e_MITO                     NaN                  NaN   \n",
       "\n",
       "                     MitoTracing.memory MitoTracing.whitelist  \\\n",
       "AV-1765_Ru1250e_MITO                NaN                   NaN   \n",
       "\n",
       "                     MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO                        NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from input template\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bd8a0e97-49ad-4295-a4ff-ca83e90732c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MitoTracing.sampleName MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO                    NaN                   NaN   \n",
       "\n",
       "                     MitoTracing.fastqR1 MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO                 NaN                 NaN   \n",
       "\n",
       "                     MitoTracing.reference  MitoTracing.includeIntrons  \\\n",
       "AV-1765_Ru1250e_MITO                   NaN                       False   \n",
       "\n",
       "                      MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "AV-1765_Ru1250e_MITO                     5000                    16   \n",
       "\n",
       "                      MitoTracing.memory MitoTracing.whitelist  \\\n",
       "AV-1765_Ru1250e_MITO                 256                   NaN   \n",
       "\n",
       "                     MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Default inputs\n",
    "inputs[f\"{prefix}.includeIntrons\"] = False\n",
    "inputs[f\"{prefix}.expectCells\"] = 5000\n",
    "inputs[f\"{prefix}.numCores\"] = 16\n",
    "# inputs[f\"{prefix}.memory\"] = 128\n",
    "inputs[f\"{prefix}.memory\"] = 256 \n",
    "\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "db50eb55-40dc-4840-bc61-63525c2afb2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>AV-1765_Ru1250e_MITO</td>\n",
       "      <td>3931_AV-1765_Ru1250e_MITO_IGO_13388_9</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MitoTracing.sampleName  \\\n",
       "AV-1765_Ru1250e_MITO   AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                                      MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO  3931_AV-1765_Ru1250e_MITO_IGO_13388_9   \n",
       "\n",
       "                                          MitoTracing.fastqR1  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                          MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                     MitoTracing.reference  MitoTracing.includeIntrons  \\\n",
       "AV-1765_Ru1250e_MITO                   NaN                       False   \n",
       "\n",
       "                      MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "AV-1765_Ru1250e_MITO                     5000                    16   \n",
       "\n",
       "                      MitoTracing.memory MitoTracing.whitelist  \\\n",
       "AV-1765_Ru1250e_MITO                 256                   NaN   \n",
       "\n",
       "                     MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample information\n",
    "inputs[f\"{prefix}.sampleName\"] = samples.index.tolist()\n",
    "\n",
    "inputs[f\"{prefix}.fastqR1\"] = samples['fastq'].str['R1']\n",
    "inputs[f\"{prefix}.fastqR2\"] = samples['fastq'].str['R2']\n",
    "\n",
    "for sample in inputs.index:\n",
    "    fastqR1_name = get_fastqs_name(inputs.loc[sample, 'MitoTracing.fastqR1'])\n",
    "    fastqR2_name = get_fastqs_name(inputs.loc[sample, 'MitoTracing.fastqR2'])\n",
    "    \n",
    "    assert(fastqR1_name == fastqR1_name)\n",
    "    inputs.loc[sample, f\"{prefix}.fastqName\"] = fastqR1_name\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "101eb0d4-4ca6-4639-bb67-afe47109a251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>AV-1765_Ru1250e_MITO</td>\n",
       "      <td>3931_AV-1765_Ru1250e_MITO_IGO_13388_9</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MitoTracing.sampleName  \\\n",
       "AV-1765_Ru1250e_MITO   AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                                      MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO  3931_AV-1765_Ru1250e_MITO_IGO_13388_9   \n",
       "\n",
       "                                          MitoTracing.fastqR1  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                          MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                        MitoTracing.reference  \\\n",
       "AV-1765_Ru1250e_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                      MitoTracing.includeIntrons  MitoTracing.expectCells  \\\n",
       "AV-1765_Ru1250e_MITO                       False                     5000   \n",
       "\n",
       "                      MitoTracing.numCores  MitoTracing.memory  \\\n",
       "AV-1765_Ru1250e_MITO                    16                 256   \n",
       "\n",
       "                     MitoTracing.whitelist MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO                   NaN            quay.io/hisplan  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reference package created by Jaeyoung\n",
    "\n",
    "cellRangerRefPkg=\"s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/reference-packages/refdata-gex-mito-GRCh38-ensemble98.tar.gz\"\n",
    "mitoFastaRefPkg=\"s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/reference-packages/mito-fasta-GRCh38.tar.gz\"\n",
    "\n",
    "reference = [dict(cellRangerRefPkg=cellRangerRefPkg,\n",
    "                  mitoFastaRefPkg=mitoFastaRefPkg)] * len(inputs)\n",
    "inputs[f\"{prefix}.reference\"] = reference\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f12841-72fe-49c6-b91c-2381f7fc62be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6581c13e-4156-4082-b9a0-14bdfd420b1e",
   "metadata": {},
   "source": [
    "## Check whitelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "03cf4850-3447-4f8a-acc0-87de499ac2f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1084d07a-a0e7-4274-81eb-2c2d3de4ecdf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "46cd6c61-b55a-4c3b-b00a-0f88df9a3b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "epithelial = True\n",
    "for sample in inputs.index:\n",
    "    for w in whitelist[sample]:\n",
    "        \n",
    "        if epithelial:\n",
    "            if '.epithelial.' in w:\n",
    "                sample_whitelist = w\n",
    "        else:\n",
    "            if '.non-epithelial.' in w:\n",
    "                sample_whitelist = w\n",
    "          \n",
    "    inputs.loc[sample, f\"{prefix}.whitelist\"] = sample_whitelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5bd15e40-eeae-4b8f-80fa-1975f89ba0b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>AV-1765_Ru1250e_MITO</td>\n",
       "      <td>3931_AV-1765_Ru1250e_MITO_IGO_13388_9</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>256</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MitoTracing.sampleName  \\\n",
       "AV-1765_Ru1250e_MITO   AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                                      MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO  3931_AV-1765_Ru1250e_MITO_IGO_13388_9   \n",
       "\n",
       "                                          MitoTracing.fastqR1  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                          MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                        MitoTracing.reference  \\\n",
       "AV-1765_Ru1250e_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                      MitoTracing.includeIntrons  MitoTracing.expectCells  \\\n",
       "AV-1765_Ru1250e_MITO                       False                     5000   \n",
       "\n",
       "                      MitoTracing.numCores  MitoTracing.memory  \\\n",
       "AV-1765_Ru1250e_MITO                    16                 256   \n",
       "\n",
       "                                        MitoTracing.whitelist  \\\n",
       "AV-1765_Ru1250e_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                     MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3371ad80-d08d-4d2a-aeba-8442cd8a16e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1ae5805b-8806-4846-a9ed-3b7400e99e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# non_epithelial = ['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581b_T1_MITO/barcodes_in_RNA.non-epithelial.RU581_Ta.txt',\n",
    "#                   's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581c-LN1_MITO/barcodes_in_RNA.non-epithelial.RU581_LNa.txt',\n",
    "#                   's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581D_MITO/barcodes_in_RNA.non-epithelial.RU581_LIV.txt']\n",
    "\n",
    "# epithelial = ['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581b_T1_MITO/barcodes_in_RNA.RU581_Ta.txt',\n",
    "#               's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581c-LN1_MITO/barcodes_in_RNA.RU581_LNa.txt',\n",
    "#               's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581D_MITO/barcodes_in_RNA.RU581_LIV.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b07e2dd7-8ed4-4dd1-9faa-a3fe7bfd99b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Whitelist needs to be in SEQC barcodes (nucleotide) format rather than CR format\n",
    "# # Path to download the whitelist file\n",
    "# path_whitelist = '/Users/sohailn/scing/bin/wdl-mito-tracing/whitelist/'\n",
    "\n",
    "# # for path_s3 in epithelial:\n",
    "# for path_s3 in non_epithelial:\n",
    "#     path_s3_dirname = os.path.dirname(path_s3)\n",
    "#     sample_name = os.path.basename(path_s3_dirname)\n",
    "\n",
    "#     filename_in = os.path.basename(path_s3)\n",
    "#     filename_out = filename_in.replace('.txt', '.seqc.txt')\n",
    "\n",
    "#     # Download barcode file\n",
    "#     cmd = f'aws s3 cp {path_s3} {path_whitelist}{filename_in}'\n",
    "#     os.system(cmd)\n",
    "\n",
    "#     barcodes = pd.read_csv(f'{path_whitelist}{filename_in}', header=None)[0].tolist()\n",
    "\n",
    "#     from seqc.sequence.encodings import DNA3Bit\n",
    "#     dna3bit = DNA3Bit()\n",
    "#     barcodes = [dna3bit.decode(x) for x in barcodes]\n",
    "    \n",
    "#     with open(f'{path_whitelist}{filename_out}', \"w\") as output:\n",
    "#         for barcode in barcodes:\n",
    "#             output.write(f'{barcode.decode(\"utf-8\")}-1\\n')\n",
    "\n",
    "#     cmd = f'aws s3 cp {path_whitelist}{filename_out} {path_s3_dirname}/{filename_out}'\n",
    "#     inputs.loc[sample_name, f\"{prefix}.whitelist\"] = f'{path_s3_dirname}/{filename_out}'\n",
    "    \n",
    "#     os.system(cmd)\n",
    "#     print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3c3633c6-cc9a-43f0-a4fa-b5b2cfc2a76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputs.loc['Ru581b_T1_MITO', f\"{prefix}.whitelist\"] = 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581b_T1_MITO/barcodes_in_RNA.RU581_Ta.seqc.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46fa392b-d4f4-4c47-aa8f-d8c7d980f62f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b019afa4-15b1-4e1d-8eb1-9ba176088dee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/barcodes_in_RNA.epithelial.RU1250_ASC1.txt'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs[f\"{prefix}.whitelist\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "eee5a4c1-8dce-4975-811c-57ed30da1994",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['epi_AV-1765_Ru1250e_MITO'], dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if epithelial:\n",
    "    inputs[f\"{prefix}.sampleName\"] = 'epi_' + inputs[f\"{prefix}.sampleName\"] \n",
    "else:\n",
    "    inputs[f\"{prefix}.sampleName\"] = 'non_epi_' + inputs[f\"{prefix}.sampleName\"]\n",
    "inputs[f\"{prefix}.sampleName\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9af5c62-35fe-4d20-a198-9819aeb2c680",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0f737133-5115-4ee6-bf93-5443e5c652d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Make label file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6ca8c556-8254-4051-93ee-5447d43abfbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     pipelineType project sample owner destination transfer  \\\n",
       "AV-1765_Ru1250e_MITO          NaN     NaN    NaN   NaN         NaN      NaN   \n",
       "\n",
       "                     comment  \n",
       "AV-1765_Ru1250e_MITO     NaN  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from labels template\n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow labels\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0731efde-30bb-419b-b5a4-b8420911f309",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>Lung Tumor Atlas</td>\n",
       "      <td>AV-1765_Ru1250e_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     pipelineType           project                sample  \\\n",
       "AV-1765_Ru1250e_MITO  MitoTracing  Lung Tumor Atlas  AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                        owner                              destination  \\\n",
       "AV-1765_Ru1250e_MITO  sohailn  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                     transfer  comment  \n",
       "AV-1765_Ru1250e_MITO        -  sohailn  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "# labels[\"destination\"] = samples['S3_path'] + \"/\" + output_dirname\n",
    "\n",
    "if epithelial:\n",
    "    labels[\"destination\"] = samples['S3_path'] + \"/\" + 'epi_' + output_dirname\n",
    "else:\n",
    "    labels[\"destination\"] = samples['S3_path'] + \"/\" + 'non_epi_' + output_dirname\n",
    "\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "82ab0b61-8acf-43d3-97ae-3e53c528e890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/epi_mito-tracing-outs'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# labels.loc['Ru581D_MITO', 'destination'] = 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581D_MITO/non_epi_mito-tracing-outs'\n",
    "# labels.loc['Ru581D_MITO', 'destination'] = 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/barcodes_rna/Ru581D_MITO/epi_mito-tracing-outs'\n",
    "\n",
    "labels['destination'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5973e09-e9bf-4415-acee-9e1809afdf5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "84007e1e-ae11-4665-ba47-aab5b803af27",
   "metadata": {},
   "source": [
    "# Submit job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "eb40233f-7d95-4e66-91b8-933d77265ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs_all = inputs.copy()\n",
    "labels_all = labels.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "56fb192f-46d2-401d-bf8e-7bf65551364c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AV-1759_Ru1083_MITO',\n",
       " 'AV-1760_MSK_LX_1083c_T_2_MITO',\n",
       " 'AV-1760_Ru263_MITO',\n",
       " 'AV-1761_POSIE_101920_T_1_MITO',\n",
       " 'AV-1762_Ru1083d_MITO',\n",
       " 'AV-1763_Ru1250C_T_1_MITO',\n",
       " 'AV-1764_MSK_LX_1250b_PM_1_MITO',\n",
       " 'AV-1764_Ru1250D_T_1_MITO',\n",
       " 'AV-1765_Ru1250e_MITO',\n",
       " 'AV-1766_MSK_LX_1250f_MITO']"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_names = inputs.index.tolist()\n",
    "sample_names.sort()\n",
    "sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "ca0f8bc8-c862-4398-89b2-152ad55bdec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1759_Ru1083_MITO</th>\n",
       "      <td>non_epi_AV-1759_Ru1083_MITO</td>\n",
       "      <td>3924_AV-1759_Ru1083_MITO_IGO_13388_1</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1760_MSK_LX_1083c_T_2_MITO</th>\n",
       "      <td>non_epi_AV-1760_MSK_LX_1083c_T_2_MITO</td>\n",
       "      <td>3925_AV-1760_MSK_LX_1083c_T_2_MITO_I...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1760_Ru263_MITO</th>\n",
       "      <td>non_epi_AV-1760_Ru263_MITO</td>\n",
       "      <td>3933_AV-1760_Ru263_MITO_IGO_13388_7</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1761_POSIE_101920_T_1_MITO</th>\n",
       "      <td>non_epi_AV-1761_POSIE_101920_T_1_MITO</td>\n",
       "      <td>3926_AV-1761_POSIE_101920_T_1_MITO_I...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1762_Ru1083d_MITO</th>\n",
       "      <td>non_epi_AV-1762_Ru1083d_MITO</td>\n",
       "      <td>3927_AV-1762_Ru1083d_MITO_IGO_13388_4</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1763_Ru1250C_T_1_MITO</th>\n",
       "      <td>non_epi_AV-1763_Ru1250C_T_1_MITO</td>\n",
       "      <td>3928_AV-1763_Ru1250C_T_1_MITO_IGO_13...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1764_MSK_LX_1250b_PM_1_MITO</th>\n",
       "      <td>non_epi_AV-1764_MSK_LX_1250b_PM_1_MITO</td>\n",
       "      <td>3929_AV-1764_MSK_LX_1250b_PM_1_MITO_...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1764_Ru1250D_T_1_MITO</th>\n",
       "      <td>non_epi_AV-1764_Ru1250D_T_1_MITO</td>\n",
       "      <td>3930_AV-1764_Ru1250D_T_1_MITO_IGO_13...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>non_epi_AV-1765_Ru1250e_MITO</td>\n",
       "      <td>3931_AV-1765_Ru1250e_MITO_IGO_13388_9</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV-1766_MSK_LX_1250f_MITO</th>\n",
       "      <td>non_epi_AV-1766_MSK_LX_1250f_MITO</td>\n",
       "      <td>3932_AV-1766_MSK_LX_1250f_MITO_IGO_1...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>128</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                MitoTracing.sampleName  \\\n",
       "AV-1759_Ru1083_MITO                        non_epi_AV-1759_Ru1083_MITO   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO    non_epi_AV-1760_MSK_LX_1083c_T_2_MITO   \n",
       "AV-1760_Ru263_MITO                          non_epi_AV-1760_Ru263_MITO   \n",
       "AV-1761_POSIE_101920_T_1_MITO    non_epi_AV-1761_POSIE_101920_T_1_MITO   \n",
       "AV-1762_Ru1083d_MITO                      non_epi_AV-1762_Ru1083d_MITO   \n",
       "AV-1763_Ru1250C_T_1_MITO              non_epi_AV-1763_Ru1250C_T_1_MITO   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  non_epi_AV-1764_MSK_LX_1250b_PM_1_MITO   \n",
       "AV-1764_Ru1250D_T_1_MITO              non_epi_AV-1764_Ru1250D_T_1_MITO   \n",
       "AV-1765_Ru1250e_MITO                      non_epi_AV-1765_Ru1250e_MITO   \n",
       "AV-1766_MSK_LX_1250f_MITO            non_epi_AV-1766_MSK_LX_1250f_MITO   \n",
       "\n",
       "                                                  MitoTracing.fastqName  \\\n",
       "AV-1759_Ru1083_MITO                3924_AV-1759_Ru1083_MITO_IGO_13388_1   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   3925_AV-1760_MSK_LX_1083c_T_2_MITO_I...   \n",
       "AV-1760_Ru263_MITO                  3933_AV-1760_Ru263_MITO_IGO_13388_7   \n",
       "AV-1761_POSIE_101920_T_1_MITO   3926_AV-1761_POSIE_101920_T_1_MITO_I...   \n",
       "AV-1762_Ru1083d_MITO              3927_AV-1762_Ru1083d_MITO_IGO_13388_4   \n",
       "AV-1763_Ru1250C_T_1_MITO        3928_AV-1763_Ru1250C_T_1_MITO_IGO_13...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  3929_AV-1764_MSK_LX_1250b_PM_1_MITO_...   \n",
       "AV-1764_Ru1250D_T_1_MITO        3930_AV-1764_Ru1250D_T_1_MITO_IGO_13...   \n",
       "AV-1765_Ru1250e_MITO              3931_AV-1765_Ru1250e_MITO_IGO_13388_9   \n",
       "AV-1766_MSK_LX_1250f_MITO       3932_AV-1766_MSK_LX_1250f_MITO_IGO_1...   \n",
       "\n",
       "                                                    MitoTracing.fastqR1  \\\n",
       "AV-1759_Ru1083_MITO             [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1760_Ru263_MITO              [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1761_POSIE_101920_T_1_MITO   [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1762_Ru1083d_MITO            [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1763_Ru1250C_T_1_MITO        [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1764_Ru1250D_T_1_MITO        [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1765_Ru1250e_MITO            [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1766_MSK_LX_1250f_MITO       [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                                    MitoTracing.fastqR2  \\\n",
       "AV-1759_Ru1083_MITO             [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1760_Ru263_MITO              [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1761_POSIE_101920_T_1_MITO   [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1762_Ru1083d_MITO            [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1763_Ru1250C_T_1_MITO        [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1764_Ru1250D_T_1_MITO        [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1765_Ru1250e_MITO            [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "AV-1766_MSK_LX_1250f_MITO       [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                                  MitoTracing.reference  \\\n",
       "AV-1759_Ru1083_MITO             {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1760_Ru263_MITO              {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1761_POSIE_101920_T_1_MITO   {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1762_Ru1083d_MITO            {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1763_Ru1250C_T_1_MITO        {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1764_Ru1250D_T_1_MITO        {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1765_Ru1250e_MITO            {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "AV-1766_MSK_LX_1250f_MITO       {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                                MitoTracing.includeIntrons  \\\n",
       "AV-1759_Ru1083_MITO                                  False   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO                        False   \n",
       "AV-1760_Ru263_MITO                                   False   \n",
       "AV-1761_POSIE_101920_T_1_MITO                        False   \n",
       "AV-1762_Ru1083d_MITO                                 False   \n",
       "AV-1763_Ru1250C_T_1_MITO                             False   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO                       False   \n",
       "AV-1764_Ru1250D_T_1_MITO                             False   \n",
       "AV-1765_Ru1250e_MITO                                 False   \n",
       "AV-1766_MSK_LX_1250f_MITO                            False   \n",
       "\n",
       "                                MitoTracing.expectCells  MitoTracing.numCores  \\\n",
       "AV-1759_Ru1083_MITO                                5000                    16   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO                      5000                    16   \n",
       "AV-1760_Ru263_MITO                                 5000                    16   \n",
       "AV-1761_POSIE_101920_T_1_MITO                      5000                    16   \n",
       "AV-1762_Ru1083d_MITO                               5000                    16   \n",
       "AV-1763_Ru1250C_T_1_MITO                           5000                    16   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO                     5000                    16   \n",
       "AV-1764_Ru1250D_T_1_MITO                           5000                    16   \n",
       "AV-1765_Ru1250e_MITO                               5000                    16   \n",
       "AV-1766_MSK_LX_1250f_MITO                          5000                    16   \n",
       "\n",
       "                                MitoTracing.memory  \\\n",
       "AV-1759_Ru1083_MITO                            128   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO                  128   \n",
       "AV-1760_Ru263_MITO                             128   \n",
       "AV-1761_POSIE_101920_T_1_MITO                  128   \n",
       "AV-1762_Ru1083d_MITO                           128   \n",
       "AV-1763_Ru1250C_T_1_MITO                       128   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO                 128   \n",
       "AV-1764_Ru1250D_T_1_MITO                       128   \n",
       "AV-1765_Ru1250e_MITO                           128   \n",
       "AV-1766_MSK_LX_1250f_MITO                      128   \n",
       "\n",
       "                                                  MitoTracing.whitelist  \\\n",
       "AV-1759_Ru1083_MITO             s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO   s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1760_Ru263_MITO              s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1761_POSIE_101920_T_1_MITO   s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1762_Ru1083d_MITO            s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1763_Ru1250C_T_1_MITO        s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1764_Ru1250D_T_1_MITO        s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1765_Ru1250e_MITO            s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "AV-1766_MSK_LX_1250f_MITO       s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                               MitoTracing.dockerRegistry  \n",
       "AV-1759_Ru1083_MITO                       quay.io/hisplan  \n",
       "AV-1760_MSK_LX_1083c_T_2_MITO             quay.io/hisplan  \n",
       "AV-1760_Ru263_MITO                        quay.io/hisplan  \n",
       "AV-1761_POSIE_101920_T_1_MITO             quay.io/hisplan  \n",
       "AV-1762_Ru1083d_MITO                      quay.io/hisplan  \n",
       "AV-1763_Ru1250C_T_1_MITO                  quay.io/hisplan  \n",
       "AV-1764_MSK_LX_1250b_PM_1_MITO            quay.io/hisplan  \n",
       "AV-1764_Ru1250D_T_1_MITO                  quay.io/hisplan  \n",
       "AV-1765_Ru1250e_MITO                      quay.io/hisplan  \n",
       "AV-1766_MSK_LX_1250f_MITO                 quay.io/hisplan  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs = pd.DataFrame(inputs_all.loc[sample_names[0:]])\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c25489ef-ec76-4e19-b6d3-fafb53e5fd23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1759_Ru1083_MITO/FASTQ/3924_AV-1759_Ru1083_MITO_IGO_13388_1_S1_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1759_Ru1083_MITO/FASTQ/3924_AV-1759_Ru1083_MITO_IGO_13388_1_S1_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_MSK_LX_1083c_T_2_MITO/FASTQ/3925_AV-1760_MSK_LX_1083c_T_2_MITO_IGO_13388_2_S2_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_MSK_LX_1083c_T_2_MITO/FASTQ/3925_AV-1760_MSK_LX_1083c_T_2_MITO_IGO_13388_2_S2_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_Ru263_MITO/FASTQ/3933_AV-1760_Ru263_MITO_IGO_13388_7_S6_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_Ru263_MITO/FASTQ/3933_AV-1760_Ru263_MITO_IGO_13388_7_S6_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1761_POSIE_101920_T_1_MITO/FASTQ/3926_AV-1761_POSIE_101920_T_1_MITO_IGO_13388_3_S8_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1761_POSIE_101920_T_1_MITO/FASTQ/3926_AV-1761_POSIE_101920_T_1_MITO_IGO_13388_3_S8_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1762_Ru1083d_MITO/FASTQ/3927_AV-1762_Ru1083d_MITO_IGO_13388_4_S3_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1762_Ru1083d_MITO/FASTQ/3927_AV-1762_Ru1083d_MITO_IGO_13388_4_S3_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1763_Ru1250C_T_1_MITO/FASTQ/3928_AV-1763_Ru1250C_T_1_MITO_IGO_13388_5_S4_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1763_Ru1250C_T_1_MITO/FASTQ/3928_AV-1763_Ru1250C_T_1_MITO_IGO_13388_5_S4_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_MSK_LX_1250b_PM_1_MITO/FASTQ/3929_AV-1764_MSK_LX_1250b_PM_1_MITO_IGO_13388_6_S5_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_MSK_LX_1250b_PM_1_MITO/FASTQ/3929_AV-1764_MSK_LX_1250b_PM_1_MITO_IGO_13388_6_S5_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_Ru1250D_T_1_MITO/FASTQ/3930_AV-1764_Ru1250D_T_1_MITO_IGO_13388_8_S9_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_Ru1250D_T_1_MITO/FASTQ/3930_AV-1764_Ru1250D_T_1_MITO_IGO_13388_8_S9_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/FASTQ/3931_AV-1765_Ru1250e_MITO_IGO_13388_9_S10_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/FASTQ/3931_AV-1765_Ru1250e_MITO_IGO_13388_9_S10_L002_R1_001.fastq.gz']),\n",
       "       list(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1766_MSK_LX_1250f_MITO/FASTQ/3932_AV-1766_MSK_LX_1250f_MITO_IGO_13388_10_S7_L001_R1_001.fastq.gz', 's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1766_MSK_LX_1250f_MITO/FASTQ/3932_AV-1766_MSK_LX_1250f_MITO_IGO_13388_10_S7_L002_R1_001.fastq.gz'])],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs[f'{prefix}.fastqR1'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "248dbbb9-ee1d-4ae8-88e5-ecfecc7cbdd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1759_Ru1083_MITO/barcodes_in_RNA.non-epithelial.RU1083_LIV.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_MSK_LX_1083c_T_2_MITO/barcodes_in_RNA.non-epithelial.RU1083_T2.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_Ru263_MITO/barcodes_in_RNA.non-epithelial.RU263_PDX.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1761_POSIE_101920_T_1_MITO/barcodes_in_RNA.non-epithelial.RU1083_T1.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1762_Ru1083d_MITO/barcodes_in_RNA.non-epithelial.RU1083_ST.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1763_Ru1250C_T_1_MITO/barcodes_in_RNA.non-epithelial.RU1250_T1.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_MSK_LX_1250b_PM_1_MITO/barcodes_in_RNA.non-epithelial.RU1250_PL.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_Ru1250D_T_1_MITO/barcodes_in_RNA.non-epithelial.RU1250_T2.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/barcodes_in_RNA.non-epithelial.RU1250_ASC1.txt',\n",
       "       's3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1766_MSK_LX_1250f_MITO/barcodes_in_RNA.non-epithelial.RU1250_ASC2.txt'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs[f'{prefix}.whitelist'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0eafa67-d6cc-49c9-be93-50957248ded1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a9ee54c6-1a2c-46f5-a450-75bcf2a7cfb5",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'labels_all' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [33]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m labels \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\u001b[43mlabels_all\u001b[49m\u001b[38;5;241m.\u001b[39mloc[sample_names[\u001b[38;5;241m0\u001b[39m:]])\n\u001b[1;32m      2\u001b[0m labels\n",
      "\u001b[0;31mNameError\u001b[0m: name 'labels_all' is not defined"
     ]
    }
   ],
   "source": [
    "labels = pd.DataFrame(labels_all.loc[sample_names[0:]])\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9008381e-ad79-4dab-8585-a59d20756c73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "db50837c-9648-4400-af53-aa436bb93adc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MitoTracing.sampleName</th>\n",
       "      <th>MitoTracing.fastqName</th>\n",
       "      <th>MitoTracing.fastqR1</th>\n",
       "      <th>MitoTracing.fastqR2</th>\n",
       "      <th>MitoTracing.reference</th>\n",
       "      <th>MitoTracing.includeIntrons</th>\n",
       "      <th>MitoTracing.expectCells</th>\n",
       "      <th>MitoTracing.numCores</th>\n",
       "      <th>MitoTracing.memory</th>\n",
       "      <th>MitoTracing.whitelist</th>\n",
       "      <th>MitoTracing.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>epi_AV-1765_Ru1250e_MITO</td>\n",
       "      <td>3931_AV-1765_Ru1250e_MITO_IGO_13388_9</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>[s3://dp-lab-data/SCRI_Projects/HTA/...</td>\n",
       "      <td>{'cellRangerRefPkg': 's3://dp-lab-da...</td>\n",
       "      <td>False</td>\n",
       "      <td>5000</td>\n",
       "      <td>16</td>\n",
       "      <td>256</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        MitoTracing.sampleName  \\\n",
       "AV-1765_Ru1250e_MITO  epi_AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                                      MitoTracing.fastqName  \\\n",
       "AV-1765_Ru1250e_MITO  3931_AV-1765_Ru1250e_MITO_IGO_13388_9   \n",
       "\n",
       "                                          MitoTracing.fastqR1  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                          MitoTracing.fastqR2  \\\n",
       "AV-1765_Ru1250e_MITO  [s3://dp-lab-data/SCRI_Projects/HTA/...   \n",
       "\n",
       "                                        MitoTracing.reference  \\\n",
       "AV-1765_Ru1250e_MITO  {'cellRangerRefPkg': 's3://dp-lab-da...   \n",
       "\n",
       "                      MitoTracing.includeIntrons  MitoTracing.expectCells  \\\n",
       "AV-1765_Ru1250e_MITO                       False                     5000   \n",
       "\n",
       "                      MitoTracing.numCores  MitoTracing.memory  \\\n",
       "AV-1765_Ru1250e_MITO                    16                 256   \n",
       "\n",
       "                                        MitoTracing.whitelist  \\\n",
       "AV-1765_Ru1250e_MITO  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                     MitoTracing.dockerRegistry  \n",
       "AV-1765_Ru1250e_MITO            quay.io/hisplan  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b7dff356-8581-4638-a8e2-9e93dde1a251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AV-1765_Ru1250e_MITO</th>\n",
       "      <td>MitoTracing</td>\n",
       "      <td>Lung Tumor Atlas</td>\n",
       "      <td>AV-1765_Ru1250e_MITO</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/SCRI_Projects/HTA/M...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     pipelineType           project                sample  \\\n",
       "AV-1765_Ru1250e_MITO  MitoTracing  Lung Tumor Atlas  AV-1765_Ru1250e_MITO   \n",
       "\n",
       "                        owner                              destination  \\\n",
       "AV-1765_Ru1250e_MITO  sohailn  s3://dp-lab-data/SCRI_Projects/HTA/M...   \n",
       "\n",
       "                     transfer  comment  \n",
       "AV-1765_Ru1250e_MITO        -  sohailn  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bf7782ae-982a-45d9-bef1-5e6dede154cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/epi_mito-tracing-outs'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels['destination'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "cb6579bc-ee47-41fb-b4a7-66435d46188f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Setting cache to true to hopefully save some time\n",
    "path_to_options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3bd30985-9d8f-4a6a-afb4-0f93357ad25b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "918936fec41b4e9b94b952b0ef943203",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())\n",
    "\n",
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(inputs.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))\n",
    "        # Sometimes causes problems if too many samples are run at once\n",
    "        time.sleep(20) # Delay for 20 seconds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "25c94191-2218-4edc-b3cf-d705e0f36721",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sohailn/scing/bin/wdl-mito-tracing/configs/AV-1765_Ru1250e_MITO_MitoTracing.inputs.json\n",
      "/Users/sohailn/scing/bin/wdl-mito-tracing/configs/AV-1765_Ru1250e_MITO_MitoTracing.labels.json\n"
     ]
    }
   ],
   "source": [
    "print(path_to_inputs)\n",
    "print(path_to_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0a34cef8-6aef-46ea-a3ec-b5fe3e674213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/wdl-mito-tracing/submit-processing.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/AV-1765_Ru1250e_MITO_MitoTracing.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/configs/AV-1765_Ru1250e_MITO_MitoTracing.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/wdl-mito-tracing/MitoTracing.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"91bee57e-4aaa-40ce-9723-e2526cc16fbf\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f011b6-0131-4443-a67b-c4ab0aa0c8a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "569b813a-e476-42f7-a9be-889ede71997f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b514db5d-62b1-4ba3-b115-f6f33b23a736",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bfe3d434-fec4-4b4a-8987-b4c6d3c3eee2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "epi_Ru1083_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1759_Ru1083_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru1083_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1759_Ru1083_MITO/non_epi_mito-tracing-outs\n",
      "epi_MSK_LX_1083c_T_2_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_MSK_LX_1083c_T_2_MITO/epi_mito-tracing-outs\n",
      "non_epi_MSK_LX_1083c_T_2_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_MSK_LX_1083c_T_2_MITO/non_epi_mito-tracing-outs\n",
      "epi_POSIE_101920_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1761_POSIE_101920_T_1_MITO/epi_mito-tracing-outs\n",
      "non_epi_POSIE_101920_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1761_POSIE_101920_T_1_MITO/non_epi_mito-tracing-outs\n",
      "epi_Ru1083d_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1762_Ru1083d_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru1083d_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1762_Ru1083d_MITO/non_epi_mito-tracing-outs\n",
      "epi_Ru1250C_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1763_Ru1250C_T_1_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru1250C_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1763_Ru1250C_T_1_MITO/non_epi_mito-tracing-outs\n",
      "epi_MSK_LX_1250b_PM_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_MSK_LX_1250b_PM_1_MITO/epi_mito-tracing-outs\n",
      "non_epi_MSK_LX_1250b_PM_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_MSK_LX_1250b_PM_1_MITO/non_epi_mito-tracing-outs\n",
      "epi_Ru1250D_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_Ru1250D_T_1_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru1250D_T_1_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1764_Ru1250D_T_1_MITO/non_epi_mito-tracing-outs\n",
      "epi_Ru1250e_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru1250e_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1765_Ru1250e_MITO/non_epi_mito-tracing-outs\n",
      "epi_MSK_LX_1250f_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1766_MSK_LX_1250f_MITO/epi_mito-tracing-outs\n",
      "non_epi_MSK_LX_1250f_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1766_MSK_LX_1250f_MITO/non_epi_mito-tracing-outs\n",
      "epi_Ru263_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_Ru263_MITO/epi_mito-tracing-outs\n",
      "non_epi_Ru263_MITO, s3://dp-lab-data/SCRI_Projects/HTA/Mito_tracing/AV-1760_Ru263_MITO/non_epi_mito-tracing-outs\n"
     ]
    }
   ],
   "source": [
    "# Making sample.csv\n",
    "# Common query col: id, request_id, Sample\n",
    "sample_id = list(range(3924, 3934))\n",
    "samples = format_sample_aws(sample_id, 'id', creds)\n",
    "\n",
    "for sample in samples.index:\n",
    "    s3_path = samples[samples.index == sample]['S3_path'].item()\n",
    "    sample = sample[8:]\n",
    "    \n",
    "    print(f'epi_{sample}, {s3_path}/epi_{output_dirname}')\n",
    "    print(f'non_epi_{sample}, {s3_path}/non_epi_{output_dirname}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
