{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Define Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Priority of GEX data if multiple outputs are found in db\n",
    "sharp_wl_priority_map = {\n",
    "        m: [\"SEQC\", \"CR_GEX\"] for m in [\"Hashtag\", \"CiteSeq\"]\n",
    "    }\n",
    "# File patterns to search for in S3 for each accompanying pipeline\n",
    "sharp_wl_pattern_map = {\n",
    "    \"SEQC\": \"_dense.csv$\",\n",
    "    \"CR_GEX\": \"/filtered_feature_bc_matrix/barcodes.tsv.gz$\",\n",
    "    \"CR_ATAC\": \"/filtered_peak_bc_matrix/barcodes.tsv\"\n",
    "}\n",
    "sharp_wl_method_map = {\n",
    "    \"SEQC\": \"SeqcDenseCountsMatrixCsv\",\n",
    "    \"CR_GEX\": \"10x\",\n",
    "    \"CR_ATAC\": \"10x\",\n",
    "}\n",
    "# Names of FASTQ inputs in WDL; order is same as fastq_file_ids\n",
    "# TODO: Ask to change all inputs to \"fastq{file_id}\" or \"uriFastq{file_id}\"\n",
    "sharp_fastq_inputs_map = {\n",
    "    m: [\"uriFastqR1\", \"uriFastqR2\"] for m in [\"Hashtag\", \"CiteSeq\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'Hashtag': ['R1', 'R2'],\n",
    "    'CiteSeq': ['R1', 'R2'],\n",
    "    'AsapSeq': ['R1', 'R2', 'R3'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get s3 path of existing GEX analysis files\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_wl_dir(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_hashtag_lib = \"peer_lab_db.hashtag_lib\"\n",
    "        table_genome_index = \"peer_lab_db.genome_index\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_stats_data}.analysis_storage\n",
    "        FROM {table_sample_data}\n",
    "        LEFT JOIN {table_stats_data} \n",
    "        ON {table_stats_data}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_hashtag_lib}\n",
    "        ON {table_hashtag_lib}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_genome_index}\n",
    "        ON {table_genome_index}.id = {table_hashtag_lib}.genomeIndex_id\n",
    "        LEFT JOIN {table_sc_tech}\n",
    "        ON {table_sc_tech}.id = {table_genome_index}.scTech_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        if result: \n",
    "            return result\n",
    "        # As backup, get AWS storage location directly from sample_data\n",
    "        else:\n",
    "            query = f\"\"\"\n",
    "            SELECT AWS_storage\n",
    "            FROM {table_sample_data}\n",
    "            WHERE {table_sample_data}.id = {sample_id}\n",
    "            \"\"\"\n",
    "            result = execute_query(query, user, password)[0][0]\n",
    "            return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get white list method and associated file\n",
    "# Throws exception if no white list exists\n",
    "def get_wl_params(\n",
    "    sample_id: str,\n",
    "    user: str,\n",
    "    password: str,\n",
    "):\n",
    "    wl_params = dict()\n",
    "\n",
    "    wl_dir = get_wl_dir(sample_id, user, password)\n",
    "    wl_patterns = [sharp_wl_pattern_map[p] for p in sharp_wl_priority_map[prefix]]\n",
    "\n",
    "    try:\n",
    "        # Check white list file exists before loading info from database\n",
    "        assert wl_dir, f\"Empty analysis storage for sample id {sample_id}\"\n",
    "        _, bucket, key, _, _ = urllib.parse.urlsplit(wl_dir)\n",
    "        # White list file and method is first entry found on S3 \n",
    "        wl = pd.DataFrame(\n",
    "            [get_s3_objects(bucket, key.strip(\"/\"), re.compile(p)) for p in wl_patterns],\n",
    "            index = sharp_wl_priority_map[prefix],\n",
    "        ).dropna(how=\"all\")\n",
    "        try:\n",
    "            wl_key = wl.iloc[0,0] # if empty, missing white list file\n",
    "            wl_params[\"uri\"] = os.path.join(\"s3://\", bucket, wl_key)\n",
    "            wl_params[\"method\"] = sharp_wl_method_map[wl.index[0]]\n",
    "        except IndexError:\n",
    "            logging.error(\n",
    "                \"Path to barcodes or counts matrix of GEX data is missing!\"\n",
    "            )\n",
    "            return\n",
    "\n",
    "    except AssertionError:\n",
    "        logging.warning(f\"Path to GEX output results is missing for {sample_id}!\")\n",
    "        return\n",
    "\n",
    "    return wl_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bc_json_manual(\n",
    "    prefix,\n",
    "    platform,\n",
    "    user, \n",
    "    password,\n",
    "):\n",
    "    try:\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcodes\n",
    "        FROM {table_sc_tech}\n",
    "        WHERE {table_sc_tech}.sc_Tech = \"{platform.upper()}_{prefix}\"\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat barcodes from collaborator excel file\n",
    "def get_bcs_manual(\n",
    "    path_to_excel,\n",
    "):\n",
    "    # Replace unrecognized characters with text description\n",
    "    barcodes = pd.read_excel(path_to_excel) \n",
    "    replace = lambda x: x.encode('ascii', 'namereplace').decode().replace(\"\\\\N\", \"\")\n",
    "    barcodes[\"Description\"] = barcodes[\"Description\"].apply(replace)\n",
    "    barcodes[\"BP Shift\"] = 0\n",
    "    return barcodes[[\"Barcode\", \"DNA_ID\", \"Description\", \"BP Shift\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bc_params_manual(\n",
    "    barcodes,\n",
    "    prefix, \n",
    "    platform,\n",
    "    creds,\n",
    "):\n",
    "    bc_params = dict()\n",
    "    \n",
    "    # Add bp shift and sequence length based on conjugation\n",
    "    bp_shift_map = {\n",
    "        \"A\": 0,\n",
    "        \"B\": 10,\n",
    "        \"C\": 10,\n",
    "        \"M\": 1, # Methanol\n",
    "    }\n",
    "    conjugation = barcodes[\"DNA_ID\"].str.get(0)\n",
    "    if conjugation.nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample has multiple hashtag barcode categories and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else:\n",
    "        bc_params[\"conjugation\"] = conjugation.values[0]\n",
    "        bc_params[\"bp_shift\"] = bp_shift_map[bc_params[\"conjugation\"]]\n",
    "        bc_params[\"seq_length\"] = bc_params[\"bp_shift\"] + barcodes[\"Barcode\"].apply(len).max()\n",
    "    \n",
    "    # JSON of bc and UMI positions are stored in database\n",
    "    bc_json = get_bc_json_manual(prefix, platform, creds['user'], creds['password'])\n",
    "    bc_pos = json.loads(bc_json)\n",
    "    bc_params[\"cb\"] = bc_pos[\"cellbarcode\"]\n",
    "    bc_params[\"umi\"] = bc_params[\"cb\"] + bc_pos[\"UMIs\"]\n",
    "    \n",
    "    return bc_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_template(\n",
    "    prefix: str, \n",
    "    conjugation: str = None,\n",
    "):\n",
    "    if prefix==\"Hashtag\":\n",
    "        return {\n",
    "            \"A\": \"hashtag-10x-v3-tsa\",\n",
    "            \"B\": \"hashtag-10x-v3-tsb\",\n",
    "            \"C\": \"hashtag-10x-tsc\",\n",
    "        }[conjugation]\n",
    "    elif prefix==\"CiteSeq\":\n",
    "        return \"citeseq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Numpy encoder for JSON from pandas series\n",
    "class NpEncoder(json.JSONEncoder):\n",
    "    def default(self, obj):\n",
    "        if isinstance(obj, np.integer):\n",
    "            return int(obj)\n",
    "        elif isinstance(obj, np.floating):\n",
    "            return float(obj)\n",
    "        elif isinstance(obj, np.ndarray):\n",
    "            return obj.tolist()\n",
    "        else:\n",
    "            return super(NpEncoder, self).default(obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from SCRIdb\n",
    "def get_s3_objects(bucket, key, pattern, full_uri=False):\n",
    "    \n",
    "    s3r = boto3.resource(\"s3\")\n",
    "    bucket_s3 = s3r.Bucket(bucket)\n",
    "    objects = []\n",
    "    for obj in bucket_s3.objects.filter(Prefix=key):\n",
    "        hit = pattern.search(obj.key)\n",
    "        if hit:\n",
    "            objects.append(obj.key)\n",
    "    if full_uri:\n",
    "        objects = [f\"s3://{bucket}/{o}\" for o in objects]\n",
    "    return objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def execute_query(query, user, password):\n",
    "    with connect(\n",
    "        host=\"peer-lab-db.cggxmlwgzzpw.us-east-1.rds.amazonaws.com\",\n",
    "        database=\"peer_lab_db\",\n",
    "        user=user,\n",
    "        password=password,\n",
    "    ) as connection:\n",
    "        with connection.cursor(buffered=True) as cursor:\n",
    "            cursor.execute(query)\n",
    "            result = cursor.fetchall()\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_species(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_species = \"peer_lab_db.species\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_species}.Species\n",
    "        FROM {table_species}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_species}.id = {table_genome_idx}.species_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sc_tech(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sc_tech}.sc_Tech\n",
    "        FROM {table_sc_tech}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_sc_tech}.id = {table_genome_idx}.scTech_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sample_id(sample_name, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.Sample=\"{sample_name}\"\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_project_id(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_SEQC_version(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 cp {loc}/seqc-results/seqc_log.txt -\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        version = re.match(r\".*SEQC=v(\\d+\\.\\d+\\.\\d+).*\", out)[1]\n",
    "        return version\n",
    "    except:\n",
    "        return \"N/A\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_file_prefix(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 ls {loc}/seqc-results/\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        \n",
    "        # Note: I'm expecting the aligned bam file to be in loc\n",
    "        bam_pattern = re.compile(r\"(.*)_Aligned\\.out\\.bam$\")\n",
    "        filename = list(filter(bam_pattern.match, out.split()))[0]\n",
    "        file_prefix = re.match(bam_pattern, filename)[1]\n",
    "        return file_prefix\n",
    "    except:\n",
    "        raise ValueError(f\"BAM file not found in {loc}\")\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference(sample_id):\n",
    "    # Get species from database to decide reference\n",
    "    species = get_species(sample_id, creds[\"user\"], creds[\"password\"])\n",
    "    \n",
    "    # Map to reference locations\n",
    "    if \"Human\" in species:\n",
    "        return \"s3://seqc-public/genomes/hg38_long_polya/annotations.gtf\"\n",
    "    elif \"Mouse\" in species:\n",
    "        return \"s3://seqc-public/genomes/mm38_long_polya/annotations.gtf\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown Species: {species}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bc_whitelist(sample_id):\n",
    "    # Get version from database to decide whitelist\n",
    "    sc_tech = get_sc_tech(sample_id, creds[\"user\"], creds[\"password\"])\n",
    "    \n",
    "    # Map to reference locations\n",
    "    if \"V3\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v3/flat/3M-february-2018.txt\"\n",
    "    elif \"V2\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v2/flat/737K-august-2016.txt\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown Technology: {sc_tech}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(\n",
    "    workflow_path: str,\n",
    "    execp: str,\n",
    "    secrets: str,\n",
    "    inputs: str,\n",
    "    labels: str,\n",
    "    options: str,\n",
    "):\n",
    "    # change working directory to the pipeline package\n",
    "    oldwd = os.getcwd()\n",
    "    os.chdir(workflow_path)\n",
    "    \n",
    "    # execute the pipeline command\n",
    "    cmd = f\"{workflow_path}/{execp} -k {secrets} -i {inputs} -l {labels} -o {options}\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)\n",
    "    out = var.__dict__\n",
    "    \n",
    "    # change working directory back\n",
    "    os.chdir(oldwd)\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get bc and UMI positions from database stored in JSON format\n",
    "def get_bc_json(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_hashtag_lib = \"peer_lab_db.hashtag_lib\"\n",
    "        table_genome_index = \"peer_lab_db.genome_index\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcodes\n",
    "        FROM {table_sample_data}\n",
    "        LEFT JOIN {table_stats_data} \n",
    "        ON {table_stats_data}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_hashtag_lib}\n",
    "        ON {table_hashtag_lib}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_genome_index}\n",
    "        ON {table_genome_index}.id = {table_hashtag_lib}.genomeIndex_id\n",
    "        LEFT JOIN {table_sc_tech}\n",
    "        ON {table_sc_tech}.id = {table_genome_index}.scTech_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get bc sequence data from database\n",
    "def get_bcs(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_hashtag_barcodes = \"peer_lab_db.hashtag_barcodes\"\n",
    "        table_hashtags = \"peer_lab_db.hashtags\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcode_sequence, concat(substring(category, -1), barcode), \n",
    "        demultiplex_label, bp_shift FROM {table_hashtags} \n",
    "        LEFT JOIN {table_hashtag_barcodes} \n",
    "        ON {table_hashtag_barcodes}.id = {table_hashtags}.hashtagBarcodes_id \n",
    "        WHERE {table_hashtags}.sampleData_id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bc_params(\n",
    "    sample_id,\n",
    "    user,\n",
    "    password,\n",
    "):\n",
    "    bc_params = dict()\n",
    "\n",
    "    # JSON of bc and UMI positions are stored in database\n",
    "    # First check dense matrix exists before loading JSON from database\n",
    "    bc_json = get_bc_json(sample_id, user, password)\n",
    "    bc_pos = json.loads(bc_json)\n",
    "    bc_params[\"cb\"] = bc_pos[\"cellbarcode\"]\n",
    "    bc_params[\"umi\"] = bc_params[\"cb\"] + bc_pos[\"UMIs\"]\n",
    "\n",
    "    # Get bc sequence data from database\n",
    "    bcs = get_bcs(sample_id, user, password)\n",
    "    if not bcs:\n",
    "        logging.warning(f\"Barcodes data Empty:\\n\\t {db_connect.cur.statement}\")\n",
    "        return\n",
    "    for bc in bcs:\n",
    "        try:\n",
    "            assert bc[0], \"AssertionError: Missing sequence barcodes!\"\n",
    "            assert bc[1], \"AssertionError: Missing barcode IDs\"\n",
    "        except AssertionError as err:\n",
    "            logging.warning(f\"{err}:\\n\\t {db_connect.cur.statement}\")\n",
    "            return\n",
    "\n",
    "    barcodes = pd.DataFrame(bcs, columns=[\"sequence\", \"code\", \"label\", \"bp_shift\"])\n",
    "    conjugation = barcodes[\"code\"].str.get(0)\n",
    "    if conjugation.nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample has multiple hashtag barcode categories and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else:\n",
    "        bc_params[\"conjugation\"] = conjugation.values[0]\n",
    "\n",
    "    if barcodes[\"bp_shift\"].nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample {sample_id} has hashtag barcode categories, with bp-shift length/s \"\n",
    "            f\"{barcodes['bp_shift'].unique()}, and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else: \n",
    "        bc_params[\"bp_shift\"] = int(barcodes[\"bp_shift\"][0])\n",
    "        bc_params[\"seq_length\"] = bc_params[\"bp_shift\"] + barcodes[\"sequence\"].apply(len).max()\n",
    "\n",
    "    return bc_params"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"Hashtag\": \"Hashtag-results\",\n",
    "    \"CiteSeq\": \"CiteSeq-results\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"Hashtag\": \"submit-hashtag.sh\",\n",
    "    \"CiteSeq\": \"submit-citeseq.sh\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to pipeline name\n",
    "pipeline_types = {\n",
    "    \"Hashtag\": \"Hashtag\",\n",
    "    \"CiteSeq\": \"CITE-seq\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"CiteSeq\" # Workflow(s) to run; also .wdl filename prefix(es)\n",
    "pipeline_type = pipeline_types[prefix] # field in *.labels.json\n",
    "output_dirname = results_dirs[prefix]\n",
    "\n",
    "# If need to add comment, put here\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sharp-specific parameters\n",
    "options_prefix=\"Sharp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = f\"{Path.home()}/scing/bin/sharp-0.1.1\" # CHANGE THIS\n",
    "path_to_exec = sh_files[prefix]\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = f\"{workflow_dir}/{options_prefix}.options.aws.json\"\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS\n",
    "barcodes_path = f\"{Path.home()}/scing/data/barcodes\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Samples on which to run Sharp\n",
    "# Notes: \n",
    "# - Assumes data is transferred to AWS s3 (this should be an s3 location)\n",
    "# - Assumes directory name is also name of sample\n",
    "# - Workflows above will be run on all samples below\n",
    "# sample_paths = [\n",
    "#     \"s3://dp-lab-data/SCRI_Projects/HTAN_CITEseq/CI210127_CD45pos_citeseq_CITE\",\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: AWS_ACCESS_KEY_ID=ASIAYPY66CWMKHJQGBB4\n",
      "env: AWS_SECRET_ACCESS_KEY=0nbHN00aFVulHp4+YCSy0RvhgCGM727gYP3RGqSZ\n",
      "env: AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEGgaCXVzLWVhc3QtMSJIMEYCIQDxy4jpyfEB/jM9OAPA8MaqUXoBwb85rfdmM2jnCMsbLQIhAK1wUjV40LfkU8YVMww7NEXIwHxZ0EgOHYguj2eMFfJxKvgBCMD//////////wEQAxoMNTgzNjQzNTY3NTEyIgyXMJrVvZpPwzJN68EqzAEvkprlhoEsf/wIOuRrDT2dxOc8apURXE7FxyGu21YIeQ1uWx70qKA6oWYjMzLvihBS5hc7h+LeDbfKH6tnA93f2L/X5gzbDkFjGZAETOMhvbkwuco3Ly7120Maf7BzxFhI2icfYAaMUUTMcazjA/Pvg6nGsAckCvDAjJYHCYqlCtsWBZp4h/6qKdoCUttuW1zBzMYhMLkaehJHVn0XJRT4km41FSjc+tUMwP/n6qLg9Wm8201qRdjQo2Cn8mUHzLaJ4doyXKlibIflvWgwq+a0lwY6lwG2EC5bGDs0YEb0+XMzFdKcct+fiPnnqgxZd/i3lUhLxv1imL6GP8Hu3toPJgkazFpoPN5+iCBOTavcSTnlKWSzX6+TGtJta85foMoS7+N9qTwem3MJBKZoW8I5kLcmBtMcqig6x3x+NZOHox6xrG/ddwngg1sEc5tzkozICNL9RqzW8HH8ZJqqQ6BQySfbKFoj3u5F0EBm\n",
      "2021-10-07 15:31:32 agc-583643567512-us-east-1\n",
      "2021-10-07 15:28:07 cdktoolkit-stagingbucket-d49u1xfb0sc4\n",
      "2019-10-09 13:04:06 cf-templates-umiwbnq3566w-us-east-1\n",
      "2019-10-10 12:46:54 dp-daily-reports\n",
      "2017-09-26 10:15:15 dp-lab-data\n",
      "2019-05-23 12:34:58 dp-lab-data-public\n",
      "2018-03-01 13:29:07 dp-lab-glacier\n",
      "2020-11-06 23:02:27 dp-lab-gwf-core\n",
      "2021-07-15 18:14:01 dp-lab-gwf-core2\n",
      "2017-09-26 10:17:18 dp-lab-home\n",
      "2019-08-14 16:42:43 dp-lab-test\n",
      "2019-04-25 12:35:35 elasticbeanstalk-us-east-1-583643567512\n",
      "2019-12-13 15:53:57 scri-computational\n",
      "2017-09-26 10:23:50 seqc-public\n"
     ]
    }
   ],
   "source": [
    "%env AWS_ACCESS_KEY_ID=ASIAYPY66CWMKHJQGBB4\n",
    "%env AWS_SECRET_ACCESS_KEY=0nbHN00aFVulHp4+YCSy0RvhgCGM727gYP3RGqSZ\n",
    "%env AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEGgaCXVzLWVhc3QtMSJIMEYCIQDxy4jpyfEB/jM9OAPA8MaqUXoBwb85rfdmM2jnCMsbLQIhAK1wUjV40LfkU8YVMww7NEXIwHxZ0EgOHYguj2eMFfJxKvgBCMD//////////wEQAxoMNTgzNjQzNTY3NTEyIgyXMJrVvZpPwzJN68EqzAEvkprlhoEsf/wIOuRrDT2dxOc8apURXE7FxyGu21YIeQ1uWx70qKA6oWYjMzLvihBS5hc7h+LeDbfKH6tnA93f2L/X5gzbDkFjGZAETOMhvbkwuco3Ly7120Maf7BzxFhI2icfYAaMUUTMcazjA/Pvg6nGsAckCvDAjJYHCYqlCtsWBZp4h/6qKdoCUttuW1zBzMYhMLkaehJHVn0XJRT4km41FSjc+tUMwP/n6qLg9Wm8201qRdjQo2Cn8mUHzLaJ4doyXKlibIflvWgwq+a0lwY6lwG2EC5bGDs0YEb0+XMzFdKcct+fiPnnqgxZd/i3lUhLxv1imL6GP8Hu3toPJgkazFpoPN5+iCBOTavcSTnlKWSzX6+TGtJta85foMoS7+N9qTwem3MJBKZoW8I5kLcmBtMcqig6x3x+NZOHox6xrG/ddwngg1sEc5tzkozICNL9RqzW8HH8ZJqqQ6BQySfbKFoj3u5F0EBm\n",
    "!aws s3 ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Execution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Method 1: From SCRIdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "def get_sample_name(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Sample\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_names = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_names.append(result[0])\n",
    "        return sample_names\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "    \n",
    "    \n",
    "def get_aws_path(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_paths = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_paths.append(result[0])\n",
    "        return sample_paths\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "            \n",
    "def get_sample_id(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_ids = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_ids.append(result[0])\n",
    "        return sample_ids\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def format_sample_aws(querys, query_col, creds):\n",
    "    sample_names = []\n",
    "    sample_paths = []\n",
    "    sample_ids = []\n",
    "    \n",
    "    for query in querys:\n",
    "        sample_names += get_sample_name(query, query_col, creds)\n",
    "        sample_paths += get_aws_path(query, query_col, creds)\n",
    "        sample_ids += get_sample_id(query, query_col, creds)\n",
    "        \n",
    "    sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "    \n",
    "    samples = pd.DataFrame(\n",
    "        dict(S3_path=sample_paths, Sample_ID=sample_ids),\n",
    "        index=sample_names,\n",
    "        dtype=str,\n",
    "    )\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN2</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN3</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_path Sample_ID\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947\n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948\n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...      3949\n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...      3950\n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...      3951"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "sample_id = list(range(3947, 3952))\n",
    "\n",
    "samples = format_sample_aws(sample_id, 'id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = samples.rename(columns={'S3_path': 'S3_Path'})\n",
    "if prefix == 'Hashtag':\n",
    "    samples['S3_Path'] += '_HTO'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>FASTQs</th>\n",
       "      <th>Whitelist_Params</th>\n",
       "      <th>Barcode_Params</th>\n",
       "      <th>Barcodes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'cb': 16, 'umi': 28, 'conjugation':...</td>\n",
       "      <td>[(GTCAACTCTTTAGCG, C0251, Pt19_Tumor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'cb': 16, 'umi': 28, 'conjugation':...</td>\n",
       "      <td>[(GTCAACTCTTTAGCG, C0251, Pt60_LN, 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3949</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'cb': 16, 'umi': 28, 'conjugation':...</td>\n",
       "      <td>[(GTCAACTCTTTAGCG, C0251, Pt37_A, 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN2</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3950</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'cb': 16, 'umi': 28, 'conjugation':...</td>\n",
       "      <td>[(GTCAACTCTTTAGCG, C0251, Pt33_B, 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN3</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3951</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'cb': 16, 'umi': 28, 'conjugation':...</td>\n",
       "      <td>[(GTCAACTCTTTAGCG, C0251, Pt42_A, 10...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_Path Sample_ID  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948   \n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...      3949   \n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...      3950   \n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...      3951   \n",
       "\n",
       "                                                  FASTQs  \\\n",
       "SV-1721_SV_LN11  {'R1': ['s3://dp-lab-data/collaborat...   \n",
       "SV-1721_SV_LN12  {'R1': ['s3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN1   {'R1': ['s3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN2   {'R1': ['s3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN3   {'R1': ['s3://dp-lab-data/collaborat...   \n",
       "\n",
       "                                        Whitelist_Params  \\\n",
       "SV-1721_SV_LN11  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1721_SV_LN12  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN1   {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN2   {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1723_SV_LN3   {'uri': 's3://dp-lab-data/collaborat...   \n",
       "\n",
       "                                          Barcode_Params  \\\n",
       "SV-1721_SV_LN11  {'cb': 16, 'umi': 28, 'conjugation':...   \n",
       "SV-1721_SV_LN12  {'cb': 16, 'umi': 28, 'conjugation':...   \n",
       "SV-1723_SV_LN1   {'cb': 16, 'umi': 28, 'conjugation':...   \n",
       "SV-1723_SV_LN2   {'cb': 16, 'umi': 28, 'conjugation':...   \n",
       "SV-1723_SV_LN3   {'cb': 16, 'umi': 28, 'conjugation':...   \n",
       "\n",
       "                                                Barcodes  \n",
       "SV-1721_SV_LN11  [(GTCAACTCTTTAGCG, C0251, Pt19_Tumor...  \n",
       "SV-1721_SV_LN12  [(GTCAACTCTTTAGCG, C0251, Pt60_LN, 1...  \n",
       "SV-1723_SV_LN1   [(GTCAACTCTTTAGCG, C0251, Pt37_A, 10...  \n",
       "SV-1723_SV_LN2   [(GTCAACTCTTTAGCG, C0251, Pt33_B, 10...  \n",
       "SV-1723_SV_LN3   [(GTCAACTCTTTAGCG, C0251, Pt42_A, 10...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Get information for all samples\n",
    "# sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "# sample_names = [os.path.basename(s) for s in sample_paths]\n",
    "# sample_names = [re.match(r'(.*)_.+$', s)[1] for s in sample_names] # remove library suffix (e.g. _CITE, _HTO, etc.)\n",
    "# # TODO: assert basename is in peer_lab_db.sample_data.Sample\n",
    "# # assert(all(check_sample_name(s) for s in sample_names))\n",
    "# samples = pd.DataFrame(\n",
    "#     sample_paths,\n",
    "#     index=sample_names,\n",
    "#     columns=[\"S3_Path\"],\n",
    "#     dtype=str,\n",
    "# )\n",
    "# Get FASTQ paths from S3\n",
    "# Note: Uses same FASTQ file ids for all samples\n",
    "fastq_file_ids = fastq_map[prefix]\n",
    "samples[\"FASTQs\"] = samples[\"S3_Path\"].apply(lambda x: get_fastqs(x, fastq_file_ids))\n",
    "\n",
    "# samples[\"Sample_ID\"] = pd.Series(samples.index).apply(\n",
    "#     lambda x: get_sample_id(x, creds['user'], creds['password'])\n",
    "# ).values\n",
    "samples[\"Whitelist_Params\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_wl_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "samples[\"Barcode_Params\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_bc_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "samples[\"Barcodes\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_bcs(x, creds['user'], creds['password'])\n",
    ")\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fa6d1f8820a443d7b22bfecedb97f2e1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Function to reformat barcode labels for Sharp\n",
    "def reformat_bc_label(label):\n",
    "    label = label.encode('ascii', 'namereplace').decode()\n",
    "    label = label.replace(\"\\\\N\", \"\").replace(\" \", \"_\")\n",
    "    return label\n",
    "\n",
    "for name, sample in tqdm(samples.iterrows(), total=len(samples)):\n",
    "    \n",
    "    # Reformat barcodes\n",
    "    barcodes = pd.DataFrame(\n",
    "        sample[\"Barcodes\"],\n",
    "        columns=[\"sequence\", \"code\", \"label\", \"bp_shift\"]\n",
    "    )\n",
    "    barcodes[\"label\"] = barcodes[\"label\"].apply(reformat_bc_label)\n",
    "    \n",
    "    # Save to CSV\n",
    "    path_to_csv = f\"{barcodes_path}/{sample['Sample_ID']}_tag-list.csv\"\n",
    "    barcodes.to_csv(path_to_csv, header=False, index=False)\n",
    "    \n",
    "    # Upload tag-list to AWS\n",
    "    cmd = f\"aws s3 cp {path_to_csv} {sample['S3_Path']}/{output_dirname}/tag-list.csv\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "conjugation = samples[\"Barcode_Params\"].apply(lambda x: x[\"conjugation\"]).values[0] # conjugation must be same for all samples\n",
    "template_prefix = get_template(prefix, conjugation)\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "\n",
    "platform = \"10x\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.sampleName\"] = inputs.index\n",
    "inputs[f\"{prefix}.scRnaSeqPlatform\"] = platform # may need to change\n",
    "inputs[f\"{prefix}.lengthR1\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.lengthR2\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"seq_length\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhitelistUri\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"uri\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhiteListMethod\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"method\"])\n",
    "inputs[f\"{prefix}.{'tagList' if (prefix=='CiteSeq') else 'hashTagList'}\"] = \\\n",
    "    samples[\"S3_Path\"] + f\"/{output_dirname}/tag-list.csv\" #TODO: Ask about changing all to 'tagList'\n",
    "inputs[f\"{prefix}.cbStartPos\"] = 1\n",
    "inputs[f\"{prefix}.cbEndPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"cb\"])\n",
    "inputs[f\"{prefix}.umiStartPos\"] = inputs[f\"{prefix}.cbEndPos\"]+1\n",
    "inputs[f\"{prefix}.umiEndPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.trimPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"bp_shift\"])\n",
    "inputs[f\"{prefix}.translate10XBarcodes\"] = \\\n",
    "    (inputs[f\"{prefix}.scRnaSeqPlatform\"] == \"10x_v3\") & \\\n",
    "    (samples[\"Barcode_Params\"].apply(lambda x: x[\"conjugation\"] == \"B\"))\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "for file_id in fastq_file_ids: # Set FASTQs\n",
    "    inputs[f\"{prefix}.uriFastq{file_id}\"] = samples[\"FASTQs\"].apply(lambda x: x[file_id])\n",
    "\n",
    "# ********************\n",
    "# Defaults\n",
    "# Note: These may need to be changed on a per-sample or per-execution basis\n",
    "\n",
    "inputs[f\"{prefix}.slidingWindowSearch\"] = False\n",
    "inputs[f\"{prefix}.cbCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.umiCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.numExpectedCells\"] = 0\n",
    "# Need trick to set dictionary for each row\n",
    "common_resource_spec = {\n",
    "    \"cpu\": 32,\n",
    "    \"memory\": -1,\n",
    "}\n",
    "inputs[f\"{prefix}.resourceSpec\"] = inputs.iloc[:, 0].apply(lambda x: common_resource_spec)\n",
    "if prefix == \"Hashtag\":\n",
    "    inputs[f\"{prefix}.minCount\"] = 10\n",
    "\n",
    "# ********************\n",
    "\n",
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds[\"user\"], creds[\"password\"]))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_Path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hashtag.uriFastqR1</th>\n",
       "      <th>Hashtag.uriFastqR2</th>\n",
       "      <th>Hashtag.sampleName</th>\n",
       "      <th>Hashtag.scRnaSeqPlatform</th>\n",
       "      <th>Hashtag.lengthR1</th>\n",
       "      <th>Hashtag.lengthR2</th>\n",
       "      <th>Hashtag.cellBarcodeWhitelistUri</th>\n",
       "      <th>Hashtag.cellBarcodeWhiteListMethod</th>\n",
       "      <th>Hashtag.hashTagList</th>\n",
       "      <th>Hashtag.cbStartPos</th>\n",
       "      <th>...</th>\n",
       "      <th>Hashtag.umiEndPos</th>\n",
       "      <th>Hashtag.trimPos</th>\n",
       "      <th>Hashtag.slidingWindowSearch</th>\n",
       "      <th>Hashtag.translate10XBarcodes</th>\n",
       "      <th>Hashtag.cbCollapsingDistance</th>\n",
       "      <th>Hashtag.umiCollapsingDistance</th>\n",
       "      <th>Hashtag.numExpectedCells</th>\n",
       "      <th>Hashtag.minCount</th>\n",
       "      <th>Hashtag.resourceSpec</th>\n",
       "      <th>Hashtag.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1721_SV_LN11</td>\n",
       "      <td>10x</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1721_SV_LN12</td>\n",
       "      <td>10x</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN1</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1723_SV_LN1</td>\n",
       "      <td>10x</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN2</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1723_SV_LN2</td>\n",
       "      <td>10x</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN3</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1723_SV_LN3</td>\n",
       "      <td>10x</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Hashtag.uriFastqR1  \\\n",
       "SV-1721_SV_LN11  [s3://dp-lab-data/collaborators/vard...   \n",
       "SV-1721_SV_LN12  [s3://dp-lab-data/collaborators/vard...   \n",
       "SV-1723_SV_LN1   [s3://dp-lab-data/collaborators/vard...   \n",
       "SV-1723_SV_LN2   [s3://dp-lab-data/collaborators/vard...   \n",
       "SV-1723_SV_LN3   [s3://dp-lab-data/collaborators/vard...   \n",
       "\n",
       "                                      Hashtag.uriFastqR2 Hashtag.sampleName  \\\n",
       "SV-1721_SV_LN11  [s3://dp-lab-data/collaborators/vard...    SV-1721_SV_LN11   \n",
       "SV-1721_SV_LN12  [s3://dp-lab-data/collaborators/vard...    SV-1721_SV_LN12   \n",
       "SV-1723_SV_LN1   [s3://dp-lab-data/collaborators/vard...     SV-1723_SV_LN1   \n",
       "SV-1723_SV_LN2   [s3://dp-lab-data/collaborators/vard...     SV-1723_SV_LN2   \n",
       "SV-1723_SV_LN3   [s3://dp-lab-data/collaborators/vard...     SV-1723_SV_LN3   \n",
       "\n",
       "                Hashtag.scRnaSeqPlatform  Hashtag.lengthR1  Hashtag.lengthR2  \\\n",
       "SV-1721_SV_LN11                      10x                28                25   \n",
       "SV-1721_SV_LN12                      10x                28                25   \n",
       "SV-1723_SV_LN1                       10x                28                25   \n",
       "SV-1723_SV_LN2                       10x                28                25   \n",
       "SV-1723_SV_LN3                       10x                28                25   \n",
       "\n",
       "                         Hashtag.cellBarcodeWhitelistUri  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...   \n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...   \n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...   \n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...   \n",
       "\n",
       "                Hashtag.cellBarcodeWhiteListMethod  \\\n",
       "SV-1721_SV_LN11                                10x   \n",
       "SV-1721_SV_LN12                                10x   \n",
       "SV-1723_SV_LN1                                 10x   \n",
       "SV-1723_SV_LN2                                 10x   \n",
       "SV-1723_SV_LN3                                 10x   \n",
       "\n",
       "                                     Hashtag.hashTagList  Hashtag.cbStartPos  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "\n",
       "                 ...  Hashtag.umiEndPos  Hashtag.trimPos  \\\n",
       "SV-1721_SV_LN11  ...                 28               10   \n",
       "SV-1721_SV_LN12  ...                 28               10   \n",
       "SV-1723_SV_LN1   ...                 28               10   \n",
       "SV-1723_SV_LN2   ...                 28               10   \n",
       "SV-1723_SV_LN3   ...                 28               10   \n",
       "\n",
       "                 Hashtag.slidingWindowSearch  Hashtag.translate10XBarcodes  \\\n",
       "SV-1721_SV_LN11                        False                         False   \n",
       "SV-1721_SV_LN12                        False                         False   \n",
       "SV-1723_SV_LN1                         False                         False   \n",
       "SV-1723_SV_LN2                         False                         False   \n",
       "SV-1723_SV_LN3                         False                         False   \n",
       "\n",
       "                 Hashtag.cbCollapsingDistance  Hashtag.umiCollapsingDistance  \\\n",
       "SV-1721_SV_LN11                             1                              1   \n",
       "SV-1721_SV_LN12                             1                              1   \n",
       "SV-1723_SV_LN1                              1                              1   \n",
       "SV-1723_SV_LN2                              1                              1   \n",
       "SV-1723_SV_LN3                              1                              1   \n",
       "\n",
       "                 Hashtag.numExpectedCells  Hashtag.minCount  \\\n",
       "SV-1721_SV_LN11                         0                10   \n",
       "SV-1721_SV_LN12                         0                10   \n",
       "SV-1723_SV_LN1                          0                10   \n",
       "SV-1723_SV_LN2                          0                10   \n",
       "SV-1723_SV_LN3                          0                10   \n",
       "\n",
       "                      Hashtag.resourceSpec  Hashtag.dockerRegistry  \n",
       "SV-1721_SV_LN11  {'cpu': 32, 'memory': -1}         quay.io/hisplan  \n",
       "SV-1721_SV_LN12  {'cpu': 32, 'memory': -1}         quay.io/hisplan  \n",
       "SV-1723_SV_LN1   {'cpu': 32, 'memory': -1}         quay.io/hisplan  \n",
       "SV-1723_SV_LN2   {'cpu': 32, 'memory': -1}         quay.io/hisplan  \n",
       "SV-1723_SV_LN3   {'cpu': 32, 'memory': -1}         quay.io/hisplan  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>Locally advanced gastric cancer</td>\n",
       "      <td>SV-1721_SV_LN11</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>Locally advanced gastric cancer</td>\n",
       "      <td>SV-1721_SV_LN12</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN1</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>single cell immune profiling of PBMC...</td>\n",
       "      <td>SV-1723_SV_LN1</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN2</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>single cell immune profiling of PBMC...</td>\n",
       "      <td>SV-1723_SV_LN2</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN3</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>single cell immune profiling of PBMC...</td>\n",
       "      <td>SV-1723_SV_LN3</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pipelineType                                  project  \\\n",
       "SV-1721_SV_LN11      Hashtag          Locally advanced gastric cancer   \n",
       "SV-1721_SV_LN12      Hashtag          Locally advanced gastric cancer   \n",
       "SV-1723_SV_LN1       Hashtag  single cell immune profiling of PBMC...   \n",
       "SV-1723_SV_LN2       Hashtag  single cell immune profiling of PBMC...   \n",
       "SV-1723_SV_LN3       Hashtag  single cell immune profiling of PBMC...   \n",
       "\n",
       "                          sample    owner  \\\n",
       "SV-1721_SV_LN11  SV-1721_SV_LN11  sohailn   \n",
       "SV-1721_SV_LN12  SV-1721_SV_LN12  sohailn   \n",
       "SV-1723_SV_LN1    SV-1723_SV_LN1  sohailn   \n",
       "SV-1723_SV_LN2    SV-1723_SV_LN2  sohailn   \n",
       "SV-1723_SV_LN3    SV-1723_SV_LN3  sohailn   \n",
       "\n",
       "                                             destination transfer  comment  \n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...        -  sohailn  \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...        -  sohailn  \n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...        -  sohailn  \n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...        -  sohailn  \n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...        -  sohailn  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/Hashtag-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN12_HTO/Hashtag-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN1_HTO/Hashtag-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN2_HTO/Hashtag-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN3_HTO/Hashtag-results']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels['destination'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c623c42539d348d5a97720b9a4d7315f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "import time\n",
    "\n",
    "with tqdm(samples.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))\n",
    "        time.sleep(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-hashtag.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1721_SV_LN11_Hashtag.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1721_SV_LN11_Hashtag.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"72d6f6f0-b149-47f8-9017-cc0ee3ff98d7\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-hashtag.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1721_SV_LN12_Hashtag.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1721_SV_LN12_Hashtag.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"8c2f2354-c7a7-44aa-a776-2f315ca4f297\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-hashtag.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN1_Hashtag.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN1_Hashtag.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"a278b116-d2b9-462d-b24b-8ef490141f58\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-hashtag.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN2_Hashtag.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN2_Hashtag.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"4a1d348b-78b5-41b5-a4df-40c9fc19ff63\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-hashtag.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN3_Hashtag.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN3_Hashtag.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"768c8842-9f79-4309-b2e4-7f5751ecc0fc\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "### Method 2: From Excel File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual inputs\n",
    "platform = \"10x_v3\"\n",
    "#path_to_excel = f\"{Path.home()}/scing/barcodes/JR-1217_3080_Barcodes.xlsx\" \n",
    "# path_to_excel = f\"{Path.home()}/scing/barcodes/3213_IM-1356.xlsx\"\n",
    "path_to_excel = f'~/Downloads/SV-1721 CITEseq.xlsx'\n",
    "# path_to_excel = f'~/Downloads/SV-1723 CITEseq.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN1</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN2</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1723_SV_LN3</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_path Sample_ID\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947\n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948\n",
       "SV-1723_SV_LN1   s3://dp-lab-data/collaborators/vardh...      3949\n",
       "SV-1723_SV_LN2   s3://dp-lab-data/collaborators/vardh...      3950\n",
       "SV-1723_SV_LN3   s3://dp-lab-data/collaborators/vardh...      3951"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "sample_id = list(range(3947, 3952))\n",
    "\n",
    "samples = format_sample_aws(sample_id, 'id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = samples.rename(columns={'S3_path': 'S3_Path'})\n",
    "samples['S3_Path'] += '_HTO'\n",
    "# if prefix == 'Hashtag':\n",
    "#     samples['S3_Path'] += '_HTO'\n",
    "    \n",
    "# elif prefix == 'CiteSEq':\n",
    "#     samples['S3_Path'] += '_CITE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_Path Sample_ID\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947\n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = samples.loc[samples.index.str.startswith('SV-1721')]\n",
    "# samples = samples.loc[samples.index.str.startswith('SV-1723')]\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get information for all samples\n",
    "# # sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "# # sample_names = [os.path.basename(s) for s in sample_paths]\n",
    "# # sample_names = [re.match(r'(.*)_.+$', s)[1] for s in sample_names] # remove library suffix (e.g. _CITE, _HTO, etc.)\n",
    "# # TODO: assert basename is in peer_lab_db.sample_data.Sample\n",
    "# # assert(all(check_sample_name(s) for s in sample_names))\n",
    "# samples = pd.DataFrame(\n",
    "#     sample_paths,\n",
    "#     index=sample_names,\n",
    "#     columns=[\"S3_Path\"],\n",
    "#     dtype=str,\n",
    "# )\n",
    "# samples[\"Sample_ID\"] = pd.Series(samples.index).apply(\n",
    "#     lambda x: get_sample_id(x, creds['user'], creds['password'])\n",
    "# ).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read barcodes from file\n",
    "# Note: Must be subset to HTO or CITE barcodes before next step!\n",
    "# barcodes = get_bcs_manual(path_to_excel)\n",
    "# barcodes = barcodes[\n",
    "#     barcodes[\"Description\"].str.contains(\"SS1\") |\n",
    "#     barcodes[\"Description\"].str.contains(\"SS2\")\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~/Downloads/SV-1721 CITEseq.xlsx\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>DNA_ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>BP Shift</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTCTTTGTCAGTGCA</td>\n",
       "      <td>C0006</td>\n",
       "      <td>anti-human CD86</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTTGTCCGACAATAC</td>\n",
       "      <td>C0007</td>\n",
       "      <td>anti-human CD274 (B7-H1, PD-L1)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TGATAGAAACAGACC</td>\n",
       "      <td>C0020</td>\n",
       "      <td>anti-human CD270 (HVEM, TR2)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATCACATCGTTGCCA</td>\n",
       "      <td>C0023</td>\n",
       "      <td>anti-human CD155 (PVR)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACCTTCCGTCTAAG</td>\n",
       "      <td>C0024</td>\n",
       "      <td>anti-human CD112 (Nectin-2)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>GAGTCGAGAAATCAT</td>\n",
       "      <td>C0918</td>\n",
       "      <td>anti-human HLA-E</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>TCCCACTTCCGCTTT</td>\n",
       "      <td>C0920</td>\n",
       "      <td>anti-human CD82</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CTACTTCCCTGTCAA</td>\n",
       "      <td>C0944</td>\n",
       "      <td>anti-human CD101 (BB27)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>GCCGCATGAGAAACA</td>\n",
       "      <td>C1046</td>\n",
       "      <td>anti-human CD88 (C5aR)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CTGATGAGATGTCAG</td>\n",
       "      <td>C1052</td>\n",
       "      <td>anti-human CD224</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Barcode DNA_ID                      Description  BP Shift\n",
       "0    GTCTTTGTCAGTGCA  C0006                  anti-human CD86         0\n",
       "1    GTTGTCCGACAATAC  C0007  anti-human CD274 (B7-H1, PD-L1)         0\n",
       "2    TGATAGAAACAGACC  C0020     anti-human CD270 (HVEM, TR2)         0\n",
       "3    ATCACATCGTTGCCA  C0023           anti-human CD155 (PVR)         0\n",
       "4    AACCTTCCGTCTAAG  C0024      anti-human CD112 (Nectin-2)         0\n",
       "..               ...    ...                              ...       ...\n",
       "132  GAGTCGAGAAATCAT  C0918                 anti-human HLA-E         0\n",
       "133  TCCCACTTCCGCTTT  C0920                  anti-human CD82         0\n",
       "134  CTACTTCCCTGTCAA  C0944          anti-human CD101 (BB27)         0\n",
       "135  GCCGCATGAGAAACA  C1046           anti-human CD88 (C5aR)         0\n",
       "136  CTGATGAGATGTCAG  C1052                 anti-human CD224         0\n",
       "\n",
       "[137 rows x 4 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes = pd.read_excel(path_to_excel, header=None) \n",
    "print(path_to_excel)\n",
    "barcodes.columns = ['Description', 'Barcode', 'DNA_ID', 'description', 'citeseq']\n",
    "replace = lambda x: x.encode('ascii', 'namereplace').decode().replace(\"\\\\N\", \"\")\n",
    "barcodes[\"Description\"] = barcodes[\"Description\"].apply(replace)\n",
    "barcodes[\"BP Shift\"] = 0\n",
    "barcodes = barcodes[[\"Barcode\", \"DNA_ID\", \"Description\", \"BP Shift\"]]\n",
    "barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39230bb1a70941beb108a2e3ee0d331e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aws s3 cp ~/Downloads/SV-1721_CITEseq.csv s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/CiteSeq-results/tag-list.csv\n",
      "aws s3 cp ~/Downloads/SV-1721_CITEseq.csv s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN12_HTO/CiteSeq-results/tag-list.csv\n"
     ]
    }
   ],
   "source": [
    "# Upload barcodes for all samples\n",
    "# Note: Uploads same barcodes to s3 for all samples\n",
    "path_to_csv = path_to_excel.replace(\".xlsx\", \".csv\")\n",
    "path_to_csv = path_to_csv.replace(' ', '_')\n",
    "barcodes.to_csv(path_to_csv, header=False, index=False)\n",
    "for s3_path in tqdm(samples[\"S3_Path\"]):\n",
    "    cmd = f\"aws s3 cp {path_to_csv} {s3_path}/{output_dirname}/tag-list.csv\"\n",
    "    print(cmd)\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Whitelist_Params</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_Path Sample_ID  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948   \n",
       "\n",
       "                                        Whitelist_Params  \n",
       "SV-1721_SV_LN11  {'uri': 's3://dp-lab-data/collaborat...  \n",
       "SV-1721_SV_LN12  {'uri': 's3://dp-lab-data/collaborat...  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note: Assumes GEX data is recorded in database\n",
    "samples[\"Whitelist_Params\"] = samples[\"Sample_ID\"].apply(\n",
    "    lambda x: get_wl_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "assert ~samples[\"Whitelist_Params\"].isna().any()\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Whitelist_Params</th>\n",
       "      <th>BC_Params</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'conjugation': 'C', 'bp_shift': 10,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'conjugation': 'C', 'bp_shift': 10,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_Path Sample_ID  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948   \n",
       "\n",
       "                                        Whitelist_Params  \\\n",
       "SV-1721_SV_LN11  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1721_SV_LN12  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "\n",
       "                                               BC_Params  \n",
       "SV-1721_SV_LN11  {'conjugation': 'C', 'bp_shift': 10,...  \n",
       "SV-1721_SV_LN12  {'conjugation': 'C', 'bp_shift': 10,...  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note: Adds same barcode parameters for all samples\n",
    "samples[\"BC_Params\"] = samples[\"Sample_ID\"].apply(\n",
    "    lambda x: get_bc_params_manual(barcodes, prefix, platform, creds)\n",
    ")\n",
    "assert ~samples[\"BC_Params\"].isna().any()\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Whitelist_Params</th>\n",
       "      <th>BC_Params</th>\n",
       "      <th>FASTQs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3947</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'conjugation': 'C', 'bp_shift': 10,...</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>3948</td>\n",
       "      <td>{'uri': 's3://dp-lab-data/collaborat...</td>\n",
       "      <td>{'conjugation': 'C', 'bp_shift': 10,...</td>\n",
       "      <td>{'R1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 S3_Path Sample_ID  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...      3947   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...      3948   \n",
       "\n",
       "                                        Whitelist_Params  \\\n",
       "SV-1721_SV_LN11  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "SV-1721_SV_LN12  {'uri': 's3://dp-lab-data/collaborat...   \n",
       "\n",
       "                                               BC_Params  \\\n",
       "SV-1721_SV_LN11  {'conjugation': 'C', 'bp_shift': 10,...   \n",
       "SV-1721_SV_LN12  {'conjugation': 'C', 'bp_shift': 10,...   \n",
       "\n",
       "                                                  FASTQs  \n",
       "SV-1721_SV_LN11  {'R1': ['s3://dp-lab-data/collaborat...  \n",
       "SV-1721_SV_LN12  {'R1': ['s3://dp-lab-data/collaborat...  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get FASTQ paths from S3\n",
    "# Note: Uses same FASTQ file ids for all samples\n",
    "fastq_file_ids = fastq_map[prefix]\n",
    "samples[\"FASTQs\"] = samples[\"S3_Path\"].apply(lambda x: get_fastqs(x, fastq_file_ids))\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'R1': ['s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L001_R1_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L002_R1_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L003_R1_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L004_R1_001.fastq.gz'],\n",
       " 'R2': ['s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L001_R2_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L002_R2_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L003_R2_001.fastq.gz',\n",
       "  's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/FASTQ/3947_SV-1721_SV_LN11_HTO_IGO_12437_AS_7_S25_L004_R2_001.fastq.gz']}"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.iloc[0]['FASTQs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "conjugation = samples[\"BC_Params\"].apply(lambda x: x[\"conjugation\"]).values[0] # conjugation must be same for all samples\n",
    "template_prefix = get_template(prefix, conjugation)\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.sampleName\"] = inputs.index\n",
    "inputs[f\"{prefix}.scRnaSeqPlatform\"] = platform # may need to change\n",
    "inputs[f\"{prefix}.lengthR1\"] = samples[\"BC_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.lengthR2\"] = samples[\"BC_Params\"].apply(lambda x: x[\"seq_length\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhitelistUri\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"uri\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhiteListMethod\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"method\"])\n",
    "inputs[f\"{prefix}.{'tagList' if (prefix=='CiteSeq') else 'hashTagList'}\"] = \\\n",
    "    samples[\"S3_Path\"] + f\"/{output_dirname}/tag-list.csv\" #TODO: Ask about changing all to 'tagList'\n",
    "inputs[f\"{prefix}.cbStartPos\"] = 1\n",
    "inputs[f\"{prefix}.cbEndPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"cb\"])\n",
    "inputs[f\"{prefix}.umiStartPos\"] = inputs[f\"{prefix}.cbEndPos\"]+1\n",
    "inputs[f\"{prefix}.umiEndPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.trimPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"bp_shift\"])\n",
    "inputs[f\"{prefix}.translate10XBarcodes\"] = \\\n",
    "    (inputs[f\"{prefix}.scRnaSeqPlatform\"] == \"10x_v3\") & \\\n",
    "    (samples[\"BC_Params\"].apply(lambda x: x[\"conjugation\"] == \"B\"))\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "for file_id in fastq_file_ids: # Set FASTQs\n",
    "    inputs[f\"{prefix}.uriFastq{file_id}\"] = samples[\"FASTQs\"].apply(lambda x: x[file_id])\n",
    "\n",
    "# ********************\n",
    "# Defaults\n",
    "# Note: These may need to be changed on a per-sample or per-execution basis\n",
    "\n",
    "inputs[f\"{prefix}.slidingWindowSearch\"] = False\n",
    "inputs[f\"{prefix}.cbCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.umiCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.numExpectedCells\"] = 0\n",
    "# Need trick to set dictionary for each row\n",
    "common_resource_spec = {\n",
    "    \"cpu\": 32,\n",
    "    \"memory\": -1,\n",
    "}\n",
    "inputs[f\"{prefix}.resourceSpec\"] = inputs.iloc[:, 0].apply(lambda x: common_resource_spec)\n",
    "if prefix == \"Hashtag\":\n",
    "    inputs[f\"{prefix}.minCount\"] = 10\n",
    "\n",
    "# ********************\n",
    "\n",
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds[\"user\"], creds[\"password\"]))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_Path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CiteSeq.uriFastqR1</th>\n",
       "      <th>CiteSeq.uriFastqR2</th>\n",
       "      <th>CiteSeq.sampleName</th>\n",
       "      <th>CiteSeq.scRnaSeqPlatform</th>\n",
       "      <th>CiteSeq.lengthR1</th>\n",
       "      <th>CiteSeq.lengthR2</th>\n",
       "      <th>CiteSeq.cellBarcodeWhitelistUri</th>\n",
       "      <th>CiteSeq.cellBarcodeWhiteListMethod</th>\n",
       "      <th>CiteSeq.tagList</th>\n",
       "      <th>CiteSeq.cbStartPos</th>\n",
       "      <th>...</th>\n",
       "      <th>CiteSeq.umiStartPos</th>\n",
       "      <th>CiteSeq.umiEndPos</th>\n",
       "      <th>CiteSeq.trimPos</th>\n",
       "      <th>CiteSeq.slidingWindowSearch</th>\n",
       "      <th>CiteSeq.translate10XBarcodes</th>\n",
       "      <th>CiteSeq.cbCollapsingDistance</th>\n",
       "      <th>CiteSeq.umiCollapsingDistance</th>\n",
       "      <th>CiteSeq.numExpectedCells</th>\n",
       "      <th>CiteSeq.resourceSpec</th>\n",
       "      <th>CiteSeq.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1721_SV_LN11</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/vard...</td>\n",
       "      <td>SV-1721_SV_LN12</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>10x</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      CiteSeq.uriFastqR1  \\\n",
       "SV-1721_SV_LN11  [s3://dp-lab-data/collaborators/vard...   \n",
       "SV-1721_SV_LN12  [s3://dp-lab-data/collaborators/vard...   \n",
       "\n",
       "                                      CiteSeq.uriFastqR2 CiteSeq.sampleName  \\\n",
       "SV-1721_SV_LN11  [s3://dp-lab-data/collaborators/vard...    SV-1721_SV_LN11   \n",
       "SV-1721_SV_LN12  [s3://dp-lab-data/collaborators/vard...    SV-1721_SV_LN12   \n",
       "\n",
       "                CiteSeq.scRnaSeqPlatform  CiteSeq.lengthR1  CiteSeq.lengthR2  \\\n",
       "SV-1721_SV_LN11                   10x_v3                28                25   \n",
       "SV-1721_SV_LN12                   10x_v3                28                25   \n",
       "\n",
       "                         CiteSeq.cellBarcodeWhitelistUri  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...   \n",
       "\n",
       "                CiteSeq.cellBarcodeWhiteListMethod  \\\n",
       "SV-1721_SV_LN11                                10x   \n",
       "SV-1721_SV_LN12                                10x   \n",
       "\n",
       "                                         CiteSeq.tagList  CiteSeq.cbStartPos  \\\n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...                   1   \n",
       "\n",
       "                 ...  CiteSeq.umiStartPos  CiteSeq.umiEndPos  CiteSeq.trimPos  \\\n",
       "SV-1721_SV_LN11  ...                   17                 28               10   \n",
       "SV-1721_SV_LN12  ...                   17                 28               10   \n",
       "\n",
       "                 CiteSeq.slidingWindowSearch  CiteSeq.translate10XBarcodes  \\\n",
       "SV-1721_SV_LN11                        False                         False   \n",
       "SV-1721_SV_LN12                        False                         False   \n",
       "\n",
       "                 CiteSeq.cbCollapsingDistance  CiteSeq.umiCollapsingDistance  \\\n",
       "SV-1721_SV_LN11                             1                              1   \n",
       "SV-1721_SV_LN12                             1                              1   \n",
       "\n",
       "                 CiteSeq.numExpectedCells       CiteSeq.resourceSpec  \\\n",
       "SV-1721_SV_LN11                         0  {'cpu': 32, 'memory': -1}   \n",
       "SV-1721_SV_LN12                         0  {'cpu': 32, 'memory': -1}   \n",
       "\n",
       "                CiteSeq.dockerRegistry  \n",
       "SV-1721_SV_LN11        quay.io/hisplan  \n",
       "SV-1721_SV_LN12        quay.io/hisplan  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN11</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>Locally advanced gastric cancer</td>\n",
       "      <td>SV-1721_SV_LN11</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SV-1721_SV_LN12</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>Locally advanced gastric cancer</td>\n",
       "      <td>SV-1721_SV_LN12</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/vardh...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pipelineType                          project  \\\n",
       "SV-1721_SV_LN11     CITE-seq  Locally advanced gastric cancer   \n",
       "SV-1721_SV_LN12     CITE-seq  Locally advanced gastric cancer   \n",
       "\n",
       "                          sample    owner  \\\n",
       "SV-1721_SV_LN11  SV-1721_SV_LN11  sohailn   \n",
       "SV-1721_SV_LN12  SV-1721_SV_LN12  sohailn   \n",
       "\n",
       "                                             destination transfer  comment  \n",
       "SV-1721_SV_LN11  s3://dp-lab-data/collaborators/vardh...        -  sohailn  \n",
       "SV-1721_SV_LN12  s3://dp-lab-data/collaborators/vardh...        -  sohailn  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f2653ca7da3a4aa8b6d437310e123962",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import time\n",
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(samples.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))\n",
    "            \n",
    "        time.sleep(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN1_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN1_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"9f17b923-4124-422c-ab30-35c1d30e3261\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN2_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN2_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"980049ce-7178-4e37-97ab-2dd2ef54a844\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/sharp-0.1.1/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN3_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/configs/SV-1723_SV_LN3_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/sharp-0.1.1/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"0ee16f6b-dff7-4758-affe-551a6103540c\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN1_HTO/CiteSeq-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN2_HTO/CiteSeq-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/scImmuneProfilingPbmcsNAcTreatedCovidPts/SV-1723_SV_LN3_HTO/CiteSeq-results']"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels['destination'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN11_HTO/CiteSeq-results',\n",
       " 's3://dp-lab-data/collaborators/vardhans/LocallyAdvancedGastricCancer/SV-1721_SV_LN12_HTO/CiteSeq-results']"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels['destination'].tolist()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
