{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# Define Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Priority of GEX data if multiple outputs are found in db\n",
    "sharp_wl_priority_map = {\n",
    "        m: [\"SEQC\", \"CR_GEX\"] for m in [\"Hashtag\", \"CiteSeq\"]\n",
    "    }\n",
    "# File patterns to search for in S3 for each accompanying pipeline\n",
    "sharp_wl_pattern_map = {\n",
    "    \"SEQC\": \"_dense.csv$\",\n",
    "    \"CR_GEX\": \"/filtered_feature_bc_matrix/barcodes.tsv.gz$\",\n",
    "    \"CR_ATAC\": \"/filtered_peak_bc_matrix/barcodes.tsv$\"\n",
    "}\n",
    "sharp_wl_method_map = {\n",
    "    \"SEQC\": \"SeqcDenseCountsMatrixCsv\",\n",
    "    \"CR_GEX\": \"10x\",\n",
    "    \"CR_ATAC\": \"10x\",\n",
    "}\n",
    "# Names of FASTQ inputs in WDL; order is same as fastq_file_ids\n",
    "# TODO: Ask to change all inputs to \"fastq{file_id}\" or \"uriFastq{file_id}\"\n",
    "sharp_fastq_inputs_map = {\n",
    "    m: [\"uriFastqR1\", \"uriFastqR2\"] for m in [\"Hashtag\", \"CiteSeq\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'Hashtag': ['R1', 'R2'],\n",
    "    'CiteSeq': ['R1', 'R2'],\n",
    "    'AsapSeq': ['R1', 'R2', 'R3'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get s3 path of existing GEX analysis files\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_wl_dir(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_hashtag_lib = \"peer_lab_db.hashtag_lib\"\n",
    "        table_genome_index = \"peer_lab_db.genome_index\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        '''\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_stats_data}.analysis_storage\n",
    "        FROM {table_sample_data}\n",
    "        LEFT JOIN {table_stats_data} \n",
    "        ON {table_stats_data}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_hashtag_lib}\n",
    "        ON {table_hashtag_lib}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_genome_index}\n",
    "        ON {table_genome_index}.id = {table_hashtag_lib}.genomeIndex_id\n",
    "        LEFT JOIN {table_sc_tech}\n",
    "        ON {table_sc_tech}.id = {table_genome_index}.scTech_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        if result: \n",
    "            return result\n",
    "        # As backup, get AWS storage location directly from sample_data\n",
    "        else:\n",
    "        '''\n",
    "        query = f\"\"\"\n",
    "        SELECT AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get white list method and associated file\n",
    "# Throws exception if no white list exists\n",
    "def get_wl_params(\n",
    "    sample_id: str,\n",
    "    user: str,\n",
    "    password: str,\n",
    "):\n",
    "    wl_params = dict()\n",
    "\n",
    "    wl_dir = get_wl_dir(sample_id, user, password)\n",
    "    wl_patterns = [sharp_wl_pattern_map[p] for p in sharp_wl_priority_map[prefix]]\n",
    "\n",
    "    try:\n",
    "        # Check white list file exists before loading info from database\n",
    "        assert wl_dir, f\"Empty analysis storage for sample id {sample_id}\"\n",
    "        _, bucket, key, _, _ = urllib.parse.urlsplit(wl_dir)\n",
    "        # White list file and method is first entry found on S3 \n",
    "        wl = pd.DataFrame(\n",
    "            [get_s3_objects(bucket, key.strip(\"/\"), re.compile(p)) for p in wl_patterns],\n",
    "            index = sharp_wl_priority_map[prefix],\n",
    "        ).dropna(how=\"all\")\n",
    "        try:\n",
    "            wl_key = wl.iloc[0,0] # if empty, missing white list file\n",
    "            wl_params[\"uri\"] = os.path.join(\"s3://\", bucket, wl_key)\n",
    "            wl_params[\"method\"] = sharp_wl_method_map[wl.index[0]]\n",
    "        except IndexError:\n",
    "            logging.error(\n",
    "                \"Path to barcodes or counts matrix of GEX data is missing!\"\n",
    "            )\n",
    "            return\n",
    "\n",
    "    except AssertionError:\n",
    "        logging.warning(f\"Path to GEX output results is missing for {sample_id}!\")\n",
    "        return\n",
    "\n",
    "    return wl_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bc_json_manual(\n",
    "    prefix,\n",
    "    platform,\n",
    "    user, \n",
    "    password,\n",
    "):\n",
    "    try:\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcodes\n",
    "        FROM {table_sc_tech}\n",
    "        WHERE {table_sc_tech}.sc_Tech = \"{platform.upper()}_{prefix}\"\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat barcodes from collaborator excel file\n",
    "def get_bcs_manual(\n",
    "    path_to_excel,\n",
    "):\n",
    "    # Replace unrecognized characters with text description\n",
    "    barcodes = pd.read_excel(path_to_excel) \n",
    "    replace = lambda x: x.encode('ascii', 'namereplace').decode().replace(\"\\\\N\", \"\")\n",
    "    barcodes[\"Description\"] = barcodes[\"Description\"].apply(replace)\n",
    "    barcodes[\"BP Shift\"] = 0\n",
    "    return barcodes[[\"Barcode\", \"DNA_ID\", \"Description\", \"BP Shift\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bc_params_manual(\n",
    "    barcodes,\n",
    "    prefix, \n",
    "    platform,\n",
    "    creds,\n",
    "):\n",
    "    bc_params = dict()\n",
    "    \n",
    "    # Add bp shift and sequence length based on conjugation\n",
    "    bp_shift_map = {\n",
    "        \"A\": 0,\n",
    "        \"B\": 10,\n",
    "        \"C\": 10,\n",
    "        \"M\": 1, # Methanol\n",
    "    }\n",
    "    conjugation = barcodes[\"DNA_ID\"].str.get(0)\n",
    "    if conjugation.nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample has multiple hashtag barcode categories and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else:\n",
    "        bc_params[\"conjugation\"] = conjugation.values[0]\n",
    "        bc_params[\"bp_shift\"] = bp_shift_map[bc_params[\"conjugation\"]]\n",
    "        bc_params[\"seq_length\"] = bc_params[\"bp_shift\"] + barcodes[\"Barcode\"].apply(len).max()\n",
    "    \n",
    "    # JSON of bc and UMI positions are stored in database\n",
    "    bc_json = get_bc_json_manual(prefix, platform, creds['user'], creds['password'])\n",
    "    bc_pos = json.loads(bc_json)\n",
    "    bc_params[\"cb\"] = bc_pos[\"cellbarcode\"]\n",
    "    bc_params[\"umi\"] = bc_params[\"cb\"] + bc_pos[\"UMIs\"]\n",
    "    \n",
    "    return bc_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_template(\n",
    "    prefix: str, \n",
    "    conjugation: str = None,\n",
    "):\n",
    "    if prefix==\"Hashtag\":\n",
    "        return {\n",
    "            \"A\": \"hashtag-10x-v3-tsa\",\n",
    "            \"B\": \"hashtag-10x-v3-tsb\",\n",
    "            \"C\": \"hashtag-10x-tsc\",\n",
    "        }[conjugation]\n",
    "    elif prefix==\"CiteSeq\":\n",
    "        return \"citeseq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Numpy encoder for JSON from pandas series\n",
    "class NpEncoder(json.JSONEncoder):\n",
    "    def default(self, obj):\n",
    "        if isinstance(obj, np.integer):\n",
    "            return int(obj)\n",
    "        elif isinstance(obj, np.floating):\n",
    "            return float(obj)\n",
    "        elif isinstance(obj, np.ndarray):\n",
    "            return obj.tolist()\n",
    "        else:\n",
    "            return super(NpEncoder, self).default(obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from SCRIdb\n",
    "def get_s3_objects(bucket, key, pattern, full_uri=False):\n",
    "    \n",
    "    s3r = boto3.resource(\"s3\")\n",
    "    bucket_s3 = s3r.Bucket(bucket)\n",
    "    objects = []\n",
    "    for obj in bucket_s3.objects.filter(Prefix=key):\n",
    "        hit = pattern.search(obj.key)\n",
    "        if hit:\n",
    "            objects.append(obj.key)\n",
    "    if full_uri:\n",
    "        objects = [f\"s3://{bucket}/{o}\" for o in objects]\n",
    "    return objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def execute_query(query, user, password):\n",
    "    with connect(\n",
    "        host=\"peer-lab-db.cggxmlwgzzpw.us-east-1.rds.amazonaws.com\",\n",
    "        database=\"peer_lab_db\",\n",
    "        user=user,\n",
    "        password=password,\n",
    "    ) as connection:\n",
    "        with connection.cursor(buffered=True) as cursor:\n",
    "            cursor.execute(query)\n",
    "            result = cursor.fetchall()\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_species(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_species = \"peer_lab_db.species\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_species}.Species\n",
    "        FROM {table_species}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_species}.id = {table_genome_idx}.species_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sc_tech(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sc_tech}.sc_Tech\n",
    "        FROM {table_sc_tech}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_sc_tech}.id = {table_genome_idx}.scTech_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sample_id(sample_name, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.Sample=\"{sample_name}\"\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_project_id(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_SEQC_version(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 cp {loc}/seqc-results/seqc_log.txt -\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        version = re.match(r\".*SEQC=v(\\d+\\.\\d+\\.\\d+).*\", out)[1]\n",
    "        return version\n",
    "    except:\n",
    "        return \"N/A\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_file_prefix(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 ls {loc}/seqc-results/\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        \n",
    "        # Note: I'm expecting the aligned bam file to be in loc\n",
    "        bam_pattern = re.compile(r\"(.*)_Aligned\\.out\\.bam$\")\n",
    "        filename = list(filter(bam_pattern.match, out.split()))[0]\n",
    "        file_prefix = re.match(bam_pattern, filename)[1]\n",
    "        return file_prefix\n",
    "    except:\n",
    "        raise ValueError(f\"BAM file not found in {loc}\")\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference(sample_id):\n",
    "    # Get species from database to decide reference\n",
    "    species = get_species(sample_id, creds[\"user\"], creds[\"password\"])\n",
    "    \n",
    "    # Map to reference locations\n",
    "    if \"Human\" in species:\n",
    "        return \"s3://seqc-public/genomes/hg38_long_polya/annotations.gtf\"\n",
    "    elif \"Mouse\" in species:\n",
    "        return \"s3://seqc-public/genomes/mm38_long_polya/annotations.gtf\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown Species: {species}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bc_whitelist(sample_id):\n",
    "    # Get version from database to decide whitelist\n",
    "    sc_tech = get_sc_tech(sample_id, creds[\"user\"], creds[\"password\"])\n",
    "    \n",
    "    # Map to reference locations\n",
    "    if \"V3\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v3/flat/3M-february-2018.txt\"\n",
    "    elif \"V2\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v2/flat/737K-august-2016.txt\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown Technology: {sc_tech}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(\n",
    "    workflow_path: str,\n",
    "    execp: str,\n",
    "    secrets: str,\n",
    "    inputs: str,\n",
    "    labels: str,\n",
    "    options: str,\n",
    "):\n",
    "    # change working directory to the pipeline package\n",
    "    oldwd = os.getcwd()\n",
    "    os.chdir(workflow_path)\n",
    "    \n",
    "    # execute the pipeline command\n",
    "    cmd = f\"{workflow_path}/{execp} -k {secrets} -i {inputs} -l {labels} -o {options}\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)\n",
    "    out = var.__dict__\n",
    "    \n",
    "    # change working directory back\n",
    "    os.chdir(oldwd)\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get bc and UMI positions from database stored in JSON format\n",
    "def get_bc_json(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_stats_data = \"peer_lab_db.stats_data\"\n",
    "        table_hashtag_lib = \"peer_lab_db.hashtag_lib\"\n",
    "        table_genome_index = \"peer_lab_db.genome_index\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcodes\n",
    "        FROM {table_sample_data}\n",
    "        LEFT JOIN {table_stats_data} \n",
    "        ON {table_stats_data}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_hashtag_lib}\n",
    "        ON {table_hashtag_lib}.sampleData_id = {table_sample_data}.id\n",
    "        LEFT JOIN {table_genome_index}\n",
    "        ON {table_genome_index}.id = {table_hashtag_lib}.genomeIndex_id\n",
    "        LEFT JOIN {table_sc_tech}\n",
    "        ON {table_sc_tech}.id = {table_genome_index}.scTech_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get bc sequence data from database\n",
    "def get_bcs(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_hashtag_barcodes = \"peer_lab_db.hashtag_barcodes\"\n",
    "        table_hashtags = \"peer_lab_db.hashtags\"\n",
    "        query = f\"\"\"\n",
    "        SELECT barcode_sequence, concat(substring(category, -1), barcode), \n",
    "        demultiplex_label, bp_shift FROM {table_hashtags} \n",
    "        LEFT JOIN {table_hashtag_barcodes} \n",
    "        ON {table_hashtag_barcodes}.id = {table_hashtags}.hashtagBarcodes_id \n",
    "        WHERE {table_hashtags}.sampleData_id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bc_params(\n",
    "    sample_id,\n",
    "    user,\n",
    "    password,\n",
    "):\n",
    "    bc_params = dict()\n",
    "\n",
    "    # JSON of bc and UMI positions are stored in database\n",
    "    # First check dense matrix exists before loading JSON from database\n",
    "    bc_json = get_bc_json(sample_id, user, password)\n",
    "    bc_pos = json.loads(bc_json)\n",
    "    bc_params[\"cb\"] = bc_pos[\"cellbarcode\"]\n",
    "    bc_params[\"umi\"] = bc_params[\"cb\"] + bc_pos[\"UMIs\"]\n",
    "\n",
    "    # Get bc sequence data from database\n",
    "    bcs = get_bcs(sample_id, user, password)\n",
    "    if not bcs:\n",
    "        logging.warning(f\"Barcodes data Empty:\\n\\t {db_connect.cur.statement}\")\n",
    "        return\n",
    "    for bc in bcs:\n",
    "        try:\n",
    "            assert bc[0], \"AssertionError: Missing sequence barcodes!\"\n",
    "            assert bc[1], \"AssertionError: Missing barcode IDs\"\n",
    "        except AssertionError as err:\n",
    "            logging.warning(f\"{err}:\\n\\t {db_connect.cur.statement}\")\n",
    "            return\n",
    "\n",
    "    barcodes = pd.DataFrame(bcs, columns=[\"sequence\", \"code\", \"label\", \"bp_shift\"])\n",
    "    conjugation = barcodes[\"code\"].str.get(0)\n",
    "    if conjugation.nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample has multiple hashtag barcode categories and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else:\n",
    "        bc_params[\"conjugation\"] = conjugation.values[0]\n",
    "\n",
    "    if barcodes[\"bp_shift\"].nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample {sample_id} has hashtag barcode categories, with bp-shift length/s \"\n",
    "            f\"{barcodes['bp_shift'].unique()}, and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else: \n",
    "        bc_params[\"bp_shift\"] = int(barcodes[\"bp_shift\"][0])\n",
    "        bc_params[\"seq_length\"] = bc_params[\"bp_shift\"] + barcodes[\"sequence\"].apply(len).max()\n",
    "\n",
    "    return bc_params"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** In the \"Execution\" section below, there are two methods to process Sharp datasets:\n",
    "1. Using barcode information that has already been entered into SCRIdb.\n",
    "2. Using information provided manually in an Excel file (see formatting guide).\n",
    "\n",
    "The \"Setup\" section can be run irrespective of which method you are using. The primary input that needs to be provided in this section is a list of all S3 paths for each sample to process. Each path should look something like:\n",
    "\n",
    "`s3://dp-lab-data/collaborators/your-collaborator/your-project-name/your-sample-name`\n",
    "\n",
    "You will also need to set the `prefix` variable to the workflow you want to run (e.g., `CiteSeq` or `Hashtag`). Other inputs in the \"Setup\" section include local paths to relevant files and directories; likely, these will only need to be set once."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps from .wdl name (prefix) to results dirname\n",
    "results_dirs = {\n",
    "    \"Hashtag\": \"Hashtag-results-SEQC-test-2\",\n",
    "    \"CiteSeq\": \"CiteSeq-results\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to shell script\n",
    "sh_files = {\n",
    "    \"Hashtag\": \"submit-hashtag.sh\",\n",
    "    \"CiteSeq\": \"submit-citeseq.sh\",\n",
    "}\n",
    "\n",
    "# Maps from .wdl name (prefix) to pipeline name\n",
    "pipeline_types = {\n",
    "    \"Hashtag\": \"Hashtag\",\n",
    "    \"CiteSeq\": \"CITE-seq\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"Hashtag\" # Workflow(s) to run; also .wdl filename prefix(es)\n",
    "pipeline_type = pipeline_types[prefix] # field in *.labels.json\n",
    "output_dirname = results_dirs[prefix]\n",
    "\n",
    "# If need to add comment, put here\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sharp-specific parameters\n",
    "options_prefix=\"Sharp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = f\"{Path.home()}/scing/bin/sharp-0.1.1\" # CHANGE THIS\n",
    "path_to_exec = sh_files[prefix]\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = f\"{workflow_dir}/{options_prefix}.options.aws.json\"\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS\n",
    "barcodes_path = f\"{Path.home()}/scing/data/barcodes\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'user': 'moormana',\n",
       " 'password': 'v815cCVLLP!',\n",
       " 'host': 'peer-lab-db.cggxmlwgzzpw.us-east-1.rds.amazonaws.com',\n",
       " 'database': 'peer_lab_db',\n",
       " 'path_submit_data': '/Users/moormana/Documents/Pipeline/SCRIdb/submit_data.csv',\n",
       " 'dockerizedSEQC': '/Users/moormana/Documents/GitHub/seqc',\n",
       " 'email': 'moormana@mskcc.org',\n",
       " 'pem': '/Users/moormana/.ssh/moormana.pem',\n",
       " 'ami': 'ami-02f92579154b6edf8',\n",
       " 'instance_type': 'r5.4xlarge'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "creds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Samples on which to run Sharp\n",
    "# Notes: \n",
    "# - Assumes data is transferred to AWS s3 (this should be an s3 location)\n",
    "# - Assumes directory name is also name of sample\n",
    "# - Workflows above will be run on all samples below\n",
    "common_dir = \"\"\n",
    "sample_paths = [\n",
    "    \"s3://dp-lab-data/collaborators/tammela/C57BL6/XZ-2284_20230405_Hyperoxia_day1_HTO\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Execution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Method 1: From SCRIdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_bc_params(\n",
    "    sample_id,\n",
    "    user,\n",
    "    password,\n",
    "):\n",
    "    bc_params = dict()\n",
    "\n",
    "    # JSON of bc and UMI positions are stored in database\n",
    "    # First check dense matrix exists before loading JSON from database\n",
    "    bc_json = get_bc_json(sample_id, user, password)\n",
    "    bc_pos = json.loads(bc_json)\n",
    "    bc_params[\"cb\"] = bc_pos[\"cellbarcode\"]\n",
    "    bc_params[\"umi\"] = bc_params[\"cb\"] + bc_pos[\"UMIs\"]\n",
    "\n",
    "    # Get bc sequence data from database\n",
    "    bcs = get_bcs(sample_id, user, password)\n",
    "    print(bcs)\n",
    "    if not bcs:\n",
    "        logging.warning(f\"Barcodes data Empty:\\n\\t {db_connect.cur.statement}\")\n",
    "        return\n",
    "    for bc in bcs:\n",
    "        try:\n",
    "            assert bc[0], \"AssertionError: Missing sequence barcodes!\"\n",
    "            assert bc[1], \"AssertionError: Missing barcode IDs\"\n",
    "        except AssertionError as err:\n",
    "            logging.warning(f\"{err}:\\n\\t {db_connect.cur.statement}\")\n",
    "            return\n",
    "\n",
    "    barcodes = pd.DataFrame(bcs, columns=[\"sequence\", \"code\", \"label\", \"bp_shift\"])\n",
    "    conjugation = barcodes[\"code\"].str.get(0)\n",
    "    if conjugation.nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample has multiple hashtag barcode categories and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else:\n",
    "        bc_params[\"conjugation\"] = conjugation.values[0]\n",
    "\n",
    "    if barcodes[\"bp_shift\"].nunique() != 1:\n",
    "        logging.warning(\n",
    "            f\"Sample {sample_id} has hashtag barcode categories, with bp-shift length/s \"\n",
    "            f\"{barcodes['bp_shift'].unique()}, and will not be processed!\"\n",
    "        )\n",
    "        return\n",
    "    else: \n",
    "        bc_params[\"bp_shift\"] = int(barcodes[\"bp_shift\"][0])\n",
    "        bc_params[\"seq_length\"] = bc_params[\"bp_shift\"] + barcodes[\"sequence\"].apply(len).max()\n",
    "\n",
    "    return bc_params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('AAAGCATTCTTCACG', 'B0304', '1636', 10), ('CTTTGTCTTTGTGAG', 'B0305', '1637', 10), ('ACCCACCAGTAAGAC', 'B0301', '1634', 10), ('CTTGCCGCATGTCAT', 'B0303', '1635', 10), ('TATGCTGCCACGGTA', 'B0306', '1638', 10), ('GAGTCTGCCAGTATC', 'B0307', '1639', 10), ('TATAGAACGCCAGGC', 'B0308', '1645', 10), ('TGCCTATGAAACAAG', 'B0309', '1646', 10)]\n"
     ]
    }
   ],
   "source": [
    "# Get information for all samples\n",
    "sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "sample_names = [os.path.basename(s) for s in sample_paths]\n",
    "sample_names = [re.match(r'(.*)_.+$', s)[1] for s in sample_names] # remove library suffix (e.g. _CITE, _HTO, etc.)\n",
    "# TODO: assert basename is in peer_lab_db.sample_data.Sample\n",
    "# assert(all(check_sample_name(s) for s in sample_names))\n",
    "samples = pd.DataFrame(\n",
    "    sample_paths,\n",
    "    index=sample_names,\n",
    "    columns=[\"S3_Path\"],\n",
    "    dtype=str,\n",
    ")\n",
    "# Get FASTQ paths from S3\n",
    "# Note: Uses same FASTQ file ids for all samples\n",
    "fastq_file_ids = fastq_map[prefix]\n",
    "samples[\"FASTQs\"] = samples[\"S3_Path\"].apply(lambda x: get_fastqs(x, fastq_file_ids))\n",
    "\n",
    "samples[\"Sample_ID\"] = pd.Series(samples.index).apply(\n",
    "    lambda x: get_sample_id(x, creds['user'], creds['password'])\n",
    ").values\n",
    "samples[\"Whitelist_Params\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_wl_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "samples[\"Barcode_Params\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_bc_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "samples[\"Barcodes\"] = samples['Sample_ID'].apply(\n",
    "    lambda x: get_bcs(x, creds['user'], creds['password'])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['s3://dp-lab-data/collaborators/tammela/C57BL6/XZ-2284_20230405_Hyperoxia_day1/seqc-results/4872_XZ-2284_20230405_Hyperoxia_day1_IGO_12437_CX_30_dense.csv'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[\"Whitelist_Params\"].apply(lambda x: x['uri']).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ebcfc59577f94ba79c3e720dc5f96805",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Function to reformat barcode labels for Sharp\n",
    "def reformat_bc_label(label):\n",
    "    label = label.encode('ascii', 'namereplace').decode()\n",
    "    label = label.replace(\"\\\\N\", \"\").replace(\" \", \"_\")\n",
    "    return label\n",
    "\n",
    "for name, sample in tqdm(samples.iterrows(), total=len(samples)):\n",
    "    \n",
    "    # Reformat barcodes\n",
    "    barcodes = pd.DataFrame(\n",
    "        sample[\"Barcodes\"],\n",
    "        columns=[\"sequence\", \"code\", \"label\", \"bp_shift\"]\n",
    "    )\n",
    "    barcodes[\"label\"] = barcodes[\"label\"].apply(reformat_bc_label)\n",
    "    \n",
    "    # Save to CSV\n",
    "    path_to_csv = f\"{barcodes_path}/{sample['Sample_ID']}_tag-list.csv\"\n",
    "    barcodes.to_csv(path_to_csv, header=False, index=False)\n",
    "    \n",
    "    # Upload tag-list to AWS\n",
    "    cmd = f\"aws s3 cp {path_to_csv} {sample['S3_Path']}/{output_dirname}/tag-list.csv\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'uri': 's3://dp-lab-data/collaborators/tammela/C57BL6/XZ-2284_20230405_Hyperoxia_day1/seqc-results/4872_XZ-2284_20230405_Hyperoxia_day1_IGO_12437_CX_30_dense.csv',\n",
       " 'method': 'SeqcDenseCountsMatrixCsv'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['Whitelist_Params'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "conjugation = samples[\"Barcode_Params\"].apply(lambda x: x[\"conjugation\"]).values[0] # conjugation must be same for all samples\n",
    "template_prefix = get_template(prefix, conjugation)\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "std_inputs_fields.remove(f\"{prefix}.denseCountMatrix\")\n",
    "    \n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "\n",
    "platform = \"10x_v3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.sampleName\"] = inputs.index\n",
    "inputs[f\"{prefix}.scRnaSeqPlatform\"] = platform # may need to change\n",
    "inputs[f\"{prefix}.lengthR1\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.lengthR2\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"seq_length\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhitelistUri\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"uri\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhiteListMethod\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"method\"])\n",
    "inputs[f\"{prefix}.{'tagList' if (prefix=='CiteSeq') else 'hashTagList'}\"] = \\\n",
    "    samples[\"S3_Path\"] + f\"/{output_dirname}/tag-list.csv\" #TODO: Ask about changing all to 'tagList'\n",
    "inputs[f\"{prefix}.cbStartPos\"] = 1\n",
    "inputs[f\"{prefix}.cbEndPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"cb\"])\n",
    "inputs[f\"{prefix}.umiStartPos\"] = inputs[f\"{prefix}.cbEndPos\"]+1\n",
    "inputs[f\"{prefix}.umiEndPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.trimPos\"] = samples[\"Barcode_Params\"].apply(lambda x: x[\"bp_shift\"])\n",
    "inputs[f\"{prefix}.translate10XBarcodes\"] = \\\n",
    "    (inputs[f\"{prefix}.scRnaSeqPlatform\"] == \"10x_v3\") & \\\n",
    "    (samples[\"Barcode_Params\"].apply(lambda x: x[\"conjugation\"] == \"B\"))\n",
    "\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "for file_id in fastq_file_ids: # Set FASTQs\n",
    "    inputs[f\"{prefix}.uriFastq{file_id}\"] = samples[\"FASTQs\"].apply(lambda x: x[file_id])\n",
    "\n",
    "# ********************\n",
    "# Defaults\n",
    "# Note: These may need to be changed on a per-sample or per-execution basis\n",
    "\n",
    "inputs[f\"{prefix}.slidingWindowSearch\"] = False\n",
    "inputs[f\"{prefix}.cbCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.umiCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.numExpectedCells\"] = 5000\n",
    "# Need trick to set dictionary for each row\n",
    "common_resource_spec = {\n",
    "    \"cpu\": 32,\n",
    "    \"memory\": -1,\n",
    "}\n",
    "inputs[f\"{prefix}.resourceSpec\"] = inputs.iloc[:, 0].apply(lambda x: common_resource_spec)\n",
    "if prefix == \"Hashtag\":\n",
    "    inputs[f\"{prefix}.minCount\"] = 1\n",
    "\n",
    "# ********************\n",
    "\n",
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds[\"user\"], creds[\"password\"]))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_Path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "assert set(inputs.columns).issuperset(std_inputs_fields) & (inputs.notna().values.all())\n",
    "assert set(labels.columns).issuperset(std_labels_fields) & (labels.notna().values.all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>XZ-2284_20230405_Hyperoxia_day1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hashtag.uriFastqR1</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/tamm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.uriFastqR2</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/tamm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.sampleName</th>\n",
       "      <td>XZ-2284_20230405_Hyperoxia_day1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.scRnaSeqPlatform</th>\n",
       "      <td>10x_v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.lengthR1</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.lengthR2</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.cellBarcodeWhitelistUri</th>\n",
       "      <td>s3://dp-lab-data/collaborators/tamme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.cellBarcodeWhiteListMethod</th>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.translate10XBarcodes</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.hashTagList</th>\n",
       "      <td>s3://dp-lab-data/collaborators/tamme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.cbStartPos</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.cbEndPos</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.umiStartPos</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.umiEndPos</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.trimPos</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.slidingWindowSearch</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.cbCollapsingDistance</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.umiCollapsingDistance</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.numExpectedCells</th>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.minCount</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.resourceSpec</th>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hashtag.dockerRegistry</th>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            XZ-2284_20230405_Hyperoxia_day1\n",
       "Hashtag.uriFastqR1                  [s3://dp-lab-data/collaborators/tamm...\n",
       "Hashtag.uriFastqR2                  [s3://dp-lab-data/collaborators/tamm...\n",
       "Hashtag.sampleName                          XZ-2284_20230405_Hyperoxia_day1\n",
       "Hashtag.scRnaSeqPlatform                                             10x_v3\n",
       "Hashtag.lengthR1                                                         28\n",
       "Hashtag.lengthR2                                                         25\n",
       "Hashtag.cellBarcodeWhitelistUri     s3://dp-lab-data/collaborators/tamme...\n",
       "Hashtag.cellBarcodeWhiteListMethod                 SeqcDenseCountsMatrixCsv\n",
       "Hashtag.translate10XBarcodes                                           True\n",
       "Hashtag.hashTagList                 s3://dp-lab-data/collaborators/tamme...\n",
       "Hashtag.cbStartPos                                                        1\n",
       "Hashtag.cbEndPos                                                         16\n",
       "Hashtag.umiStartPos                                                      17\n",
       "Hashtag.umiEndPos                                                        28\n",
       "Hashtag.trimPos                                                          10\n",
       "Hashtag.slidingWindowSearch                                           False\n",
       "Hashtag.cbCollapsingDistance                                              1\n",
       "Hashtag.umiCollapsingDistance                                             1\n",
       "Hashtag.numExpectedCells                                               5000\n",
       "Hashtag.minCount                                                          1\n",
       "Hashtag.resourceSpec                              {'cpu': 32, 'memory': -1}\n",
       "Hashtag.dockerRegistry                                      quay.io/hisplan"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XZ-2284_20230405_Hyperoxia_day1</th>\n",
       "      <td>Hashtag</td>\n",
       "      <td>C57BL6</td>\n",
       "      <td>XZ-2284_20230405_Hyperoxia_day1</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/tamme...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                pipelineType project  \\\n",
       "XZ-2284_20230405_Hyperoxia_day1      Hashtag  C57BL6   \n",
       "\n",
       "                                                          sample     owner  \\\n",
       "XZ-2284_20230405_Hyperoxia_day1  XZ-2284_20230405_Hyperoxia_day1  moormana   \n",
       "\n",
       "                                                             destination  \\\n",
       "XZ-2284_20230405_Hyperoxia_day1  s3://dp-lab-data/collaborators/tamme...   \n",
       "\n",
       "                                transfer   comment  \n",
       "XZ-2284_20230405_Hyperoxia_day1        -  moormana  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e78bae58fbd74aba80565e3e32c6d6dc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(samples.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts[0]['stderr']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Method 2: From Excel File"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Excel Formatting Guidelines:* \n",
    "Files should contains three columns:\n",
    "1. *Barcode*, the barcode sequence (e.g., \"TCAATCTCCTCCTAT\")\n",
    "2. *DNA_ID*, the TotalSeq ID (e.g., \"A0964\")\n",
    "3. *Description*, a human readable description of the barcode (e.g. \"anti-human L1CAM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual inputs\n",
    "platform = \"10x_v3\"\n",
    "path_to_excel = f\"{Path.home()}/scing/data/barcodes/210331_Citeseq_CB.xlsx\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get information for all samples\n",
    "sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "sample_names = [os.path.basename(s) for s in sample_paths]\n",
    "sample_names = [re.match(r'(.*)_.+$', s)[1] for s in sample_names] # remove library suffix (e.g. _CITE, _HTO, etc.)\n",
    "# TODO: assert basename is in peer_lab_db.sample_data.Sample\n",
    "# assert(all(check_sample_name(s) for s in sample_names))\n",
    "samples = pd.DataFrame(\n",
    "    sample_paths,\n",
    "    index=sample_names,\n",
    "    columns=[\"S3_Path\"],\n",
    "    dtype=str,\n",
    ")\n",
    "samples[\"Sample_ID\"] = pd.Series(samples.index).apply(\n",
    "    lambda x: get_sample_id(x, creds['user'], creds['password'])\n",
    ").values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RS-1778_1to6</th>\n",
       "      <td>s3://dp-lab-data/collaborators/chaud...</td>\n",
       "      <td>4009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              S3_Path  Sample_ID\n",
       "RS-1778_1to6  s3://dp-lab-data/collaborators/chaud...       4009"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "barcodes = pd.DataFrame(\n",
    "    [[\"TCAATCTCCTCCTAT\", \"A0964\", \"anti-human L1CAM\"]],\n",
    "    columns=[\"Barcode\", \"DNA_ID\", \"Description\"],\n",
    ")\n",
    "barcodes.to_excel(path_to_excel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read barcodes from file\n",
    "# Note: Must be subset to HTO or CITE barcodes before next step!\n",
    "barcodes = get_bcs_manual(path_to_excel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>DNA_ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>BP Shift</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGTAACGTAGAGCGA</td>\n",
       "      <td>A0224</td>\n",
       "      <td>TCRA/B</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGTTCCCGCTCAACT</td>\n",
       "      <td>A0072</td>\n",
       "      <td>CD4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTGCGCTTTCCATT</td>\n",
       "      <td>A0080</td>\n",
       "      <td>CD8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCAATCCTTCCGCTT</td>\n",
       "      <td>A0063</td>\n",
       "      <td>CD45RA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTCCGAATCATGTTG</td>\n",
       "      <td>A0087</td>\n",
       "      <td>CD45RO</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GTCCCTGCAACTTGA</td>\n",
       "      <td>A0147</td>\n",
       "      <td>CD62L</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AGTTCAGTCAACCGA</td>\n",
       "      <td>A0148</td>\n",
       "      <td>CCR7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GTGTGTTGTCCTATG</td>\n",
       "      <td>A0390</td>\n",
       "      <td>IL7R</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CCTATTCCGATTCCG</td>\n",
       "      <td>A0410</td>\n",
       "      <td>CD38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AATTCAACCGTCGCC</td>\n",
       "      <td>A0144</td>\n",
       "      <td>CXCR5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AGCTTACCTGCACGA</td>\n",
       "      <td>A0071</td>\n",
       "      <td>CCR4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GCGATGGTAGATTAT</td>\n",
       "      <td>A0140</td>\n",
       "      <td>CXCR3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GATCCCTTTGTCACT</td>\n",
       "      <td>A0143</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TTTGTCCTGTACGCC</td>\n",
       "      <td>A0085</td>\n",
       "      <td>CD25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACAGCGCCGTATTTA</td>\n",
       "      <td>A0088</td>\n",
       "      <td>PD1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GCTAGATAGATGCAA</td>\n",
       "      <td>A0032</td>\n",
       "      <td>CD154</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AATAGCGAGCAAGTA</td>\n",
       "      <td>A0159</td>\n",
       "      <td>HLA-DR</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CGCGCACCCATTAAA</td>\n",
       "      <td>A0171</td>\n",
       "      <td>ICOS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CAAGGTATGTCACTG</td>\n",
       "      <td>A0191</td>\n",
       "      <td>CD27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TTGCTTACCGCCAGA</td>\n",
       "      <td>A0089</td>\n",
       "      <td>TIGIT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Barcode DNA_ID Description  BP Shift\n",
       "0   CGTAACGTAGAGCGA  A0224      TCRA/B         0\n",
       "1   TGTTCCCGCTCAACT  A0072         CD4         0\n",
       "2   GCTGCGCTTTCCATT  A0080         CD8         0\n",
       "3   TCAATCCTTCCGCTT  A0063      CD45RA         0\n",
       "4   CTCCGAATCATGTTG  A0087      CD45RO         0\n",
       "5   GTCCCTGCAACTTGA  A0147       CD62L         0\n",
       "6   AGTTCAGTCAACCGA  A0148        CCR7         0\n",
       "7   GTGTGTTGTCCTATG  A0390        IL7R         0\n",
       "8   CCTATTCCGATTCCG  A0410        CD38         0\n",
       "9   AATTCAACCGTCGCC  A0144       CXCR5         0\n",
       "10  AGCTTACCTGCACGA  A0071        CCR4         0\n",
       "11  GCGATGGTAGATTAT  A0140       CXCR3         0\n",
       "12  GATCCCTTTGTCACT  A0143        CCR6         0\n",
       "13  TTTGTCCTGTACGCC  A0085        CD25         0\n",
       "14  ACAGCGCCGTATTTA  A0088         PD1         0\n",
       "15  GCTAGATAGATGCAA  A0032       CD154         0\n",
       "16  AATAGCGAGCAAGTA  A0159      HLA-DR         0\n",
       "17  CGCGCACCCATTAAA  A0171        ICOS         0\n",
       "18  CAAGGTATGTCACTG  A0191        CD27         0\n",
       "19  TTGCTTACCGCCAGA  A0089       TIGIT         0"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "216011434b4348869412b987994ebc2d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/11 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Upload barcodes for all samples\n",
    "# Note: Uploads same barcodes to s3 for all samples\n",
    "path_to_csv = path_to_excel.replace(\".xlsx\", \".csv\")\n",
    "barcodes.to_csv(path_to_csv, header=False, index=False)\n",
    "for s3_path in tqdm(samples[\"S3_Path\"]):\n",
    "    cmd = f\"aws s3 cp {path_to_csv} {s3_path}/{output_dirname}/tag-list.csv\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: Assumes GEX data is recorded in database\n",
    "samples[\"Whitelist_Params\"] = samples[\"Sample_ID\"].apply(\n",
    "    lambda x: get_wl_params(x, creds['user'], creds['password'])\n",
    ")\n",
    "assert ~samples[\"Whitelist_Params\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: Adds same barcode parameters for all samples\n",
    "samples[\"BC_Params\"] = samples[\"Sample_ID\"].apply(\n",
    "    lambda x: get_bc_params_manual(barcodes, prefix, platform, creds)\n",
    ")\n",
    "assert ~samples[\"BC_Params\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get FASTQ paths from S3\n",
    "# Note: Uses same FASTQ file ids for all samples\n",
    "fastq_file_ids = fastq_map[prefix]\n",
    "samples[\"FASTQs\"] = samples[\"S3_Path\"].apply(lambda x: get_fastqs(x, fastq_file_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "conjugation = samples[\"BC_Params\"].apply(lambda x: x[\"conjugation\"]).values[0] # conjugation must be same for all samples\n",
    "template_prefix = get_template(prefix, conjugation)\n",
    "with open(f\"{config_dir}/{template_prefix}.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "with open(f\"{config_dir}/{template_prefix}.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "\n",
    "platform = \"10x_v3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.sampleName\"] = inputs.index\n",
    "inputs[f\"{prefix}.scRnaSeqPlatform\"] = platform # may need to change\n",
    "inputs[f\"{prefix}.lengthR1\"] = samples[\"BC_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.lengthR2\"] = samples[\"BC_Params\"].apply(lambda x: x[\"seq_length\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhitelistUri\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"uri\"])\n",
    "inputs[f\"{prefix}.cellBarcodeWhiteListMethod\"] = samples[\"Whitelist_Params\"].apply(lambda x: x[\"method\"])\n",
    "inputs[f\"{prefix}.{'tagList' if (prefix=='CiteSeq') else 'hashTagList'}\"] = \\\n",
    "    samples[\"S3_Path\"] + f\"/{output_dirname}/tag-list.csv\" #TODO: Ask about changing all to 'tagList'\n",
    "inputs[f\"{prefix}.cbStartPos\"] = 1\n",
    "inputs[f\"{prefix}.cbEndPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"cb\"])\n",
    "inputs[f\"{prefix}.umiStartPos\"] = inputs[f\"{prefix}.cbEndPos\"]+1\n",
    "inputs[f\"{prefix}.umiEndPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"umi\"])\n",
    "inputs[f\"{prefix}.trimPos\"] = samples[\"BC_Params\"].apply(lambda x: x[\"bp_shift\"])\n",
    "inputs[f\"{prefix}.translate10XBarcodes\"] = \\\n",
    "    (inputs[f\"{prefix}.scRnaSeqPlatform\"] == \"10x_v3\") & \\\n",
    "    (samples[\"BC_Params\"].apply(lambda x: x[\"conjugation\"] == \"B\"))\n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "for file_id in fastq_file_ids: # Set FASTQs\n",
    "    inputs[f\"{prefix}.uriFastq{file_id}\"] = samples[\"FASTQs\"].apply(lambda x: x[file_id])\n",
    "\n",
    "# ********************\n",
    "# Defaults\n",
    "# Note: These may need to be changed on a per-sample or per-execution basis\n",
    "\n",
    "inputs[f\"{prefix}.slidingWindowSearch\"] = False\n",
    "inputs[f\"{prefix}.cbCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.umiCollapsingDistance\"] = 1\n",
    "inputs[f\"{prefix}.numExpectedCells\"] = 0\n",
    "# Need trick to set dictionary for each row\n",
    "common_resource_spec = {\n",
    "    \"cpu\": 32,\n",
    "    \"memory\": -1,\n",
    "}\n",
    "inputs[f\"{prefix}.resourceSpec\"] = inputs.iloc[:, 0].apply(lambda x: common_resource_spec)\n",
    "if prefix == \"Hashtag\":\n",
    "    inputs[f\"{prefix}.minCount\"] = 10\n",
    "\n",
    "# ********************\n",
    "\n",
    "# Annotate labels\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds[\"user\"], creds[\"password\"]))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_Path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CiteSeq.uriFastqR1</th>\n",
       "      <th>CiteSeq.uriFastqR2</th>\n",
       "      <th>CiteSeq.sampleName</th>\n",
       "      <th>CiteSeq.scRnaSeqPlatform</th>\n",
       "      <th>CiteSeq.lengthR1</th>\n",
       "      <th>CiteSeq.lengthR2</th>\n",
       "      <th>CiteSeq.cellBarcodeWhitelistUri</th>\n",
       "      <th>CiteSeq.cellBarcodeWhiteListMethod</th>\n",
       "      <th>CiteSeq.tagList</th>\n",
       "      <th>CiteSeq.cbStartPos</th>\n",
       "      <th>...</th>\n",
       "      <th>CiteSeq.umiStartPos</th>\n",
       "      <th>CiteSeq.umiEndPos</th>\n",
       "      <th>CiteSeq.trimPos</th>\n",
       "      <th>CiteSeq.slidingWindowSearch</th>\n",
       "      <th>CiteSeq.translate10XBarcodes</th>\n",
       "      <th>CiteSeq.cbCollapsingDistance</th>\n",
       "      <th>CiteSeq.umiCollapsingDistance</th>\n",
       "      <th>CiteSeq.numExpectedCells</th>\n",
       "      <th>CiteSeq.resourceSpec</th>\n",
       "      <th>CiteSeq.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_1</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_2</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_3</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_4</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_5</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_5</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_6</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_6</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_7</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_7</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_8</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_8</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_9</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_9</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_10</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_10</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_11</th>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/brow...</td>\n",
       "      <td>Patient_11</td>\n",
       "      <td>10x_v3</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>SeqcDenseCountsMatrixCsv</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'cpu': 32, 'memory': -1}</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 CiteSeq.uriFastqR1  \\\n",
       "Patient_1   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_2   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_3   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_4   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_5   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_6   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_7   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_8   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_9   [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_10  [s3://dp-lab-data/collaborators/brow...   \n",
       "Patient_11  [s3://dp-lab-data/collaborators/brow...   \n",
       "\n",
       "                                 CiteSeq.uriFastqR2 CiteSeq.sampleName  \\\n",
       "Patient_1   [s3://dp-lab-data/collaborators/brow...          Patient_1   \n",
       "Patient_2   [s3://dp-lab-data/collaborators/brow...          Patient_2   \n",
       "Patient_3   [s3://dp-lab-data/collaborators/brow...          Patient_3   \n",
       "Patient_4   [s3://dp-lab-data/collaborators/brow...          Patient_4   \n",
       "Patient_5   [s3://dp-lab-data/collaborators/brow...          Patient_5   \n",
       "Patient_6   [s3://dp-lab-data/collaborators/brow...          Patient_6   \n",
       "Patient_7   [s3://dp-lab-data/collaborators/brow...          Patient_7   \n",
       "Patient_8   [s3://dp-lab-data/collaborators/brow...          Patient_8   \n",
       "Patient_9   [s3://dp-lab-data/collaborators/brow...          Patient_9   \n",
       "Patient_10  [s3://dp-lab-data/collaborators/brow...         Patient_10   \n",
       "Patient_11  [s3://dp-lab-data/collaborators/brow...         Patient_11   \n",
       "\n",
       "           CiteSeq.scRnaSeqPlatform  CiteSeq.lengthR1  CiteSeq.lengthR2  \\\n",
       "Patient_1                    10x_v3                28                15   \n",
       "Patient_2                    10x_v3                28                15   \n",
       "Patient_3                    10x_v3                28                15   \n",
       "Patient_4                    10x_v3                28                15   \n",
       "Patient_5                    10x_v3                28                15   \n",
       "Patient_6                    10x_v3                28                15   \n",
       "Patient_7                    10x_v3                28                15   \n",
       "Patient_8                    10x_v3                28                15   \n",
       "Patient_9                    10x_v3                28                15   \n",
       "Patient_10                   10x_v3                28                15   \n",
       "Patient_11                   10x_v3                28                15   \n",
       "\n",
       "                    CiteSeq.cellBarcodeWhitelistUri  \\\n",
       "Patient_1   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_2   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_3   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_4   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_5   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_6   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_7   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_8   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_9   s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_10  s3://dp-lab-data/collaborators/brown...   \n",
       "Patient_11  s3://dp-lab-data/collaborators/brown...   \n",
       "\n",
       "           CiteSeq.cellBarcodeWhiteListMethod  \\\n",
       "Patient_1            SeqcDenseCountsMatrixCsv   \n",
       "Patient_2            SeqcDenseCountsMatrixCsv   \n",
       "Patient_3            SeqcDenseCountsMatrixCsv   \n",
       "Patient_4            SeqcDenseCountsMatrixCsv   \n",
       "Patient_5            SeqcDenseCountsMatrixCsv   \n",
       "Patient_6            SeqcDenseCountsMatrixCsv   \n",
       "Patient_7            SeqcDenseCountsMatrixCsv   \n",
       "Patient_8            SeqcDenseCountsMatrixCsv   \n",
       "Patient_9            SeqcDenseCountsMatrixCsv   \n",
       "Patient_10           SeqcDenseCountsMatrixCsv   \n",
       "Patient_11           SeqcDenseCountsMatrixCsv   \n",
       "\n",
       "                                    CiteSeq.tagList  CiteSeq.cbStartPos  ...  \\\n",
       "Patient_1   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_2   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_3   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_4   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_5   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_6   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_7   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_8   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_9   s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_10  s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "Patient_11  s3://dp-lab-data/collaborators/brown...                   1  ...   \n",
       "\n",
       "            CiteSeq.umiStartPos  CiteSeq.umiEndPos  CiteSeq.trimPos  \\\n",
       "Patient_1                    17                 28                0   \n",
       "Patient_2                    17                 28                0   \n",
       "Patient_3                    17                 28                0   \n",
       "Patient_4                    17                 28                0   \n",
       "Patient_5                    17                 28                0   \n",
       "Patient_6                    17                 28                0   \n",
       "Patient_7                    17                 28                0   \n",
       "Patient_8                    17                 28                0   \n",
       "Patient_9                    17                 28                0   \n",
       "Patient_10                   17                 28                0   \n",
       "Patient_11                   17                 28                0   \n",
       "\n",
       "            CiteSeq.slidingWindowSearch  CiteSeq.translate10XBarcodes  \\\n",
       "Patient_1                         False                         False   \n",
       "Patient_2                         False                         False   \n",
       "Patient_3                         False                         False   \n",
       "Patient_4                         False                         False   \n",
       "Patient_5                         False                         False   \n",
       "Patient_6                         False                         False   \n",
       "Patient_7                         False                         False   \n",
       "Patient_8                         False                         False   \n",
       "Patient_9                         False                         False   \n",
       "Patient_10                        False                         False   \n",
       "Patient_11                        False                         False   \n",
       "\n",
       "            CiteSeq.cbCollapsingDistance  CiteSeq.umiCollapsingDistance  \\\n",
       "Patient_1                              1                              1   \n",
       "Patient_2                              1                              1   \n",
       "Patient_3                              1                              1   \n",
       "Patient_4                              1                              1   \n",
       "Patient_5                              1                              1   \n",
       "Patient_6                              1                              1   \n",
       "Patient_7                              1                              1   \n",
       "Patient_8                              1                              1   \n",
       "Patient_9                              1                              1   \n",
       "Patient_10                             1                              1   \n",
       "Patient_11                             1                              1   \n",
       "\n",
       "            CiteSeq.numExpectedCells       CiteSeq.resourceSpec  \\\n",
       "Patient_1                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_2                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_3                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_4                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_5                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_6                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_7                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_8                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_9                          0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_10                         0  {'cpu': 32, 'memory': -1}   \n",
       "Patient_11                         0  {'cpu': 32, 'memory': -1}   \n",
       "\n",
       "           CiteSeq.dockerRegistry  \n",
       "Patient_1         quay.io/hisplan  \n",
       "Patient_2         quay.io/hisplan  \n",
       "Patient_3         quay.io/hisplan  \n",
       "Patient_4         quay.io/hisplan  \n",
       "Patient_5         quay.io/hisplan  \n",
       "Patient_6         quay.io/hisplan  \n",
       "Patient_7         quay.io/hisplan  \n",
       "Patient_8         quay.io/hisplan  \n",
       "Patient_9         quay.io/hisplan  \n",
       "Patient_10        quay.io/hisplan  \n",
       "Patient_11        quay.io/hisplan  \n",
       "\n",
       "[11 rows x 21 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_1</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_2</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_3</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_4</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_5</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_5</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_6</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_6</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_7</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_7</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_8</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_8</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_9</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_9</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_10</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_10</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_11</th>\n",
       "      <td>CITE-seq</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Patient_11</td>\n",
       "      <td>moormana</td>\n",
       "      <td>s3://dp-lab-data/collaborators/brown...</td>\n",
       "      <td>-</td>\n",
       "      <td>moormana</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pipelineType   project      sample     owner  \\\n",
       "Patient_1      CITE-seq  COVID-19   Patient_1  moormana   \n",
       "Patient_2      CITE-seq  COVID-19   Patient_2  moormana   \n",
       "Patient_3      CITE-seq  COVID-19   Patient_3  moormana   \n",
       "Patient_4      CITE-seq  COVID-19   Patient_4  moormana   \n",
       "Patient_5      CITE-seq  COVID-19   Patient_5  moormana   \n",
       "Patient_6      CITE-seq  COVID-19   Patient_6  moormana   \n",
       "Patient_7      CITE-seq  COVID-19   Patient_7  moormana   \n",
       "Patient_8      CITE-seq  COVID-19   Patient_8  moormana   \n",
       "Patient_9      CITE-seq  COVID-19   Patient_9  moormana   \n",
       "Patient_10     CITE-seq  COVID-19  Patient_10  moormana   \n",
       "Patient_11     CITE-seq  COVID-19  Patient_11  moormana   \n",
       "\n",
       "                                        destination transfer   comment  \n",
       "Patient_1   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_2   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_3   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_4   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_5   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_6   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_7   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_8   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_9   s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_10  s3://dp-lab-data/collaborators/brown...        -  moormana  \n",
       "Patient_11  s3://dp-lab-data/collaborators/brown...        -  moormana  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "473536c800f34b96bd48c7818c3a9875",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/11 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(samples.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}_{prefix}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}_{prefix}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = path_to_exec,\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_1_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_1_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"96c593ea-639f-4be5-86f4-8ef0832ee733\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_2_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_2_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"dceb37ea-6b08-4cd3-8056-5a624f5b7d6b\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_3_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_3_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"f011abec-1dc7-45ae-8393-005a6a05bb64\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_4_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_4_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"32a24faa-f3a6-40d1-a9e5-62032f474d34\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_5_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_5_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"07d5932b-081a-484f-a87b-2f17ac1b2117\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_6_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_6_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"ec9cfd36-ea2c-472a-a1e1-39915c01087b\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_7_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_7_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"a9e844dd-8477-4f32-a478-c06aa04505b0\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_8_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_8_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"6298f21d-49d6-46be-b3e4-48089ab12d1f\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_9_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_9_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"892c5e59-e851-4f43-93e8-5d0b3d6137f6\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_10_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_10_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"4a1f6873-3eae-4ac5-90a4-cef9be298dae\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/moormana/scing/bin/sharp-0.0.13/submit-citeseq.sh',\n",
       "   '-k',\n",
       "   '/Users/moormana/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_11_CiteSeq.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/configs/Patient_11_CiteSeq.labels.json',\n",
       "   '-o',\n",
       "   '/Users/moormana/scing/bin/sharp-0.0.13/Sharp.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"37ca9552-0887-4299-991f-42e86a3cd824\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
