{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e799a381-2d5f-4497-bbc4-99df4f1c0370",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f82a05bc-cc8a-4d55-b7fe-9513f475077c",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Define Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c8f9ae8c-dc67-4a8e-a3f7-dc21eddb54e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "fastq_map = {\n",
    "    'CellRangerVdj': ['I1','R1','R2'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c8ab0226-5393-411a-88c0-3f8b154b6c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_fastqs(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "    fastq_file_ids: list, # FASTQ file ids needed for this run type (e.g. I1, R1, R2, etc.)\n",
    "    folder: str = \"\",\n",
    "):\n",
    "    fastq_map = dict()\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(f\"{path}/{folder}\")\n",
    "    for fid in fastq_file_ids:\n",
    "        files = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\"_{fid}_\\d{{3}}.fastq.gz$\")\n",
    "        )\n",
    "        try:\n",
    "            assert files, f\"AssertionError: Missing `{fid}` archives!\"\n",
    "            fastq_map[fid] = [os.path.join(\"s3://\", bucket, str(f)) for f in files]\n",
    "        except AssertionError as err:\n",
    "            logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "            return\n",
    "    return fastq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e555cf05-fde7-4875-b4c5-46436c9a42ec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from SCRIdb\n",
    "def get_s3_objects(bucket, key, pattern, full_uri=False):\n",
    "    \n",
    "    s3r = boto3.resource(\"s3\")\n",
    "    bucket_s3 = s3r.Bucket(bucket)\n",
    "    objects = []\n",
    "    for obj in bucket_s3.objects.filter(Prefix=key):\n",
    "        hit = pattern.search(obj.key)\n",
    "        if hit:\n",
    "            objects.append(obj.key)\n",
    "    if full_uri:\n",
    "        objects = [f\"s3://{bucket}/{o}\" for o in objects]\n",
    "    return objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "18e8874c-f137-4afb-8b40-68fe9fb896d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract FASTQ sample name from list of files\n",
    "# Note: FASTQ name is file name up to lane id (e.g. L001, L002, etc.)\n",
    "def get_fastqs_name(fastqs):\n",
    "    fastq_name_re = r\".*/(.*)_S\\d+_L\\d{3}_[A-Za-z]\\d_\\d{3}.fastq.gz$\"\n",
    "    fastq_names = [re.match(fastq_name_re, x)[1] for x in fastqs]\n",
    "    assert len(set(fastq_names)) == 1 # make sure all names are same\n",
    "    return fastq_names[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e8688b7-ec81-4414-88dc-34d91ecb411a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Numpy encoder for JSON from pandas series\n",
    "class NpEncoder(json.JSONEncoder):\n",
    "    def default(self, obj):\n",
    "        if isinstance(obj, np.integer):\n",
    "            return int(obj)\n",
    "        elif isinstance(obj, np.floating):\n",
    "            return float(obj)\n",
    "        elif isinstance(obj, np.ndarray):\n",
    "            return obj.tolist()\n",
    "        else:\n",
    "            return super(NpEncoder, self).default(obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "33af155b-7eab-4a73-ae3f-e818a412a316",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def execute_query(query, user, password):\n",
    "    with connect(\n",
    "        host=\"peer-lab-db.cggxmlwgzzpw.us-east-1.rds.amazonaws.com\",\n",
    "        database=\"peer_lab_db\",\n",
    "        user=user,\n",
    "        password=password,\n",
    "    ) as connection:\n",
    "        with connection.cursor(buffered=True) as cursor:\n",
    "            cursor.execute(query)\n",
    "            result = cursor.fetchall()\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3040e204-9470-465a-863d-6f4a297a1637",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_species(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_species = \"peer_lab_db.species\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_species}.Species\n",
    "        FROM {table_species}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_species}.id = {table_genome_idx}.species_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f214cb46-2889-483b-bb00-ed65b9b96125",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sc_tech(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_sc_tech = \"peer_lab_db.sc_tech\"\n",
    "        table_genome_idx = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sc_tech}.sc_Tech\n",
    "        FROM {table_sc_tech}\n",
    "        LEFT JOIN {table_genome_idx}\n",
    "        ON {table_sc_tech}.id = {table_genome_idx}.scTech_id\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_genome_idx}.id = {table_sample_data}.genomeIndex_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9be132ae-a9c7-46fd-bb1c-04358f4b7d34",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_sample_id(sample_name, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.Sample=\"{sample_name}\"\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ef099bfd-ee43-4dce-bc55-3045e9f4e9b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get species from database for given sample\n",
    "from mysql.connector import connect, Error\n",
    "\n",
    "def get_project_id(sample_id, user, password):\n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "865f4379-343e-4010-9d76-eb7357f9beae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_SEQC_version(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 cp {loc}/seqc-results/seqc_log.txt -\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        version = re.match(r\".*SEQC=v(\\d+\\.\\d+\\.\\d+).*\", out)[1]\n",
    "        return version\n",
    "    except:\n",
    "        return \"N/A\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0511ae48-826d-463d-b3a3-b080b6fd2d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_file_prefix(loc):\n",
    "    try:\n",
    "        cmd = f\"aws s3 ls {loc}/seqc-results/\"\n",
    "        out = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True).__dict__[\"stdout\"]\n",
    "        \n",
    "        # Note: I'm expecting the aligned bam file to be in loc\n",
    "        bam_pattern = re.compile(r\"(.*)_Aligned\\.out\\.bam$\")\n",
    "        filename = list(filter(bam_pattern.match, out.split()))[0]\n",
    "        file_prefix = re.match(bam_pattern, filename)[1]\n",
    "        return file_prefix\n",
    "    except:\n",
    "        raise ValueError(f\"BAM file not found in {loc}\")\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0c0c8d18-97f8-4ca9-8de8-7d5f0d837b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTQ reads/indices required for each workflow\n",
    "# Shoudl replace with JSON file\n",
    "cr_reference_map = {\n",
    "    'CellRangerArc':\n",
    "    {\n",
    "        'Human': \"https://cf.10xgenomics.com/supp/cell-arc/refdata-cellranger-arc-GRCh38-2020-A.tar.gz\",\n",
    "        'Mouse': \"https://cf.10xgenomics.com/supp/cell-arc/refdata-cellranger-arc-mm10-2020-A-2.0.0.tar.gz\",\n",
    "    },\n",
    "    'CellRangerATAC':\n",
    "    {\n",
    "        'Human': \"https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-arc-GRCh38-2020-A-2.0.0.tar.gz\",\n",
    "        'Mouse': \"https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-arc-mm10-2020-A-2.0.0.tar.gz\",\n",
    "    },\n",
    "    'CellRangerGex':\n",
    "    {\n",
    "        'Human': \"https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz\",\n",
    "        'Mouse': \"https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-mm10-2020-A.tar.gz\",\n",
    "    },\n",
    "    'CellRangerVdj':\n",
    "    {\n",
    "        'Human': \"GRCh38\"\n",
    "    },\n",
    "}\n",
    "\n",
    "def get_cr_reference(sample_id, prefix, user, password):\n",
    "    # Get species from database to decide reference\n",
    "    species = get_species(sample_id, user, password)\n",
    "    \n",
    "    # Map to reference locations\n",
    "    try:\n",
    "        return cr_reference_map[prefix][species]\n",
    "    except:\n",
    "        raise ValueError(f\"Unknown Species: {species}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bf3f9f33-b976-4d11-897b-79199a7e1068",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bc_whitelist(sample_id):\n",
    "    # Get version from database to decide whitelist\n",
    "    sc_tech = get_sc_tech(sample_id, creds[\"user\"], creds[\"password\"])\n",
    "    \n",
    "    # Map to reference locations\n",
    "    if \"V3\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v3/flat/3M-february-2018.txt\"\n",
    "    elif \"V2\" in sc_tech:\n",
    "        return \"s3://seqc-public/barcodes/ten_x_v2/flat/737K-august-2016.txt\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown Technology: {sc_tech}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c53b08e4-73fb-4fc6-85f6-cb7e6292d000",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(\n",
    "    workflow_path: str,\n",
    "    execp: str,\n",
    "    secrets: str,\n",
    "    inputs: str,\n",
    "    labels: str,\n",
    "    options: str,\n",
    "):\n",
    "    # change working directory to the pipeline package\n",
    "    oldwd = os.getcwd()\n",
    "    os.chdir(workflow_path)\n",
    "    \n",
    "    # execute the pipeline command\n",
    "    cmd = f\"{workflow_path}/{execp} -k {secrets} -i {inputs} -l {labels} -o {options}\"\n",
    "    var = subprocess.run(shlex.split(cmd), universal_newlines=True, capture_output=True)\n",
    "    out = var.__dict__\n",
    "    \n",
    "    # change working directory back\n",
    "    os.chdir(oldwd)\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80414572-8a8e-44a8-ada1-3b636086da98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3c47d694-4856-42e3-8458-68708048ca30",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Additional functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b389052-37ff-49b0-90e7-473c3725f0d6",
   "metadata": {
    "tags": []
   },
   "source": [
    "### DB query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "68bd2fa1-53e2-41aa-8f8e-da549db23f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "def get_sample_name(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Sample\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_names = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_names.append(result[0])\n",
    "        return sample_names\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "    \n",
    "    \n",
    "def get_aws_path(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.AWS_storage\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_paths = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_paths.append(result[0])\n",
    "        return sample_paths\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "            \n",
    "def get_sample_id(query, query_col, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        sample_ids = []\n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            sample_ids.append(result[0])\n",
    "        return sample_ids\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def format_sample_aws(querys, query_col, creds):\n",
    "    sample_names = []\n",
    "    sample_paths = []\n",
    "    sample_ids = []\n",
    "    \n",
    "    for query in querys:\n",
    "        sample_names += get_sample_name(query, query_col, creds)\n",
    "        sample_paths += get_aws_path(query, query_col, creds)\n",
    "        sample_ids += get_sample_id(query, query_col, creds)\n",
    "        \n",
    "    sample_paths = [s.strip('/') for s in sample_paths] # remove trailing slash if exists\n",
    "    \n",
    "    samples = pd.DataFrame(\n",
    "        dict(S3_Path=sample_paths, Sample_ID=sample_ids),\n",
    "        index=sample_names,\n",
    "        dtype=str,\n",
    "    )\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fa96390b-c1e5-48f6-af5b-6eed769f5020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomeIndex_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.genomeIndex_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            genomeIndex_id = result[0]\n",
    "            \n",
    "        return genomeIndex_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_scTech_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.scTech_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            scTech_id = result[0]\n",
    "            \n",
    "        return scTech_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_index(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.gIndex\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            index = result[0]\n",
    "            \n",
    "        return index\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_assay(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Run_name\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            assay = result[0]\n",
    "            \n",
    "        return assay\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_barcode(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sc_tech\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.barcodes\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            barcode = result[0]\n",
    "            \n",
    "        return barcode\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_species_id(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.genome_index\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.species_id\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            species_id = result[0]\n",
    "            \n",
    "        return species_id\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        \n",
    "def get_species(query, query_col, creds):\n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.species\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_sample_data}.Species\n",
    "        FROM {table_sample_data}\n",
    "        WHERE {table_sample_data}.{query_col}=\"{query}\"\n",
    "        \"\"\"\n",
    "        \n",
    "        results = execute_query(query, user, password)\n",
    "        for result in results:\n",
    "            barcode = result[0]\n",
    "            \n",
    "        return barcode\n",
    "            \n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")\n",
    "\n",
    "        \n",
    "def format_assay_barcode(samples, creds):\n",
    "    \n",
    "    # samples['genomeIndex_id'] = np.nan\n",
    "    samples['Reference'] = np.nan\n",
    "    # samples['scTech_id'] = np.nan\n",
    "    samples['assay'] = np.nan\n",
    "    # samples['barcode'] = np.nan\n",
    "    samples['species'] = np.nan\n",
    "    \n",
    "    for sample, row in samples.iterrows():\n",
    "        sample_id = row['Sample_ID']\n",
    "        \n",
    "        genomeIndex_id  = get_genomeIndex_id(sample_id, 'id', creds)\n",
    "        index = get_index(genomeIndex_id, 'id', creds)\n",
    "        scTech_id = get_scTech_id(genomeIndex_id, 'id', creds)\n",
    "        assay = get_assay(scTech_id, 'id', creds)\n",
    "        barcode = get_barcode(scTech_id, 'id', creds)\n",
    "        species_id = get_species_id(genomeIndex_id, 'id', creds)\n",
    "        species = get_species(species_id, 'id', creds)\n",
    "        \n",
    "\n",
    "        samples.loc[sample, 'species'] = species\n",
    "        # samples.loc[sample, 'genomeIndex_id'] = genomeIndex_id\n",
    "        if index:\n",
    "            samples.loc[sample, 'Reference'] = index.strip()\n",
    "        else:\n",
    "            samples.loc[sample, 'Reference'] = cr_reference_map[prefix][species]\n",
    "        # samples.loc[sample, 'scTech_id'] = scTech_id\n",
    "        if assay:\n",
    "            samples.loc[sample, 'assay'] = assay.strip()\n",
    "        # samples.loc[sample, 'barcode'] = barcode.strip()\n",
    "        \n",
    "\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9d4a59dc-fb1a-4e84-8fc9-95238eb22291",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_project_id(sample_id, creds):\n",
    "    \n",
    "    user = creds['user']\n",
    "    password = creds['password']\n",
    "    \n",
    "    try:\n",
    "        table_sample_data = \"peer_lab_db.sample_data\"\n",
    "        table_project_data = \"peer_lab_db.project_data\"\n",
    "        query = f\"\"\"\n",
    "        SELECT {table_project_data}.projectName\n",
    "        FROM {table_project_data}\n",
    "        LEFT JOIN {table_sample_data}\n",
    "        ON {table_project_data}.id = {table_sample_data}.projectData_id\n",
    "        WHERE {table_sample_data}.id = {sample_id}\n",
    "        \"\"\"\n",
    "        result = execute_query(query, user, password)[0][0]\n",
    "        return result\n",
    "    except Error as e:\n",
    "        print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01a8da79-682e-4ba6-861d-b9e4b54af6fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get tar gz file for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "539e3ea7-b40a-4a70-a840-ffbaf32baab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get fastq file paths on S3 for each file id\n",
    "# Returns dictionary from id to s3 path\n",
    "# Throws exception if FASTQs don't exist for any id\n",
    "def get_reference(\n",
    "    path: str, # path to directory containing FASTQ files\n",
    "):\n",
    "    print(path)\n",
    "    _, bucket, key, _, _ = urllib.parse.urlsplit(path)\n",
    "    reference = get_s3_objects(\n",
    "            bucket, key.lstrip(\"/\"),\n",
    "            re.compile(f\".tar.gz$\")\n",
    "        )\n",
    "    try:\n",
    "        reference = reference[0]\n",
    "        reference = os.path.join(\"s3://\", bucket, reference)\n",
    "    except AssertionError as err:\n",
    "        logging.warning(\"%s\\n\\t %s\", err, path)\n",
    "        return\n",
    "    return reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98cd9e7f-2171-4710-b329-b4812aeafd67",
   "metadata": {},
   "source": [
    "# Process Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28995755-3d1b-45f9-8fac-0c13bac4f9c9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9db6a119-0ad5-4965-b988-9c85de2591b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "\n",
    "prefix = \"CellRangerVdj\" # Workflow to run; also .wdl filename prefix\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "output_dirname = \"cr-vdj-results\"\n",
    "\n",
    "# If need to add comment, put here\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a067d6c1-49da-4901-ba59-bd21c5a69871",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\" # CHANGE THIS\n",
    "workflow_dir = f\"{Path.home()}/scing/bin/cellranger-vdj-6.1.2\" # CHANGE THIS\n",
    "path_to_exec = f\"{workflow_dir}/submit.sh\" # CHANGE THIS FOR SHARP\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = f\"{workflow_dir}/{prefix}.options.aws.json\"\n",
    "\n",
    "# Other file locations\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\" # CHANGE THIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d3fa829f-dfab-4f1a-b943-1f8ef0106985",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e2aca850-d8df-4247-a961-9adc5dffa741",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: AWS_ACCESS_KEY_ID=ASIAYPY66CWMP62NDHHJ\n",
      "env: AWS_SECRET_ACCESS_KEY=+vN4OePqnFaMmtmn8j+vX07vT/ytL6KXegHoWtXM\n",
      "env: AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEBkaCXVzLWVhc3QtMSJHMEUCIQDnicGk6x51cqXMOOaoToeNBclj4YkWARak9yD1JUOdOwIgZVm9Bf97AxeXMiBYTta58H2UEK3hjr7HxQcG9VhgjQ4q7wEIIhADGgw1ODM2NDM1Njc1MTIiDKEqQMj0zJvIE5N8yirMAbmsr69d2hSRrXUDgOoQSs6uR71H4B4L/+IGd7k07ajoW0QlInUZEvWqwqEiCr41zB0b4+2ibcm5QdKAqZxowl/QRGfNltScM2RsGXAlIVzvAfHAGHU3lVOi9KYd/0MRLBDqyghKojzUvrxnHHF0ZuwuUTFcS/Bl36xjYfSuq78jgebZ8CFqQ9ii5/Z3af8eiFjpxSZIX7fRetUhwCk3/2E6SH38kX9Ijr6qufVNqugWVxQtNrqnYh3aRvOebt7o6n3hnyj+jjol+l74yzCMqIqVBjqYAeIl6o8n8l/YvF5E0ftbBoxB9+sxyOFpIuPGVA9xrtmTsiKVIT0asvr/IJ4t/isujGKQVVxSrIzINYuHUcs6Vtkf958uDeJwSSg7o2QsXV2fmCbWCM8zmRqYuOcTVwXEDrz9hQ0r3rdj1QJMjLNzh4LhN2uRS2rLuUNZ+g3Qgplw0IfiiKROEovxz8TnXt3SlHugXn0brEBA\n",
      "2021-10-07 15:31:32 agc-583643567512-us-east-1\n",
      "2021-10-07 15:28:07 cdktoolkit-stagingbucket-d49u1xfb0sc4\n",
      "2019-10-09 13:04:06 cf-templates-umiwbnq3566w-us-east-1\n",
      "2019-10-10 12:46:54 dp-daily-reports\n",
      "2017-09-26 10:15:15 dp-lab-data\n",
      "2019-05-23 12:34:58 dp-lab-data-public\n",
      "2018-03-01 13:29:07 dp-lab-glacier\n",
      "2020-11-06 23:02:27 dp-lab-gwf-core\n",
      "2021-07-15 18:14:01 dp-lab-gwf-core2\n",
      "2017-09-26 10:17:18 dp-lab-home\n",
      "2019-08-14 16:42:43 dp-lab-test\n",
      "2019-04-25 12:35:35 elasticbeanstalk-us-east-1-583643567512\n",
      "2019-12-13 15:53:57 scri-computational\n",
      "2017-09-26 10:23:50 seqc-public\n"
     ]
    }
   ],
   "source": [
    "%env AWS_ACCESS_KEY_ID=ASIAYPY66CWMP62NDHHJ\n",
    "%env AWS_SECRET_ACCESS_KEY=+vN4OePqnFaMmtmn8j+vX07vT/ytL6KXegHoWtXM\n",
    "%env AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEBkaCXVzLWVhc3QtMSJHMEUCIQDnicGk6x51cqXMOOaoToeNBclj4YkWARak9yD1JUOdOwIgZVm9Bf97AxeXMiBYTta58H2UEK3hjr7HxQcG9VhgjQ4q7wEIIhADGgw1ODM2NDM1Njc1MTIiDKEqQMj0zJvIE5N8yirMAbmsr69d2hSRrXUDgOoQSs6uR71H4B4L/+IGd7k07ajoW0QlInUZEvWqwqEiCr41zB0b4+2ibcm5QdKAqZxowl/QRGfNltScM2RsGXAlIVzvAfHAGHU3lVOi9KYd/0MRLBDqyghKojzUvrxnHHF0ZuwuUTFcS/Bl36xjYfSuq78jgebZ8CFqQ9ii5/Z3af8eiFjpxSZIX7fRetUhwCk3/2E6SH38kX9Ijr6qufVNqugWVxQtNrqnYh3aRvOebt7o6n3hnyj+jjol+l74yzCMqIqVBjqYAeIl6o8n8l/YvF5E0ftbBoxB9+sxyOFpIuPGVA9xrtmTsiKVIT0asvr/IJ4t/isujGKQVVxSrIzINYuHUcs6Vtkf958uDeJwSSg7o2QsXV2fmCbWCM8zmRqYuOcTVwXEDrz9hQ0r3rdj1QJMjLNzh4LhN2uRS2rLuUNZ+g3Qgplw0IfiiKROEovxz8TnXt3SlHugXn0brEBA\n",
    "!aws s3 ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc47d702-f319-42bf-8434-68f7d662159b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "228b2db2-d5d5-4233-8ade-8ad84f459c40",
   "metadata": {},
   "source": [
    "## Get sample Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8e3c30d9-273e-4ffe-82a4-e23adb3d1ac0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               S3_Path  \\\n",
       "SD-1680_Patient_D_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "\n",
       "                              Sample_ID  \n",
       "SD-1680_Patient_D_nucseq_GE        3808  \n",
       "SD-1680_Patient_F_nucseq_GE        3809  \n",
       "SD-1680_Patient_I_nucseq_GE        3810  \n",
       "SD-1680_Patient_D_nucseq_2_GE      3811  \n",
       "SD-1680_Patient_F_nucseq_2_GE      3812  \n",
       "SD-1680_Patient_I_nucseq_2_GE      3813  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Common query col: id, request_id, Sample\n",
    "sample_ids = list(range(3808, 3814))\n",
    "samples = format_sample_aws(sample_ids, 'id', creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "07074e44-a339-468d-81a2-d10005209528",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>assay</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3808</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3809</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3810</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3811</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3812</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3813</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               S3_Path  \\\n",
       "SD-1680_Patient_D_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "\n",
       "                              Sample_ID Reference       assay species  \n",
       "SD-1680_Patient_D_nucseq_GE        3808    GRCh38  five_prime   Human  \n",
       "SD-1680_Patient_F_nucseq_GE        3809    GRCh38  five_prime   Human  \n",
       "SD-1680_Patient_I_nucseq_GE        3810    GRCh38  five_prime   Human  \n",
       "SD-1680_Patient_D_nucseq_2_GE      3811    GRCh38  five_prime   Human  \n",
       "SD-1680_Patient_F_nucseq_2_GE      3812    GRCh38  five_prime   Human  \n",
       "SD-1680_Patient_I_nucseq_2_GE      3813    GRCh38  five_prime   Human  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = format_assay_barcode(samples, creds)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e2a13aa2-8568-44e8-8316-615b22e4ee2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S3_Path</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>assay</th>\n",
       "      <th>species</th>\n",
       "      <th>FASTQs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3808</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3809</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3810</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3811</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3812</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>3813</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>five_prime</td>\n",
       "      <td>Human</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               S3_Path  \\\n",
       "SD-1680_Patient_D_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "\n",
       "                              Sample_ID Reference       assay species  \\\n",
       "SD-1680_Patient_D_nucseq_GE        3808    GRCh38  five_prime   Human   \n",
       "SD-1680_Patient_F_nucseq_GE        3809    GRCh38  five_prime   Human   \n",
       "SD-1680_Patient_I_nucseq_GE        3810    GRCh38  five_prime   Human   \n",
       "SD-1680_Patient_D_nucseq_2_GE      3811    GRCh38  five_prime   Human   \n",
       "SD-1680_Patient_F_nucseq_2_GE      3812    GRCh38  five_prime   Human   \n",
       "SD-1680_Patient_I_nucseq_2_GE      3813    GRCh38  five_prime   Human   \n",
       "\n",
       "                                                                FASTQs  \n",
       "SD-1680_Patient_D_nucseq_GE    {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "SD-1680_Patient_F_nucseq_GE    {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "SD-1680_Patient_I_nucseq_GE    {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "SD-1680_Patient_D_nucseq_2_GE  {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "SD-1680_Patient_F_nucseq_2_GE  {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "SD-1680_Patient_I_nucseq_2_GE  {'I1': ['s3://dp-lab-data/collaborat...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get FASTQ paths from S3\n",
    "# Note: Uses same FASTQ file ids for all samples\n",
    "fastq_file_ids = fastq_map[prefix]\n",
    "samples[\"FASTQs\"] = samples[\"S3_Path\"].apply(lambda x: get_fastqs(x, fastq_file_ids, \"FASTQ\"))\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db07203d-2a0f-4b0d-a49e-564d0aca90bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d6be9b75-456e-49dd-bd33-a2766e84cc7c",
   "metadata": {},
   "source": [
    "## Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9ea75c1e-9110-47e1-a66d-09ff9a80b8b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellRangerVdj.sampleName</th>\n",
       "      <th>CellRangerVdj.fastqNames</th>\n",
       "      <th>CellRangerVdj.referenceGenome</th>\n",
       "      <th>CellRangerVdj.inputFastq</th>\n",
       "      <th>CellRangerVdj.chain</th>\n",
       "      <th>CellRangerVdj.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              CellRangerVdj.sampleName  \\\n",
       "SD-1680_Patient_D_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_F_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_I_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_D_nucseq_2_GE                      NaN   \n",
       "SD-1680_Patient_F_nucseq_2_GE                      NaN   \n",
       "SD-1680_Patient_I_nucseq_2_GE                      NaN   \n",
       "\n",
       "                              CellRangerVdj.fastqNames  \\\n",
       "SD-1680_Patient_D_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_F_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_I_nucseq_GE                        NaN   \n",
       "SD-1680_Patient_D_nucseq_2_GE                      NaN   \n",
       "SD-1680_Patient_F_nucseq_2_GE                      NaN   \n",
       "SD-1680_Patient_I_nucseq_2_GE                      NaN   \n",
       "\n",
       "                              CellRangerVdj.referenceGenome  \\\n",
       "SD-1680_Patient_D_nucseq_GE                             NaN   \n",
       "SD-1680_Patient_F_nucseq_GE                             NaN   \n",
       "SD-1680_Patient_I_nucseq_GE                             NaN   \n",
       "SD-1680_Patient_D_nucseq_2_GE                           NaN   \n",
       "SD-1680_Patient_F_nucseq_2_GE                           NaN   \n",
       "SD-1680_Patient_I_nucseq_2_GE                           NaN   \n",
       "\n",
       "                              CellRangerVdj.inputFastq CellRangerVdj.chain  \\\n",
       "SD-1680_Patient_D_nucseq_GE                        NaN                 NaN   \n",
       "SD-1680_Patient_F_nucseq_GE                        NaN                 NaN   \n",
       "SD-1680_Patient_I_nucseq_GE                        NaN                 NaN   \n",
       "SD-1680_Patient_D_nucseq_2_GE                      NaN                 NaN   \n",
       "SD-1680_Patient_F_nucseq_2_GE                      NaN                 NaN   \n",
       "SD-1680_Patient_I_nucseq_2_GE                      NaN                 NaN   \n",
       "\n",
       "                              CellRangerVdj.dockerRegistry  \n",
       "SD-1680_Patient_D_nucseq_GE                            NaN  \n",
       "SD-1680_Patient_F_nucseq_GE                            NaN  \n",
       "SD-1680_Patient_I_nucseq_GE                            NaN  \n",
       "SD-1680_Patient_D_nucseq_2_GE                          NaN  \n",
       "SD-1680_Patient_F_nucseq_2_GE                          NaN  \n",
       "SD-1680_Patient_I_nucseq_2_GE                          NaN  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load minimum inputs and labels fields from templates\n",
    "with open(f\"{config_dir}/template.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6bdf4037-df51-4f71-82a5-3187f7266f76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellRangerVdj.sampleName</th>\n",
       "      <th>CellRangerVdj.fastqNames</th>\n",
       "      <th>CellRangerVdj.referenceGenome</th>\n",
       "      <th>CellRangerVdj.inputFastq</th>\n",
       "      <th>CellRangerVdj.chain</th>\n",
       "      <th>CellRangerVdj.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_D_nucseq_GE</td>\n",
       "      <td>3808_SD-1680_Patient_D_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_F_nucseq_GE</td>\n",
       "      <td>3809_SD-1680_Patient_F_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_I_nucseq_GE</td>\n",
       "      <td>3810_SD-1680_Patient_I_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_D_nucseq_2_GE</td>\n",
       "      <td>3811_SD-1680_Patient_D_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_F_nucseq_2_GE</td>\n",
       "      <td>3812_SD-1680_Patient_F_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_I_nucseq_2_GE</td>\n",
       "      <td>3813_SD-1680_Patient_I_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    CellRangerVdj.sampleName  \\\n",
       "SD-1680_Patient_D_nucseq_GE      SD-1680_Patient_D_nucseq_GE   \n",
       "SD-1680_Patient_F_nucseq_GE      SD-1680_Patient_F_nucseq_GE   \n",
       "SD-1680_Patient_I_nucseq_GE      SD-1680_Patient_I_nucseq_GE   \n",
       "SD-1680_Patient_D_nucseq_2_GE  SD-1680_Patient_D_nucseq_2_GE   \n",
       "SD-1680_Patient_F_nucseq_2_GE  SD-1680_Patient_F_nucseq_2_GE   \n",
       "SD-1680_Patient_I_nucseq_2_GE  SD-1680_Patient_I_nucseq_2_GE   \n",
       "\n",
       "                                              CellRangerVdj.fastqNames  \\\n",
       "SD-1680_Patient_D_nucseq_GE    3808_SD-1680_Patient_D_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_F_nucseq_GE    3809_SD-1680_Patient_F_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_I_nucseq_GE    3810_SD-1680_Patient_I_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  3811_SD-1680_Patient_D_nucseq_2_GE_I...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  3812_SD-1680_Patient_F_nucseq_2_GE_I...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  3813_SD-1680_Patient_I_nucseq_2_GE_I...   \n",
       "\n",
       "                              CellRangerVdj.referenceGenome  \\\n",
       "SD-1680_Patient_D_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_F_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_I_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_D_nucseq_2_GE                        GRCh38   \n",
       "SD-1680_Patient_F_nucseq_2_GE                        GRCh38   \n",
       "SD-1680_Patient_I_nucseq_2_GE                        GRCh38   \n",
       "\n",
       "                                              CellRangerVdj.inputFastq  \\\n",
       "SD-1680_Patient_D_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_F_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_I_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "\n",
       "                              CellRangerVdj.chain CellRangerVdj.dockerRegistry  \n",
       "SD-1680_Patient_D_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_F_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_I_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_D_nucseq_2_GE                auto              quay.io/hisplan  \n",
       "SD-1680_Patient_F_nucseq_2_GE                auto              quay.io/hisplan  \n",
       "SD-1680_Patient_I_nucseq_2_GE                auto              quay.io/hisplan  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs[f'{prefix}.sampleName'] = samples.index\n",
    "inputs[f'{prefix}.inputFastq'] = samples[\"FASTQs\"].apply(lambda x: np.ravel(list(x.values())))\n",
    "inputs[f'{prefix}.fastqNames'] = inputs[f\"{prefix}.inputFastq\"].apply(lambda x: get_fastqs_name(x))\n",
    "inputs[f'{prefix}.referenceGenome'] = samples['Reference']\n",
    "inputs[f'{prefix}.chain'] = 'auto'\n",
    "inputs[f'{prefix}.dockerRegistry'] = common_docker_registry\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f54b5e4-6ff1-4469-8c16-cc975d71766d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "829ad858-65ca-49f8-9b96-2b428e2a25f9",
   "metadata": {},
   "source": [
    "## Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ed61284c-ec00-4b2e-9522-f20ee0a8682c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              pipelineType project sample owner destination  \\\n",
       "SD-1680_Patient_D_nucseq_GE            NaN     NaN    NaN   NaN         NaN   \n",
       "SD-1680_Patient_F_nucseq_GE            NaN     NaN    NaN   NaN         NaN   \n",
       "SD-1680_Patient_I_nucseq_GE            NaN     NaN    NaN   NaN         NaN   \n",
       "SD-1680_Patient_D_nucseq_2_GE          NaN     NaN    NaN   NaN         NaN   \n",
       "SD-1680_Patient_F_nucseq_2_GE          NaN     NaN    NaN   NaN         NaN   \n",
       "SD-1680_Patient_I_nucseq_2_GE          NaN     NaN    NaN   NaN         NaN   \n",
       "\n",
       "                              transfer comment  \n",
       "SD-1680_Patient_D_nucseq_GE        NaN     NaN  \n",
       "SD-1680_Patient_F_nucseq_GE        NaN     NaN  \n",
       "SD-1680_Patient_I_nucseq_GE        NaN     NaN  \n",
       "SD-1680_Patient_D_nucseq_2_GE      NaN     NaN  \n",
       "SD-1680_Patient_F_nucseq_2_GE      NaN     NaN  \n",
       "SD-1680_Patient_I_nucseq_2_GE      NaN     NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(f\"{config_dir}/template.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b9c33977-ad27-4c05-864f-08652d001f5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_D_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_F_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_I_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_D_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_F_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_I_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                pipelineType           project  \\\n",
       "SD-1680_Patient_D_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_F_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_I_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_D_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_F_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_I_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "\n",
       "                                                      sample    owner  \\\n",
       "SD-1680_Patient_D_nucseq_GE      SD-1680_Patient_D_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_F_nucseq_GE      SD-1680_Patient_F_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_I_nucseq_GE      SD-1680_Patient_I_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_D_nucseq_2_GE  SD-1680_Patient_D_nucseq_2_GE  sohailn   \n",
       "SD-1680_Patient_F_nucseq_2_GE  SD-1680_Patient_F_nucseq_2_GE  sohailn   \n",
       "SD-1680_Patient_I_nucseq_2_GE  SD-1680_Patient_I_nucseq_2_GE  sohailn   \n",
       "\n",
       "                                                           destination  \\\n",
       "SD-1680_Patient_D_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "\n",
       "                              transfer  comment  \n",
       "SD-1680_Patient_D_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_F_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_I_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_D_nucseq_2_GE        -  sohailn  \n",
       "SD-1680_Patient_F_nucseq_2_GE        -  sohailn  \n",
       "SD-1680_Patient_I_nucseq_2_GE        -  sohailn  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples[\"Sample_ID\"].apply(lambda x: get_project_id(x, creds))\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['S3_Path'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88ad7ccd-f144-4696-bc2b-e7945059103f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "524a26f6-fe12-4065-8c79-b665c99336a3",
   "metadata": {},
   "source": [
    "## Execution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403c3948-bcb5-45a1-b6f5-dfa8aefb4e86",
   "metadata": {},
   "source": [
    "Make sure to check that your FASTQ files aren't archived, otherwise your job will fail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "78242f27-c831-4d7f-b4e3-14dfdb470543",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellRangerVdj.sampleName</th>\n",
       "      <th>CellRangerVdj.fastqNames</th>\n",
       "      <th>CellRangerVdj.referenceGenome</th>\n",
       "      <th>CellRangerVdj.inputFastq</th>\n",
       "      <th>CellRangerVdj.chain</th>\n",
       "      <th>CellRangerVdj.dockerRegistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_D_nucseq_GE</td>\n",
       "      <td>3808_SD-1680_Patient_D_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_F_nucseq_GE</td>\n",
       "      <td>3809_SD-1680_Patient_F_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>SD-1680_Patient_I_nucseq_GE</td>\n",
       "      <td>3810_SD-1680_Patient_I_nucseq_GE_IGO...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_D_nucseq_2_GE</td>\n",
       "      <td>3811_SD-1680_Patient_D_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_F_nucseq_2_GE</td>\n",
       "      <td>3812_SD-1680_Patient_F_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>SD-1680_Patient_I_nucseq_2_GE</td>\n",
       "      <td>3813_SD-1680_Patient_I_nucseq_2_GE_I...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/VanD...</td>\n",
       "      <td>auto</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    CellRangerVdj.sampleName  \\\n",
       "SD-1680_Patient_D_nucseq_GE      SD-1680_Patient_D_nucseq_GE   \n",
       "SD-1680_Patient_F_nucseq_GE      SD-1680_Patient_F_nucseq_GE   \n",
       "SD-1680_Patient_I_nucseq_GE      SD-1680_Patient_I_nucseq_GE   \n",
       "SD-1680_Patient_D_nucseq_2_GE  SD-1680_Patient_D_nucseq_2_GE   \n",
       "SD-1680_Patient_F_nucseq_2_GE  SD-1680_Patient_F_nucseq_2_GE   \n",
       "SD-1680_Patient_I_nucseq_2_GE  SD-1680_Patient_I_nucseq_2_GE   \n",
       "\n",
       "                                              CellRangerVdj.fastqNames  \\\n",
       "SD-1680_Patient_D_nucseq_GE    3808_SD-1680_Patient_D_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_F_nucseq_GE    3809_SD-1680_Patient_F_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_I_nucseq_GE    3810_SD-1680_Patient_I_nucseq_GE_IGO...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  3811_SD-1680_Patient_D_nucseq_2_GE_I...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  3812_SD-1680_Patient_F_nucseq_2_GE_I...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  3813_SD-1680_Patient_I_nucseq_2_GE_I...   \n",
       "\n",
       "                              CellRangerVdj.referenceGenome  \\\n",
       "SD-1680_Patient_D_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_F_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_I_nucseq_GE                          GRCh38   \n",
       "SD-1680_Patient_D_nucseq_2_GE                        GRCh38   \n",
       "SD-1680_Patient_F_nucseq_2_GE                        GRCh38   \n",
       "SD-1680_Patient_I_nucseq_2_GE                        GRCh38   \n",
       "\n",
       "                                              CellRangerVdj.inputFastq  \\\n",
       "SD-1680_Patient_D_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_F_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_I_nucseq_GE    [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  [s3://dp-lab-data/collaborators/VanD...   \n",
       "\n",
       "                              CellRangerVdj.chain CellRangerVdj.dockerRegistry  \n",
       "SD-1680_Patient_D_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_F_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_I_nucseq_GE                  auto              quay.io/hisplan  \n",
       "SD-1680_Patient_D_nucseq_2_GE                auto              quay.io/hisplan  \n",
       "SD-1680_Patient_F_nucseq_2_GE                auto              quay.io/hisplan  \n",
       "SD-1680_Patient_I_nucseq_2_GE                auto              quay.io/hisplan  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bfe55491-699a-4db9-aca3-bf6cf337086c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pipelineType</th>\n",
       "      <th>project</th>\n",
       "      <th>sample</th>\n",
       "      <th>owner</th>\n",
       "      <th>destination</th>\n",
       "      <th>transfer</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_D_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_F_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_I_nucseq_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_D_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_D_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_F_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_F_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SD-1680_Patient_I_nucseq_2_GE</th>\n",
       "      <td>CellRangerVdj</td>\n",
       "      <td>AML T cell pilot</td>\n",
       "      <td>SD-1680_Patient_I_nucseq_2_GE</td>\n",
       "      <td>sohailn</td>\n",
       "      <td>s3://dp-lab-data/collaborators/VanDe...</td>\n",
       "      <td>-</td>\n",
       "      <td>sohailn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                pipelineType           project  \\\n",
       "SD-1680_Patient_D_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_F_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_I_nucseq_GE    CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_D_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_F_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "SD-1680_Patient_I_nucseq_2_GE  CellRangerVdj  AML T cell pilot   \n",
       "\n",
       "                                                      sample    owner  \\\n",
       "SD-1680_Patient_D_nucseq_GE      SD-1680_Patient_D_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_F_nucseq_GE      SD-1680_Patient_F_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_I_nucseq_GE      SD-1680_Patient_I_nucseq_GE  sohailn   \n",
       "SD-1680_Patient_D_nucseq_2_GE  SD-1680_Patient_D_nucseq_2_GE  sohailn   \n",
       "SD-1680_Patient_F_nucseq_2_GE  SD-1680_Patient_F_nucseq_2_GE  sohailn   \n",
       "SD-1680_Patient_I_nucseq_2_GE  SD-1680_Patient_I_nucseq_2_GE  sohailn   \n",
       "\n",
       "                                                           destination  \\\n",
       "SD-1680_Patient_D_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_GE    s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_D_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_F_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "SD-1680_Patient_I_nucseq_2_GE  s3://dp-lab-data/collaborators/VanDe...   \n",
       "\n",
       "                              transfer  comment  \n",
       "SD-1680_Patient_D_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_F_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_I_nucseq_GE          -  sohailn  \n",
       "SD-1680_Patient_D_nucseq_2_GE        -  sohailn  \n",
       "SD-1680_Patient_F_nucseq_2_GE        -  sohailn  \n",
       "SD-1680_Patient_I_nucseq_2_GE        -  sohailn  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a9a9d00a-9373-4cb9-bb37-102d148998cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd58a333e16b4889acceae3156685a1d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())\n",
    "\n",
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(inputs.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = \"submit.sh\",\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1504efd7-5a1f-4213-a9a4-cde5c81706f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_2_GE.inputs.json\n",
      "/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_2_GE.labels.json\n"
     ]
    }
   ],
   "source": [
    "print(path_to_inputs)\n",
    "print(path_to_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1e933029-8a8c-4ff9-92e8-2b1312bb4e98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_D_nucseq_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_D_nucseq_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"f6e876d3-7521-4a38-8ab4-b27da700482a\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_F_nucseq_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_F_nucseq_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"0a17bb98-2945-41b3-870c-ecce71817d45\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"663f4190-6ce0-48ba-b27b-e08700584c7a\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_D_nucseq_2_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_D_nucseq_2_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"ddd92731-b99d-4bb1-9d3c-571877e84b5a\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_F_nucseq_2_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_F_nucseq_2_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"e078f4bb-8bcc-4e1f-99fd-731be185ffcd\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''},\n",
       " {'args': ['/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/submit.sh',\n",
       "   '-k',\n",
       "   '/Users/sohailn/.cromwell/cromwell-secrets.json',\n",
       "   '-i',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_2_GE.inputs.json',\n",
       "   '-l',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/configs/SD-1680_Patient_I_nucseq_2_GE.labels.json',\n",
       "   '-o',\n",
       "   '/Users/sohailn/scing/bin/cellranger-vdj-6.1.2/CellRangerVdj.options.aws.json'],\n",
       "  'returncode': 0,\n",
       "  'stdout': '{\"id\":\"3c1f3d6c-9c99-414a-ad90-9f23a856b338\",\"status\":\"Submitted\"}\\n',\n",
       "  'stderr': ''}]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stdouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d21cf2-967a-488e-b009-d838ca77d277",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ddfdb475-ceec-4235-afa1-4c23d48d523c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_D_nucseq_GE/cr-vdj-results',\n",
       " 's3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_F_nucseq_GE/cr-vdj-results',\n",
       " 's3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_I_nucseq_GE/cr-vdj-results',\n",
       " 's3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_D_nucseq_2_GE/cr-vdj-results',\n",
       " 's3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_F_nucseq_2_GE/cr-vdj-results',\n",
       " 's3://dp-lab-data/collaborators/VanDenBrink/AmlCellPilot/SD-1680_Patient_I_nucseq_2_GE/cr-vdj-results']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels['destination'].values.tolist()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
