{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d13e6985-95f3-45eb-8162-0e24a2a049d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import re, subprocess, boto3, json, shlex, mysql, os, urllib, logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from s3path import S3Path\n",
    "from pathlib import Path\n",
    "from tqdm.notebook import tqdm\n",
    "from packaging import version\n",
    "\n",
    "import glob\n",
    "import os\n",
    "\n",
    "pd.set_option(\"display.max_colwidth\", 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6547d881-c5e0-494d-9189-0059c14e126e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fb891f-0934-4489-98fb-4713ab904a5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bf6e1460-34e1-4654-9fa5-2e82eb23902b",
   "metadata": {},
   "source": [
    "# AWS setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7052097b-da4a-444c-aebb-1b585266c353",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: AWS_ACCESS_KEY_ID=ASIAYPY66CWMAGM24WHE\n",
      "env: AWS_SECRET_ACCESS_KEY=jWUmwaWT+IQhV+71PWwfi1We6vSusB78GBwNw15e\n",
      "env: AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEHIaCXVzLWVhc3QtMSJIMEYCIQC5kbf1i5ib760fqEmYsdjvX/g7ogw0RaGEz2BUIy+R/wIhAOi31VVCUdy7zmAZddGoXvKRm7JtoJzwErPPPRahY4l/KvgBCNv//////////wEQAxoMNTgzNjQzNTY3NTEyIgx4r2eNH51B/0SJBpAqzAEM95AVY9oz0nVlU4hjvRWFYWEUuXzOkPwo6RagXIByGUj0SP4HWQX0kvraLFqa34JYyijiRhgoPu7xHiGOjdHuS93qSG4mSQhLDgMTa+K7ze+8I407NeMlNxiiS5KvKCVaTt6U76rBe4kGauNcMSqT7sIbpO4btLXtwI41CnqbbXKTgeJc87BwxcvWUqW8WE9D61uZ0Sv2qmvGoL+jkFOfOGw/bMSnMOtNtWtcYHOEFI0WW3i1DeTamyOoiJ6ZfYObSQ1V+SNEZNYhW64wwo+7ngY6lwEZ//N+c9HWTzOSb9crQjpV+hRMjDy1O+FLXll70i5dHZe5RxsmFfm8PYjYaz4AZQ6Tl716sEm25s1CCbTiYx3x2OeGIuj7XLbeE8lNDqDemhlAdkhXjuCUDFk/wAz+z4rwqu3PvsA0YB+Ut2jG/l+1RYzfaXBgSWxbl6c1oJuDz5ngUewaTQGDRFomu7/u8dCN6j89zovp\n",
      "2021-10-07 15:31:32 agc-583643567512-us-east-1\n",
      "2021-10-07 15:28:07 cdktoolkit-stagingbucket-d49u1xfb0sc4\n",
      "2019-10-09 13:04:06 cf-templates-umiwbnq3566w-us-east-1\n",
      "2019-10-10 12:46:54 dp-daily-reports\n",
      "2017-09-26 10:15:15 dp-lab-data\n",
      "2019-05-23 12:34:58 dp-lab-data-public\n",
      "2018-03-01 13:29:07 dp-lab-glacier\n",
      "2020-11-06 23:02:27 dp-lab-gwf-core\n",
      "2021-07-15 18:14:01 dp-lab-gwf-core2\n",
      "2017-09-26 10:17:18 dp-lab-home\n",
      "2019-08-14 16:42:43 dp-lab-test\n",
      "2019-04-25 12:35:35 elasticbeanstalk-us-east-1-583643567512\n",
      "2019-12-13 15:53:57 scri-computational\n",
      "2017-09-26 10:23:50 seqc-public\n"
     ]
    }
   ],
   "source": [
    "# Load aws\n",
    "\n",
    "%env AWS_ACCESS_KEY_ID=ASIAYPY66CWMAGM24WHE\n",
    "%env AWS_SECRET_ACCESS_KEY=jWUmwaWT+IQhV+71PWwfi1We6vSusB78GBwNw15e\n",
    "%env AWS_SESSION_TOKEN=IQoJb3JpZ2luX2VjEHIaCXVzLWVhc3QtMSJIMEYCIQC5kbf1i5ib760fqEmYsdjvX/g7ogw0RaGEz2BUIy+R/wIhAOi31VVCUdy7zmAZddGoXvKRm7JtoJzwErPPPRahY4l/KvgBCNv//////////wEQAxoMNTgzNjQzNTY3NTEyIgx4r2eNH51B/0SJBpAqzAEM95AVY9oz0nVlU4hjvRWFYWEUuXzOkPwo6RagXIByGUj0SP4HWQX0kvraLFqa34JYyijiRhgoPu7xHiGOjdHuS93qSG4mSQhLDgMTa+K7ze+8I407NeMlNxiiS5KvKCVaTt6U76rBe4kGauNcMSqT7sIbpO4btLXtwI41CnqbbXKTgeJc87BwxcvWUqW8WE9D61uZ0Sv2qmvGoL+jkFOfOGw/bMSnMOtNtWtcYHOEFI0WW3i1DeTamyOoiJ6ZfYObSQ1V+SNEZNYhW64wwo+7ngY6lwEZ//N+c9HWTzOSb9crQjpV+hRMjDy1O+FLXll70i5dHZe5RxsmFfm8PYjYaz4AZQ6Tl716sEm25s1CCbTiYx3x2OeGIuj7XLbeE8lNDqDemhlAdkhXjuCUDFk/wAz+z4rwqu3PvsA0YB+Ut2jG/l+1RYzfaXBgSWxbl6c1oJuDz5ngUewaTQGDRFomu7/u8dCN6j89zovp\n",
    "!aws s3 ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "131f63b2-9e4c-4b36-b9b2-e6a6693dd4a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "087ae658-2d9d-4b3f-b159-ccb06f4c5626",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "931b1af4-a619-4308-848b-6a55bc65072e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which pipeline are you running\n",
    "\n",
    "prefix = \"CellRangerArc\" # Workflow to run; also .wdl filename prefix\n",
    "output_dirname = \"cr-arc-results\"\n",
    "\n",
    "workflow_dir = glob.glob(f\"{Path.home()}/scing/bin/cellranger-arc-*\")[0]\n",
    "path_to_exec = f\"{workflow_dir}/submit.sh\" # CHANGE THIS FOR SHARP\n",
    "\n",
    "# Locations of workflow-related directories and files\n",
    "path_to_cromwell_secrets = f\"{Path.home()}/.cromwell/cromwell-secrets.json\"\n",
    "db_credentials_path = f\"{Path.home()}/.config.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aaee197b-c610-46a0-b786-9bfc6aafa07d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of docker files\n",
    "common_docker_registry = \"quay.io/hisplan\"\n",
    "pipeline_type = prefix # field in *.labels.json\n",
    "comment = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "963b6641-1c6d-430f-99f4-4077b24fa1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Workflow file paths\n",
    "config_dir = f\"{workflow_dir}/configs\"\n",
    "path_to_options = f\"{workflow_dir}/{prefix}.options.aws.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "809159de-a318-434b-8d3d-f660573a3990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set credentials based on SCRIdb CLI config file\n",
    "with open(db_credentials_path) as f:\n",
    "    creds = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f359259-186e-4beb-92c8-bce6b6b03fd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e652e114-578e-4a34-8624-bf001533af26",
   "metadata": {},
   "source": [
    "# Sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b47a3ba3-ee28-4a98-afff-2b90cd10822c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AWS_storage</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>sc_tech</th>\n",
       "      <th>project_id</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RB-2041_WildType_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4440</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>GRCh38-1.1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RB-2041_mRB54_1003_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4441</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>GRCh38-1.1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         AWS_storage    id  \\\n",
       "Sample                                                                       \n",
       "RB-2041_WildType_DOGMAseq    s3://dp-lab-data/collaborators/sfeir...  4440   \n",
       "RB-2041_mRB54_1003_DOGMAseq  s3://dp-lab-data/collaborators/sfeir...  4441   \n",
       "\n",
       "                            species     sc_tech        project_id  \\\n",
       "Sample                                                              \n",
       "RB-2041_WildType_DOGMAseq     human  10X_scATAC  scATAC-seq pilot   \n",
       "RB-2041_mRB54_1003_DOGMAseq   human  10X_scATAC  scATAC-seq pilot   \n",
       "\n",
       "                                reference  \n",
       "Sample                                     \n",
       "RB-2041_WildType_DOGMAseq    GRCh38-1.1.0  \n",
       "RB-2041_mRB54_1003_DOGMAseq  GRCh38-1.1.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can supply a list of IDs or request IDs or even AWS paths\n",
    "\n",
    "# request_ids = ['PM-1779']\n",
    "# samples = sample_scridb_info(request_ids, 'request_id', creds)\n",
    "\n",
    "# sample_ids = [4138]\n",
    "# samples = sample_scridb_info(sample_ids, 'id', creds)\n",
    "\n",
    "aws_storage = ['s3://dp-lab-data/collaborators/sfeira/ScatacSeqPilot/RB-2041_mRB54_1003_DOGMAseq/',\n",
    "              's3://dp-lab-data/collaborators/sfeira/ScatacSeqPilot/RB-2041_WildType_DOGMAseq/']\n",
    "samples = sample_scridb_info(aws_storage, 'AWS_storage', creds)\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "91a0a018-06ff-4b7e-92a9-46be980ca402",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s3://dp-lab-data/SCRI_Projects/10XFixedScrnaseqPilot/TX-1886_D34M_FAfixed_scRNA']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['AWS_storage'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66be82d8-2707-44ae-b7ac-7c14e7525a0f",
   "metadata": {},
   "source": [
    "<b>IMPORTANT NOTE</b> \n",
    "\n",
    "Check the name of the folder you are running. Typically the folder that is stored in the database is just the GEX. So if another library is generated (multiome ATAC, VDJ, hashtag, etc) then it needs to be manually changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b851f925-3a5a-4cec-9ef3-4bdc53eb295b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RB-2041_WildType_DOGMAseq\n",
      "                           PRE RB-2041_WildType_DOGMAseq/\n",
      "\n",
      "RB-2041_mRB54_1003_DOGMAseq\n",
      "                           PRE RB-2041_mRB54_1003_DOGMAseq/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Check the name of the folder you are running\n",
    "# Especially if there are multiple libraries (i.e ATAC, TCR_VDJ, etc.)\n",
    "\n",
    "for sample, row in samples.iterrows():\n",
    "    print(sample)\n",
    "    path = os.path.split(row['AWS_storage'])[0] + '/'\n",
    "    os.system(f'aws s3 ls {path} | grep {sample}')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b09d8d6-46ef-403f-816b-13f34a603e9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AWS_storage</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>sc_tech</th>\n",
       "      <th>project_id</th>\n",
       "      <th>reference</th>\n",
       "      <th>FASTQs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RB-2041_WildType_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4440</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>GRCh38-1.1.0</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RB-2041_mRB54_1003_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4441</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>GRCh38-1.1.0</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         AWS_storage    id  \\\n",
       "Sample                                                                       \n",
       "RB-2041_WildType_DOGMAseq    s3://dp-lab-data/collaborators/sfeir...  4440   \n",
       "RB-2041_mRB54_1003_DOGMAseq  s3://dp-lab-data/collaborators/sfeir...  4441   \n",
       "\n",
       "                            species     sc_tech        project_id  \\\n",
       "Sample                                                              \n",
       "RB-2041_WildType_DOGMAseq     human  10X_scATAC  scATAC-seq pilot   \n",
       "RB-2041_mRB54_1003_DOGMAseq   human  10X_scATAC  scATAC-seq pilot   \n",
       "\n",
       "                                reference  \\\n",
       "Sample                                      \n",
       "RB-2041_WildType_DOGMAseq    GRCh38-1.1.0   \n",
       "RB-2041_mRB54_1003_DOGMAseq  GRCh38-1.1.0   \n",
       "\n",
       "                                                              FASTQs  \n",
       "Sample                                                                \n",
       "RB-2041_WildType_DOGMAseq    {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "RB-2041_mRB54_1003_DOGMAseq  {'I1': ['s3://dp-lab-data/collaborat...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[\"FASTQs\"] = samples[\"AWS_storage\"].apply(lambda x: get_fastqs(x, fastq_map[prefix], \"FASTQ\"))\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2861ee5a-1a3c-4774-9eda-28ac92cb3df7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "563dd74c-53ef-4595-8400-b28c23f52085",
   "metadata": {},
   "source": [
    "<b>IMPORTANT NOTE</b>\n",
    "\n",
    "Make sure that your files are not archived. The following command will print any FASTQ file that is archived. Unarchive the files and then come back to processing the sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "569312fd-750c-4356-8a73-1feff86601e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RB-2041_WildType_DOGMAseq\n",
      "RB-2041_mRB54_1003_DOGMAseq\n"
     ]
    }
   ],
   "source": [
    "for sample, row in samples.iterrows():\n",
    "    print(sample)\n",
    "    fastqs = np.ravel(list(row['FASTQs'].values()))\n",
    "    dirnames = set([os.path.dirname(x) for x in fastqs])\n",
    "    \n",
    "    for dirname in dirnames:\n",
    "        file = dirname.replace('s3://', '')\n",
    "        \n",
    "        bucket = file.split('/')[0]\n",
    "        pre = file.replace(f'{bucket}/', '')\n",
    "        \n",
    "        !aws s3api list-objects-v2 --bucket $bucket --prefix $pre --query \"Contents[?StorageClass!='STANDARD'].Key\" --output text "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "534b90a7-ab25-462c-9ae4-40c4294a9fd5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3f0f3377-1c19-4545-868f-5da05f0ea295",
   "metadata": {},
   "source": [
    "<b>IMPORTANT NOTE</b> \n",
    "\n",
    "For CellRanger you need to supply an HTTPS path. So if you are using a custom genome stored on AWS, you must make the reference <b> public </b>! Be sure to manually change the \"reference\" argument if it has not been updated correctly!!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5586ef8-e219-4388-b5f1-646e8d652e1e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b8007050-06f2-4325-ad13-68bb986b626c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AWS_storage</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>sc_tech</th>\n",
       "      <th>project_id</th>\n",
       "      <th>reference</th>\n",
       "      <th>FASTQs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RB-2041_WildType_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4440</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>https://cf.10xgenomics.com/supp/cell...</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RB-2041_mRB54_1003_DOGMAseq</th>\n",
       "      <td>s3://dp-lab-data/collaborators/sfeir...</td>\n",
       "      <td>4441</td>\n",
       "      <td>human</td>\n",
       "      <td>10X_scATAC</td>\n",
       "      <td>scATAC-seq pilot</td>\n",
       "      <td>https://cf.10xgenomics.com/supp/cell...</td>\n",
       "      <td>{'I1': ['s3://dp-lab-data/collaborat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         AWS_storage    id  \\\n",
       "Sample                                                                       \n",
       "RB-2041_WildType_DOGMAseq    s3://dp-lab-data/collaborators/sfeir...  4440   \n",
       "RB-2041_mRB54_1003_DOGMAseq  s3://dp-lab-data/collaborators/sfeir...  4441   \n",
       "\n",
       "                            species     sc_tech        project_id  \\\n",
       "Sample                                                              \n",
       "RB-2041_WildType_DOGMAseq     human  10X_scATAC  scATAC-seq pilot   \n",
       "RB-2041_mRB54_1003_DOGMAseq   human  10X_scATAC  scATAC-seq pilot   \n",
       "\n",
       "                                                           reference  \\\n",
       "Sample                                                                 \n",
       "RB-2041_WildType_DOGMAseq    https://cf.10xgenomics.com/supp/cell...   \n",
       "RB-2041_mRB54_1003_DOGMAseq  https://cf.10xgenomics.com/supp/cell...   \n",
       "\n",
       "                                                              FASTQs  \n",
       "Sample                                                                \n",
       "RB-2041_WildType_DOGMAseq    {'I1': ['s3://dp-lab-data/collaborat...  \n",
       "RB-2041_mRB54_1003_DOGMAseq  {'I1': ['s3://dp-lab-data/collaborat...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = update_ref(samples, prefix)\n",
    "\n",
    "if not samples['reference'].isna().any():\n",
    "    samples[\"reference\"].apply(lambda x: {\n",
    "        \"name\": re.match(r'.*refdata-cellranger-arc-(.*).tar.gz', x)[1],\n",
    "        \"location\": x,\n",
    "    }) \n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6b56024e-a4c1-488f-861a-17bf9c191223",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample\n",
       "RB-2041_WildType_DOGMAseq      {'name': 'human-rna-mitoblacklist', ...\n",
       "RB-2041_mRB54_1003_DOGMAseq    {'name': 'human-rna-mitoblacklist', ...\n",
       "Name: reference, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['reference'] = [\n",
    "    {\n",
    "        'name' : 'human-rna-mitoblacklist',\n",
    "        'location' : 's3://dp-lab-data/collaborators/sfeira/ScatacSeqPilot/human-rna-mitoblacklist/human-rna-mitoblacklist.tar.gz'\n",
    "    }] * len(samples)\n",
    "samples['reference']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83ed197c-562c-459e-b113-83fee86388ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5e6b5cd2-ca13-4ce8-808a-5404ede8e528",
   "metadata": {},
   "source": [
    "# Generate inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0561da32-5979-4b0e-a40e-35af26218060",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellRangerArc.runID</th>\n",
       "      <th>CellRangerArc.gexFastqName</th>\n",
       "      <th>CellRangerArc.gexFastqFiles</th>\n",
       "      <th>CellRangerArc.atacFastqName</th>\n",
       "      <th>CellRangerArc.atacFastqFiles</th>\n",
       "      <th>CellRangerArc.reference</th>\n",
       "      <th>CellRangerArc.dockerRegistry</th>\n",
       "      <th>CellRangerArc.sampleName</th>\n",
       "      <th>CellRangerArc.fastqFiles</th>\n",
       "      <th>CellRangerArc.fastqNames</th>\n",
       "      <th>CellRangerArc.referenceGenome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RB-2041_WildType_DOGMAseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "      <td>RB-2041_WildType_DOGMAseq</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/sfei...</td>\n",
       "      <td>4440_RB-2041_WildType_DOGMAseq_IGO_1...</td>\n",
       "      <td>{'name': 'human-rna-mitoblacklist', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RB-2041_mRB54_1003_DOGMAseq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>quay.io/hisplan</td>\n",
       "      <td>RB-2041_mRB54_1003_DOGMAseq</td>\n",
       "      <td>[s3://dp-lab-data/collaborators/sfei...</td>\n",
       "      <td>4441_RB-2041_mRB54_1003_DOGMAseq_IGO...</td>\n",
       "      <td>{'name': 'human-rna-mitoblacklist', ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            CellRangerArc.runID CellRangerArc.gexFastqName  \\\n",
       "Sample                                                                       \n",
       "RB-2041_WildType_DOGMAseq                   NaN                        NaN   \n",
       "RB-2041_mRB54_1003_DOGMAseq                 NaN                        NaN   \n",
       "\n",
       "                            CellRangerArc.gexFastqFiles  \\\n",
       "Sample                                                    \n",
       "RB-2041_WildType_DOGMAseq                           NaN   \n",
       "RB-2041_mRB54_1003_DOGMAseq                         NaN   \n",
       "\n",
       "                            CellRangerArc.atacFastqName  \\\n",
       "Sample                                                    \n",
       "RB-2041_WildType_DOGMAseq                           NaN   \n",
       "RB-2041_mRB54_1003_DOGMAseq                         NaN   \n",
       "\n",
       "                            CellRangerArc.atacFastqFiles  \\\n",
       "Sample                                                     \n",
       "RB-2041_WildType_DOGMAseq                            NaN   \n",
       "RB-2041_mRB54_1003_DOGMAseq                          NaN   \n",
       "\n",
       "                            CellRangerArc.reference  \\\n",
       "Sample                                                \n",
       "RB-2041_WildType_DOGMAseq                       NaN   \n",
       "RB-2041_mRB54_1003_DOGMAseq                     NaN   \n",
       "\n",
       "                            CellRangerArc.dockerRegistry  \\\n",
       "Sample                                                     \n",
       "RB-2041_WildType_DOGMAseq                quay.io/hisplan   \n",
       "RB-2041_mRB54_1003_DOGMAseq              quay.io/hisplan   \n",
       "\n",
       "                                CellRangerArc.sampleName  \\\n",
       "Sample                                                     \n",
       "RB-2041_WildType_DOGMAseq      RB-2041_WildType_DOGMAseq   \n",
       "RB-2041_mRB54_1003_DOGMAseq  RB-2041_mRB54_1003_DOGMAseq   \n",
       "\n",
       "                                            CellRangerArc.fastqFiles  \\\n",
       "Sample                                                                 \n",
       "RB-2041_WildType_DOGMAseq    [s3://dp-lab-data/collaborators/sfei...   \n",
       "RB-2041_mRB54_1003_DOGMAseq  [s3://dp-lab-data/collaborators/sfei...   \n",
       "\n",
       "                                            CellRangerArc.fastqNames  \\\n",
       "Sample                                                                 \n",
       "RB-2041_WildType_DOGMAseq    4440_RB-2041_WildType_DOGMAseq_IGO_1...   \n",
       "RB-2041_mRB54_1003_DOGMAseq  4441_RB-2041_mRB54_1003_DOGMAseq_IGO...   \n",
       "\n",
       "                                       CellRangerArc.referenceGenome  \n",
       "Sample                                                                \n",
       "RB-2041_WildType_DOGMAseq    {'name': 'human-rna-mitoblacklist', ...  \n",
       "RB-2041_mRB54_1003_DOGMAseq  {'name': 'human-rna-mitoblacklist', ...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standard inputs, modify fields as need be\n",
    "\n",
    "# Load minimum inputs and labels fields from templates\n",
    "with open(f\"{config_dir}/template.inputs.json\") as f:\n",
    "    std_inputs_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "inputs = pd.DataFrame(index=samples.index, columns=std_inputs_fields,)\n",
    "\n",
    "# Annotate inputs\n",
    "inputs[f\"{prefix}.sampleName\"] = samples.index\n",
    "inputs[f\"{prefix}.fastqFiles\"] = samples[\"FASTQs\"].apply(lambda x: np.ravel(list(x.values())))\n",
    "inputs[f\"{prefix}.fastqNames\"] = inputs[f\"{prefix}.fastqFiles\"].apply(lambda x: get_fastqs_name(x))\n",
    "inputs[f\"{prefix}.referenceGenome\"] = samples[\"reference\"]                        \n",
    "inputs[f\"{prefix}.dockerRegistry\"] = common_docker_registry\n",
    "\n",
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4700f244-fe4b-44a5-9d1b-fbc1b2b35afb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CellRangerArc.runID',\n",
       " 'CellRangerArc.gexFastqName',\n",
       " 'CellRangerArc.gexFastqFiles',\n",
       " 'CellRangerArc.atacFastqName',\n",
       " 'CellRangerArc.atacFastqFiles',\n",
       " 'CellRangerArc.reference',\n",
       " 'CellRangerArc.dockerRegistry']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "std_inputs_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f37043a-c883-4da4-9c87-f1912b102e91",
   "metadata": {},
   "source": [
    "# Generate labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81500471-4a2e-4eae-8b69-174608902709",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard labels, modify fields as need be\n",
    "\n",
    "with open(f\"{config_dir}/template.labels.json\") as f:\n",
    "    std_labels_fields = list(json.load(f).keys())\n",
    "    \n",
    "# Annotate all samples with workflow inputs and labels\n",
    "labels = pd.DataFrame(index=samples.index, columns=std_labels_fields,)\n",
    "\n",
    "labels[\"pipelineType\"] = pipeline_type\n",
    "labels[\"project\"] = samples['project_id']\n",
    "labels[\"sample\"] = labels.index\n",
    "labels[\"owner\"] = creds[\"user\"]\n",
    "labels[\"destination\"] = samples['AWS_storage'] + \"/\" + output_dirname\n",
    "labels[\"transfer\"] = \"-\"\n",
    "labels[\"comment\"] = creds[\"user\"]\n",
    "\n",
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688fe090-52a7-493e-bf52-617b91809c5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d7e3422e-a130-4b5b-ba68-f1d59921388c",
   "metadata": {},
   "source": [
    "# Run samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b231634-b533-4b48-afc3-f6c201bf3df7",
   "metadata": {},
   "source": [
    "Look over the samples before submitting one last time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71f5101-edb7-4c9d-ad07-408e756bd197",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61801e60-ab82-4012-afcc-22ee288b22d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8057ea75-3c2f-40c9-b6fe-d957dd401504",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf7b481-ec81-46ff-9452-d7d831972764",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "assert (std_inputs_fields == list(inputs.columns)) & (inputs.notna().values.all())\n",
    "assert (std_labels_fields == list(labels.columns)) & (labels.notna().values.all())\n",
    "\n",
    "stdouts = [] # to store all outputs\n",
    "process = True\n",
    "\n",
    "with tqdm(inputs.index) as t:\n",
    "\n",
    "    for sample_name in t:\n",
    "\n",
    "        # Write inputs and labels to file\n",
    "        path_to_inputs = f\"{config_dir}/{sample_name}.inputs.json\"\n",
    "        with open(path_to_inputs, \"w\") as f_inputs:\n",
    "            json.dump(inputs.loc[sample_name].to_dict(), f_inputs, indent=4, cls=NpEncoder)\n",
    "\n",
    "        path_to_labels = f\"{config_dir}/{sample_name}.labels.json\"\n",
    "        with open(path_to_labels, \"w\") as f_labels:\n",
    "            json.dump(labels.loc[sample_name].to_dict(), f_labels, indent=4, cls=NpEncoder)\n",
    "\n",
    "        if process:\n",
    "            stdouts.append(run(\n",
    "                workflow_path = workflow_dir,\n",
    "                execp = \"submit.sh\",\n",
    "                secrets = path_to_cromwell_secrets,\n",
    "                inputs = path_to_inputs,\n",
    "                labels = path_to_labels,\n",
    "                options = path_to_options,\n",
    "            ))\n",
    "            \n",
    "        time.sleep(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1efc63e0-a3bf-42d5-83d2-3392dc6abb20",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels['destination'].values.tolist()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
